{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "843ff879",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "from collections import Counter\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "from scipy.stats import fisher_exact\n",
    "from scipy import stats\n",
    "from scipy.stats import mannwhitneyu\n",
    "from scipy.stats import ttest_ind\n",
    "from scipy.stats import kruskal\n",
    "import scipy\n",
    "import matplotlib\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import matplotlib.patches as mpatches\n",
    "import copy\n",
    "from matplotlib import gridspec\n",
    "#import matplotlib.colors as colors\n",
    "from matplotlib.patches import ConnectionPatch\n",
    "import matplotlib.ticker\n",
    "from statsmodels.stats.anova import anova_lm\n",
    "from itertools import product"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2294ec5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### To combine rows\n",
    "def cat_rows(series):\n",
    "    return ','.join(list(series.astype(\"string\")))\n",
    "\n",
    "### Identify high VAF gene mutation in a participant\n",
    "def gene_high_vaf(row):\n",
    "    \n",
    "    if pd.isnull(row[\"vaf\"]) == False:\n",
    "        if \",\" in row[\"vaf\"]:\n",
    "            v=np.array(row[\"vaf\"].split(\",\")).astype(float)\n",
    "            all_genes_present=np.array(row[\"symbol\"].split(\",\"))\n",
    "            p_mut_all=np.array(row[\"p_mut\"].split(\",\"))\n",
    "            maxind=np.argmax(v)\n",
    "            if np.unique(all_genes_present).shape[0] == 1:\n",
    "                gene_vaf=[all_genes_present[maxind],np.max(v),p_mut_all[maxind],all_genes_present[maxind],np.max(v),p_mut_all[maxind]]\n",
    "            else:\n",
    "                gene_vaf=[\"NA\",\"NA\",\"NA\",all_genes_present[maxind],np.max(v),p_mut_all[maxind]]\n",
    "        else:\n",
    "            gene_vaf=[row[\"symbol\"],float(row[\"vaf\"]),row[\"p_mut\"],row[\"symbol\"],float(row[\"vaf\"]),row[\"p_mut\"]]\n",
    "    else:\n",
    "        gene_vaf=[\"Control\",\"NA\",\"NA\",\"Control\",\"NA\",\"NA\"]\n",
    "    return gene_vaf\n",
    "\n",
    "### categorize VAF to high/low based on a single cut-off\n",
    "def vaf_class(vaf):\n",
    "    if vaf!=\"NA\" and vaf != np.nan:\n",
    "        vaf_group=[]\n",
    "        for t in [0.05,0.10,0.15,0.20,0.25]:\n",
    "            if float(vaf)>=t:\n",
    "                vaf_group.append(1)\n",
    "            else:\n",
    "                vaf_group.append(0)\n",
    "        return vaf_group\n",
    "    else:\n",
    "            return [vaf,vaf,vaf,vaf,vaf]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e84acca",
   "metadata": {},
   "outputs": [],
   "source": [
    "############# Load UKB baseline data ##############\n",
    "ukb500k=pd.read_csv(\"/home/sruthi/ukbb_prediction_model/data/ukb500_pheno_prediction.tsv\",delimiter=\"\\t\",header='infer',low_memory=False,parse_dates=['daac','AML', 'MDS', 'MPN', 'CMML', 'dod'],dtype={\"eid\": 'string'})\n",
    "#Remove samples for which somatic variant calling failed\n",
    "ids_somatic=np.loadtxt('/home/sruthi/ukbb_prediction_model/rap/germline_variants/farm/ddx41/code/ids.txt',dtype=\"str\")\n",
    "\n",
    "ukb500k=ukb500k[ukb500k[\"eid\"].isin(ids_somatic)]\n",
    "print(ukb500k.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82ae05c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "######## Add information on genetic sex and exclude people with non-matching genetic and self-reported sex ###################\n",
    "genetic_sex=pd.read_csv(\"/home/sruthi/telomere/inputs/genetic_sex.tsv\",sep=\"\\t\",dtype=\"str\")\n",
    "genetic_sex.columns=[\"eid\",\"genetic_sex\"]\n",
    "ukb500k=ukb500k.merge(genetic_sex,how=\"left\",on=\"eid\")\n",
    "\n",
    "print(ukb500k.genetic_sex.isna().value_counts())\n",
    "ukb500k=ukb500k[(ukb500k[\"sex\"]==ukb500k[\"genetic_sex\"]) | (ukb500k.genetic_sex.isna())]\n",
    "ukb500k.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e2c0db9",
   "metadata": {},
   "outputs": [],
   "source": [
    "############ Chemotherapy/radiotherapy/malignancy phenotypes\n",
    "chemo_radio=pd.read_csv(\"../inputs/ukb_cheracan.tsv\",delimiter=\"\\t\",dtype={\"eid\":\"string\"},parse_dates=[\"chemo_date\",\"radio_date\",\"All_malignant_neoplasms\"],usecols=[\"eid\",\"chemo_date\",\"radio_date\",\"All_malignant_neoplasms\"])\n",
    "ukb500k=ukb500k.merge(chemo_radio,how=\"left\",on=\"eid\")\n",
    "\n",
    "############ Add myeloid malignancy prevalence \n",
    "mm_prevalence=pd.read_csv(\"/home/sruthi/Downloads/prevalent_diagnoses.tsv\",delimiter=\"\\t\",dtype={\"eid\":'string'})\n",
    "ukb500k=ukb500k.drop(columns=[\"AML\",\"MPN\",\"MDS\",\"CMML\"])\n",
    "ukb500k=ukb500k.merge(mm_prevalence,on=\"eid\",how=\"left\",suffixes=[\"\",\"_b\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b99ca38f",
   "metadata": {},
   "outputs": [],
   "source": [
    "######## Somatic variants in U2AF1 from pileup ############\n",
    "u2af1=pd.read_csv(\"/home/sruthi/ukbb_prediction_model/rap/hotspot_pileup/u2af1/analysis_results/u2af1_variants.tsv\",delimiter=\"\\t\",usecols=[\"eid\",\"spdi_x\",\"vaf_total\",\"HGVSp\"],dtype={\"eid\":\"str\"})\n",
    "u2af1[\"symbol\"]=\"U2AF1\"\n",
    "print(u2af1.shape)\n",
    "\n",
    "u2af1=u2af1.rename(columns={\"spdi_x\":\"c_mut\",\"HGVSp\":\"p_mut\",\"vaf_total\":\"vaf\"})\n",
    "u2af1=u2af1[u2af1[\"vaf\"]>=0.05]\n",
    "u2af1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc60f5ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "################ Somatic variant calls from Mutect2 #############\n",
    "som=pd.read_csv(\"/home/sruthi/ukbb_prediction_model/rap/germline_variants/farm/ddx41/extracted_variants/results_var_ch_all_allgenes_UKB\",delimiter=\"\\t\",header='infer',low_memory=False,dtype={\"eid\": 'string'})\n",
    "som[\"p_mut\"]=som[\"p_mut\"].fillna(value=\"NA\")\n",
    "\n",
    "################ Somatic variant calls of CHEK2\n",
    "chek2=pd.read_csv(\"/home/sruthi/telomere/code/chek2_vep/chek2_filtered_variants_simple.tsv\",delimiter=\"\\t\",dtype={\"eid\":\"str\"},usecols=[\"eid\",\"HGVSp\",\"HGVSc\",\"vaf\"])\n",
    "chek2[\"symbol\"]=\"CHEK2\"\n",
    "chek2=chek2.rename(columns={\"HGVSc\":\"c_mut\",\"HGVSp\":\"p_mut\"})\n",
    "\n",
    "\n",
    "genes_to_include=[\"ASXL1\",\"ATM\",\"BCOR\",\"BCORL1\",\"BRAF\",\"BRCC3\",\"CALR\",\"CBL\",\"DNMT3A\",\"EZH2\",\"FLT3\",\"GNAS\",\"GNB1\",\"IDH1\",\"IDH2\",\"JAK2\",\"KDM6A\",\"KIT\",\"KRAS\",\"MPL\",\"NPM1\",\"NRAS\",\"PHF6\",\"PIGA\",\"PPM1D\",\"PRPF40B\",\"RAD21\",\"RUNX1\",\"SF1\",\"SF3A1\",\"SF3B1\",\"SMC1A\",\"SMC3\",\"SRSF2\",\"STAG2\",\"STAT3\",\"TET2\",\"TP53\",\"U2AF1\",\"U2AF2\",\"ZRSR2\",\"CHEK2\"]\n",
    "\n",
    "\n",
    "print(som.shape)\n",
    "som=som[som[\"symbol\"]!=\"U2AF1\"]\n",
    "print(som.shape)\n",
    "som=pd.concat([som,u2af1,chek2],axis=0)\n",
    "print(som.shape)\n",
    "som=som[som[\"symbol\"].isin(genes_to_include)]\n",
    "print(som.shape)\n",
    "som_comb=som.groupby('eid')[['symbol',\"vaf\",\"c_mut\",\"p_mut\"]].agg(cat_rows).reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba1e3075",
   "metadata": {},
   "outputs": [],
   "source": [
    "########### Telomere length information #############\n",
    "telomere_length=pd.read_csv(\"../inputs/ts_ratios.tsv\",delimiter=\"\\t\",header='infer',low_memory=False,dtype={\"f.eid\": 'string'},usecols=[\"f.eid\",\"f.22190.0.0\",\"f.22191.0.0\",\"f.22192.0.0\",\"f.22191.1.0\",\"f.22191.2.0\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b726959",
   "metadata": {},
   "outputs": [],
   "source": [
    "########### Combine baseline, telomere length and somatic variants data ##########\n",
    "ukb500k_ts_som=ukb500k.merge(telomere_length,left_on=\"eid\",right_on=\"f.eid\",how=\"left\")\n",
    "ukb500k_ts_som=ukb500k_ts_som.merge(som_comb,on=\"eid\",how=\"left\")\n",
    "ukb500k_ts_som=ukb500k_ts_som.rename(columns={\"f.22190.0.0\":\"Unadjusted T/S ratio\",\"f.22191.0.0\":\"Adjusted T/S ratio\",\"f.22192.0.0\":\"Z-adjusted T/S log\",\"f.22191.1.0\":\"Adjusted T/S ratio 1\",\"f.22191.2.0\":\"Adjusted T/S ratio 2\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d399fc27",
   "metadata": {},
   "outputs": [],
   "source": [
    "############ UKB age bins\n",
    "ukb500k_ts_som[\"age_bins\"]=pd.cut(ukb500k_ts_som[\"age\"],bins=np.arange(36,77,4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c62f9f75",
   "metadata": {},
   "outputs": [],
   "source": [
    "########### Remove participants with blood collection date different from DAAC and participants who withdrew consent\n",
    "samples_not_matching_daac=np.loadtxt(\"/home/sruthi/reference/ukbiobank/450k_exomes/blood_sample_collection_date/samples_not_matching.tsv\",usecols=[0],dtype=\"str\")\n",
    "\n",
    "samples_no_consent=np.loadtxt(\"/home/sruthi/reference/ukbiobank/withdraw_consent_eids/w56844_2023-04-25.csv\", dtype=\"str\")\n",
    "\n",
    "ukb500k_ts_som=ukb500k_ts_som[~((ukb500k_ts_som[\"eid\"].isin(samples_not_matching_daac)) | \n",
    "ukb500k_ts_som[\"eid\"].isin(samples_no_consent))]\n",
    "ukb500k_ts_som.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6a0f780",
   "metadata": {},
   "outputs": [],
   "source": [
    "#som.loc[som[\"eid\"].isin(ukb500k_ts_som[\"eid\"]),['symbol',\"vaf\",\"c_mut\",\"p_mut\"]].to_csv(\"somatic_calls.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "659009d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "som.loc[(som[\"symbol\"].isin([\"ASXL1\",\"CALR\",\"CBL\",\"GNAS\",\"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"DNMT3A\",\"DNMT3A\",\"U2AF1\"])) & (som[\"eid\"].isin(ukb500k_ts_som[\"eid\"])),\"vaf\"].shape,som.loc[(som[\"symbol\"].isin([\"ASXL1\",\"CALR\",\"CBL\",\"GNAS\",\"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"DNMT3A\",\"DNMT3A\",\"U2AF1\"])) & (som[\"eid\"].isin(ukb500k_ts_som[\"eid\"])) & (som[\"vaf\"].astype(float)>0.02),\"vaf\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a96d9be1",
   "metadata": {},
   "outputs": [],
   "source": [
    "som.loc[((som[\"eid\"].isin(ukb500k_ts_som[\"eid\"])) & (som[\"vaf\"].astype(float)>0.02)),\"vaf\"].shape, som.loc[(som[\"eid\"].isin(ukb500k_ts_som[\"eid\"])),\"vaf\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51514148",
   "metadata": {},
   "outputs": [],
   "source": [
    "############ Age related prevalence of CH genes\n",
    "prev=ukb500k_ts_som[\"age\"].value_counts().rename(\"t\")\n",
    "ch_prev=[]\n",
    "for i,gene in enumerate([\"DNMT3A\", \"TET2\", \"ASXL1\", \"PPM1D\", \"TP53\", \"JAK2\", \"SRSF2\", \"SF3B1\", \"GNB1\",\"GNAS\",\"U2AF1\"]):\n",
    "    if gene!=\"Control\":\n",
    "        ukb500k_ts_som[\"prev\"]=\"UKB\"\n",
    "        ukb500k_ts_som.loc[ukb500k_ts_som[\"symbol\"].str.contains(gene,na=False),\"prev\"]=gene\n",
    "        gene_prev=pd.concat([prev,ukb500k_ts_som.loc[ukb500k_ts_som[\"prev\"]==gene,\"age\"].value_counts().rename(\"n\")],axis=1)\n",
    "        gene_prev[\"pct\"]=gene_prev[\"n\"]/gene_prev[\"t\"]*100\n",
    "        gene_prev[\"single_gene\"]=gene\n",
    "        print(gene_prev)\n",
    "        ch_prev.append(gene_prev)\n",
    "ch_prev=pd.concat(ch_prev)\n",
    "ch_prev=ch_prev.reset_index().rename(columns={\"index\":\"age\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf8b6b1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ukb_age=ukb500k_ts_som[[\"eid\",\"age\",\"sex\"]].copy(deep=True)\n",
    "#ch_prev.to_csv(\"ch_prevalence.tsv\",sep=\"\\t\",index=False,na_rep=\"NA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69588f93",
   "metadata": {},
   "outputs": [],
   "source": [
    "############ Mark somatic variants with highest VAF in each participant and group participants based on VAF ############\n",
    "ukb500k_ts_som[[\"single_gene\",\"single_gene_vaf\",\"single_gene_pmut\",\"multi_gene\",\"multi_gene_vaf\",\"multi_gene_pmut\"]]=ukb500k_ts_som.apply(lambda row: gene_high_vaf(row),axis=1,result_type='expand')\n",
    "#ukb500k_ts_som[[\"single_gene\",\"single_gene_vaf\",\"single_gene_pmut\"]]=ukb500k_ts_som.apply(lambda row: gene_high_vaf(row),axis=1,result_type='expand')\n",
    "\n",
    "ukb500k_ts_som[[\"multi_gene_vaf_0.05\",\"multi_gene_vaf_0.10\",\"multi_gene_vaf_0.15\",\"multi_gene_vaf_0.20\",\"multi_gene_vaf_0.25\"]]=ukb500k_ts_som.apply(lambda row:vaf_class(row[\"multi_gene_vaf\"]),axis=1,result_type='expand')\n",
    "ukb500k_ts_som[[\"single_gene_vaf_0.05\",\"single_gene_vaf_0.10\",\"single_gene_vaf_0.15\",\"single_gene_vaf_0.20\",\"single_gene_vaf_0.25\"]]=ukb500k_ts_som.apply(lambda row:vaf_class(row[\"single_gene_vaf\"]),axis=1,result_type='expand')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7313afe",
   "metadata": {},
   "outputs": [],
   "source": [
    "############ Calculate polygenic risk scores - UKB-GWAS,topmed-GWAS############\n",
    "directories =[\"/home/sruthi/telomere/polygenic_risk_score/keep_ambig/outfiles/all_chr_prs_files\",\"/home/sruthi/telomere/polygenic_risk_score/topmed/outfiles_all_trans/all_chr_prs_files\"]\n",
    "column_names=[\"total_PRS\",\"topmed_all_trans_total_PRS\"]\n",
    "prs_scores=[]\n",
    "for d,directory in enumerate(directories):\n",
    "    i=1\n",
    "    for file in os.listdir(directory):\n",
    "        filename = os.fsdecode(file)\n",
    "        if filename.endswith(\".all_score\"):\n",
    "            print (i,filename)\n",
    "            f=pd.read_csv(directory+\"/\"+filename,usecols=[\"IID\",\"Pt_1\"],dtype={\"IID\":\"string\"},header=\"infer\",sep=\" \")\n",
    "            if i==1:\n",
    "                prs=f\n",
    "                prs=prs.rename(columns={\"Pt_1\":\"total_PRS\"})\n",
    "            else:\n",
    "                prs=prs.merge(f,on=\"IID\",how=\"inner\")\n",
    "                prs[\"total_PRS\"]=prs[\"total_PRS\"]+prs[\"Pt_1\"]\n",
    "                prs=prs[[\"IID\",\"total_PRS\"]]\n",
    "            i+=1\n",
    "    prs.columns=[\"eid\",column_names[d]]\n",
    "    prs_scores.append(prs.set_index(\"eid\"))\n",
    "prs=pd.concat(prs_scores,axis=1,join='outer').reset_index()\n",
    "\n",
    "ukb500k_ts_som=ukb500k_ts_som.merge(prs,on=\"eid\",how=\"left\")\n",
    "ukb500k_ts_som.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05089ad5",
   "metadata": {},
   "outputs": [],
   "source": [
    "########### Split DNMT3A into DNMT3A_R882 AND DNMT3A_other\n",
    "ukb500k_ts_som.loc[((ukb500k_ts_som[\"single_gene\"]==\"DNMT3A\") & (ukb500k_ts_som[\"single_gene_pmut\"].str.contains(\"R882\"))),\"single_gene\"]=\"DNMT3A_R882\"\n",
    "ukb500k_ts_som.loc[(ukb500k_ts_som[\"single_gene\"]==\"DNMT3A\"),\"single_gene\"]=\"DNMT3A_other\"\n",
    "ukb500k_ts_som[\"single_gene\"].value_counts()\n",
    "\n",
    "ukb500k_ts_som.loc[((ukb500k_ts_som[\"multi_gene\"]==\"DNMT3A\") & (ukb500k_ts_som[\"multi_gene_pmut\"].str.contains(\"R882\"))),\"multi_gene\"]=\"DNMT3A_R882\"\n",
    "ukb500k_ts_som.loc[(ukb500k_ts_som[\"multi_gene\"]==\"DNMT3A\"),\"multi_gene\"]=\"DNMT3A_other\"\n",
    "ukb500k_ts_som[\"multi_gene\"].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e3b4081",
   "metadata": {},
   "outputs": [],
   "source": [
    "ukb500k_ts_som[\"single_gene\"].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64db3a5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "########## Genes to include in the analysis ###########\n",
    "imp_genes=[\"ASXL1\",\"CALR\",\"CBL\",\"GNAS\",\"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"Control\",\"DNMT3A_other\",\"DNMT3A_R882\",\"U2AF1\"]\n",
    "all_genes=[\"Control\",\"DNMT3A_other\",\"TET2\",\"DNMT3A_R882\",\"ASXL1\",\"PPM1D\",\"TP53\",\"JAK2\",\"SF3B1\",\"SRSF2\",\"GNB1\",\"GNAS\",\"CBL\",\"ATM\",\"STAT3\",\"CALR\",\"BRCC3\",\"KDM6A\",\"KRAS\",\"BCOR\", \"STAG2\",\"BCORL1\",\"U2AF1\", \"IDH2\",\"CHEK2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed2ebf2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "######### Chemotherapy/radiotherapy/malignant neoplasm prevalence in PPM1D mutation carriers\n",
    "print(ukb500k_ts_som[(ukb500k_ts_som[\"single_gene\"]==\"PPM1D\")].shape,ukb500k_ts_som[(ukb500k_ts_som[\"single_gene\"]==\"PPM1D\") & (ukb500k_ts_som[\"chemo_date\"]<ukb500k_ts_som[\"daac\"])].shape,ukb500k_ts_som[(ukb500k_ts_som[\"single_gene\"]==\"PPM1D\") & ((ukb500k_ts_som[\"chemo_date\"].isna()) | (ukb500k_ts_som[\"chemo_date\"]>ukb500k_ts_som[\"daac\"]))].shape)\n",
    "\n",
    "print(ukb500k_ts_som[(ukb500k_ts_som[\"single_gene\"]==\"PPM1D\")].shape,ukb500k_ts_som[(ukb500k_ts_som[\"single_gene\"]==\"PPM1D\") & (ukb500k_ts_som[\"radio_date\"]<ukb500k_ts_som[\"daac\"])].shape,ukb500k_ts_som[(ukb500k_ts_som[\"single_gene\"]==\"PPM1D\") & ((ukb500k_ts_som[\"radio_date\"].isna()) | (ukb500k_ts_som[\"radio_date\"]>ukb500k_ts_som[\"daac\"]))].shape)\n",
    "\n",
    "print(ukb500k_ts_som[(ukb500k_ts_som[\"single_gene\"]==\"PPM1D\")].shape,ukb500k_ts_som[(ukb500k_ts_som[\"single_gene\"]==\"PPM1D\") & (ukb500k_ts_som[\"All_malignant_neoplasms\"]<ukb500k_ts_som[\"daac\"])].shape,ukb500k_ts_som[(ukb500k_ts_som[\"single_gene\"]==\"PPM1D\") & ((ukb500k_ts_som[\"All_malignant_neoplasms\"].isna()) | (ukb500k_ts_som[\"All_malignant_neoplasms\"]>ukb500k_ts_som[\"daac\"]))].shape)\n",
    "\n",
    "print(ukb500k_ts_som[ukb500k_ts_som[\"single_gene\"].str.contains(\"PPM1D\",na=False)].shape,ukb500k_ts_som[ukb500k_ts_som[\"single_gene\"].str.contains(\"PPM1D\",na=False) & (((ukb500k_ts_som[\"chemo_date\"].notna()) & (ukb500k_ts_som[\"chemo_date\"]<ukb500k_ts_som[\"daac\"])) | ((ukb500k_ts_som[\"radio_date\"].notna()) & (ukb500k_ts_som[\"radio_date\"]<ukb500k_ts_som[\"daac\"])) | ((ukb500k_ts_som[\"All_malignant_neoplasms\"].notna()) & (ukb500k_ts_som[\"All_malignant_neoplasms\"]<ukb500k_ts_som[\"daac\"])))].shape)\n",
    "\n",
    "ukb500k_ts_som[(ukb500k_ts_som[\"single_gene\"].str.contains(\"PPM1D\",na=False)) & (((ukb500k_ts_som[\"chemo_date\"].isna()) | (ukb500k_ts_som[\"chemo_date\"]>ukb500k_ts_som[\"daac\"])) & ((ukb500k_ts_som[\"radio_date\"].isna()) | (ukb500k_ts_som[\"radio_date\"]>ukb500k_ts_som[\"daac\"])) & ((ukb500k_ts_som[\"All_malignant_neoplasms\"].isna()) | (ukb500k_ts_som[\"All_malignant_neoplasms\"]>ukb500k_ts_som[\"daac\"])))].shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a64d0ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "##########Identify people with SRSF2,TET2 and DNMT3A,TET2 mutations for the double mutant analysis\n",
    "ukb500k_ts_som[\"symbol_unique\"]=ukb500k_ts_som[\"symbol\"].map(lambda row:\",\".join([gene for gene in np.unique(np.array(row.split(\",\")))]),na_action='ignore')\n",
    "\n",
    "print(ukb500k_ts_som[\"symbol_unique\"].value_counts()[\"SRSF2,TET2\"],ukb500k_ts_som[\"symbol_unique\"].value_counts()[\"DNMT3A,TET2\"])\n",
    "\n",
    "ukb500k_ts_som[\"multi_gene_TET2\"]=ukb500k_ts_som.apply(lambda row:row[\"symbol_unique\"]+\"-\"+row[\"multi_gene\"] if (row[\"symbol_unique\"]==\"SRSF2,TET2\" or row[\"symbol_unique\"]==\"DNMT3A,TET2\") else row[\"multi_gene\"],axis=1)\n",
    "ukb500k_ts_som[\"multi_gene_TET2_vaf\"]=ukb500k_ts_som.apply(lambda row:row[\"multi_gene_vaf\"] if (row[\"symbol_unique\"]==\"SRSF2,TET2\" or row[\"symbol_unique\"]==\"DNMT3A,TET2\")else row[\"multi_gene_vaf\"],axis=1)\n",
    "print(ukb500k_ts_som.shape)\n",
    "\n",
    "ukb500k_ts_som.loc[(ukb500k_ts_som[\"multi_gene_TET2\"]==\"DNMT3A,TET2-DNMT3A_other\"),\"multi_gene_TET2\"]=\"DNMT3A,TET2-DNMT3A\"\n",
    "ukb500k_ts_som.loc[(ukb500k_ts_som[\"multi_gene_TET2\"]==\"DNMT3A,TET2-DNMT3A_R882\"),\"multi_gene_TET2\"]=\"DNMT3A,TET2-DNMT3A\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "515aa1a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "########### Data with all genes, remove duplicate columns, remove mutations in less frequent genes ###############\n",
    "ukb500k_ts_som_allgenes=ukb500k_ts_som.copy(deep=True)\n",
    "ukb500k_ts_som_allgenes=ukb500k_ts_som_allgenes.loc[:,(ukb500k_ts_som_allgenes.columns!=\"IID\")]\n",
    "ukb500k_ts_som_allgenes=ukb500k_ts_som_allgenes.loc[:,(ukb500k_ts_som_allgenes.columns!=\"f.eid\")]\n",
    "ukb500k_ts_som_allgenes=ukb500k_ts_som_allgenes[(ukb500k_ts_som_allgenes[\"single_gene\"].isin(all_genes)) & ((ukb500k_ts_som_allgenes[\"total_PRS\"].notna()) | (ukb500k_ts_som_allgenes[\"Z-adjusted T/S log\"].notna()))]\n",
    "ukb500k_ts_som_allgenes.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dcc85ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "########### Data with main genes (highest VAF), remove duplicate columns, remove mutations in less frequent genes  ###############\n",
    "ukb500k_ts_som_multigene=ukb500k_ts_som.copy(deep=True)\n",
    "ukb500k_ts_som_multigene=ukb500k_ts_som_multigene.loc[:,(ukb500k_ts_som_multigene.columns!=\"IID\")]\n",
    "ukb500k_ts_som_multigene=ukb500k_ts_som_multigene.loc[:,(ukb500k_ts_som_multigene.columns!=\"f.eid\")]\n",
    "ukb500k_ts_som_multigene=ukb500k_ts_som_multigene[(ukb500k_ts_som_multigene[\"multi_gene\"].isin(imp_genes)) & ((ukb500k_ts_som_multigene[\"total_PRS\"].notna()) | (ukb500k_ts_som_multigene[\"Z-adjusted T/S log\"].notna()))]\n",
    "ukb500k_ts_som_multigene.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7834e10",
   "metadata": {},
   "outputs": [],
   "source": [
    "############ Data with main genes (single mutation only), remove duplicate columns, remove mutations in less frequent genes #############\n",
    "ukb500k_ts_som=ukb500k_ts_som.loc[:,(ukb500k_ts_som.columns!=\"IID\")]\n",
    "ukb500k_ts_som=ukb500k_ts_som.loc[:,(ukb500k_ts_som.columns!=\"f.eid\")]\n",
    "ukb500k_ts_som=ukb500k_ts_som[(ukb500k_ts_som[\"single_gene\"].isin(imp_genes)) & ((ukb500k_ts_som[\"total_PRS\"].notna()) | (ukb500k_ts_som[\"Z-adjusted T/S log\"].notna()))]\n",
    "ukb500k_ts_som.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a23b829",
   "metadata": {},
   "outputs": [],
   "source": [
    "colors=sns.color_palette('husl',len(imp_genes))\n",
    "print(len(colors))\n",
    "gene_color={'Control':colors[0]}\n",
    "for i,gene in enumerate(sorted([x for x in imp_genes if x!=\"Control\"])):\n",
    "    gene_color[gene]=colors[i+1]\n",
    "\n",
    "colors\n",
    "\n",
    "colors=sns.color_palette('husl',len(all_genes))\n",
    "print(len(colors))\n",
    "gene_color_allgenes={'Control':colors[0]}\n",
    "for i,gene in enumerate(sorted([x for x in all_genes if x!=\"Control\"])):\n",
    "    gene_color_allgenes[gene]=colors[i+1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1479d85f",
   "metadata": {},
   "outputs": [],
   "source": [
    "########### Z-normalize PRS #################\n",
    "for y in [\"total_PRS\",\"topmed_all_trans_total_PRS\"]:\n",
    "    ukb500k_ts_som[y+\"_old\"]=ukb500k_ts_som[y]\n",
    "    ukb500k_ts_som[y]=(ukb500k_ts_som[y]-ukb500k_ts_som.loc[(ukb500k_ts_som[\"single_gene\"]==\"Control\"),y].mean())/np.std(ukb500k_ts_som.loc[(ukb500k_ts_som[\"single_gene\"]==\"Control\"),y])\n",
    "    \n",
    "    ukb500k_ts_som_allgenes[y+\"_old\"]=ukb500k_ts_som_allgenes[y]\n",
    "    ukb500k_ts_som_allgenes[y]=(ukb500k_ts_som_allgenes[y]-ukb500k_ts_som_allgenes.loc[(ukb500k_ts_som_allgenes[\"single_gene\"]==\"Control\"),y].mean())/np.std(ukb500k_ts_som_allgenes.loc[(ukb500k_ts_som_allgenes[\"single_gene\"]==\"Control\"),y])\n",
    "    \n",
    "    ukb500k_ts_som_multigene[y+\"_old\"]=ukb500k_ts_som_multigene[y]\n",
    "    ukb500k_ts_som_multigene[y]=(ukb500k_ts_som_multigene[y]-ukb500k_ts_som_multigene.loc[(ukb500k_ts_som_multigene[\"single_gene\"]==\"Control\"),y].mean())/np.std(ukb500k_ts_som_multigene.loc[(ukb500k_ts_som_multigene[\"single_gene\"]==\"Control\"),y])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09ab6dbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "############# Winsorize WBC percentages ##########\n",
    "for wbc_perc in [\"WBC\",\"LY_perc\",\"MO_perc\",\"NE_perc\",\"EO_perc\",\"BA_perc\"]:\n",
    "    ukb500k_ts_som[wbc_perc+\"_wins\"]=scipy.stats.mstats.winsorize(ukb500k_ts_som[wbc_perc],limits=[0.005,0.005],nan_policy=\"omit\")\n",
    "    ukb500k_ts_som_allgenes[wbc_perc+\"_wins\"]=scipy.stats.mstats.winsorize(ukb500k_ts_som_allgenes[wbc_perc],limits=[0.005,0.005],nan_policy=\"omit\")\n",
    "    ukb500k_ts_som_multigene[wbc_perc+\"_wins\"]=scipy.stats.mstats.winsorize(ukb500k_ts_som_multigene[wbc_perc],limits=[0.005,0.005],nan_policy=\"omit\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1460b8d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "ukb500k_ts_som.shape,ukb500k_ts_som_allgenes.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58feb5de",
   "metadata": {},
   "outputs": [],
   "source": [
    "ukb500k_ts_som[\"total_PRS\"].min(),ukb500k_ts_som[\"total_PRS\"].max(),ukb500k_ts_som[\"Z-adjusted T/S log\"].min(),ukb500k_ts_som[\"Z-adjusted T/S log\"].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a81eea10",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "########    Add mCA data   ########################################\n",
    "mca=pd.read_csv(\"/home/sruthi/reference/ukbiobank/450k_exomes/mCA/CNV_data_19808.txt\", usecols=[\"ID\",\"CHR\",\"START_MB\",\"END_MB\",\"SIZE_MB\",\"COPY_CHANGE\",   \"CELL_FRAC\"],dtype=\"string\",header='infer',delimiter=\"\\t\")\n",
    "display(mca)\n",
    "\n",
    "bridge_mca=pd.read_csv(\"/home/sruthi/reference/ukbiobank/450k_exomes/mCA/ukb56844bridge19808.txt\",dtype=\"str\",delimiter=\" \",names=[\"eid\",\"eid_return\"])\n",
    "display(bridge_mca)\n",
    "\n",
    "mca=mca.merge(bridge_mca,left_on=\"ID\",right_on=\"eid_return\",how=\"inner\")\n",
    "print(mca.eid_return.isna().value_counts())\n",
    "display(mca)\n",
    "\n",
    "mca=mca.loc[:,(mca.columns!=\"eid_return\")]\n",
    "display(mca)\n",
    "\n",
    "mca_samples_included=pd.read_csv(\"/home/sruthi/reference/ukbiobank/450k_exomes/mCA/ukb4777.samples_analyzed.txt\",dtype=\"str\",delimiter=\"\\t\",names=[\"eid_return\"])\n",
    "mca_samples_included=mca_samples_included.merge(bridge_mca,on=\"eid_return\",how=\"left\")\n",
    "print(mca_samples_included.eid_return.isna().value_counts())\n",
    "\n",
    "mca[\"CHR_COPY_CHANGE\"]=mca.apply(lambda row:row[\"CHR\"]+\":\"+row[\"COPY_CHANGE\"],axis=1)\n",
    "\n",
    "mca_comb=mca.groupby('eid').agg(cat_rows).reset_index()\n",
    "ukb_mca_age=ukb_age.merge(mca_comb,on=\"eid\",how=\"left\")\n",
    "ukb_mca_age=ukb_mca_age[ukb_mca_age[\"eid\"].isin(mca_samples_included[\"eid\"])]\n",
    "\n",
    "#mca_comb[\"CHR_COPY_CHANGE\"]=mca_comb.apply(lambda row:row[\"CHR\"]+\":\"+row[\"COPY_CHANGE\"] if \",\" not in row[\"COPY_CHANGE\"] else \"NA\",axis=1)\n",
    "mca_comb[\"CHR_COPY_CHANGE\"]=mca_comb[\"CHR_COPY_CHANGE\"].map(lambda row:row if \",\" not in row else \"NA\")\n",
    "\n",
    "display(mca[\"CHR\"].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed00785b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine with somatic variant call data\n",
    "ukb500k_ts_som=ukb500k_ts_som.merge(mca_comb,on=\"eid\",how=\"left\")\n",
    "\n",
    "ukb500k_ts_som_mca=ukb500k_ts_som[ukb500k_ts_som[\"eid\"].isin(mca_samples_included[\"eid\"])].copy(deep=True)\n",
    "print(ukb500k_ts_som_mca.shape)\n",
    "ukb500k_ts_som_mca.loc[ukb500k_ts_som_mca[\"CHR_COPY_CHANGE\"].isna(),\"CHR_COPY_CHANGE\"]=\"Control\"\n",
    "ukb500k_ts_som_mca=ukb500k_ts_som_mca[(~(ukb500k_ts_som_mca[\"CHR_COPY_CHANGE\"].str.contains(\"unknown\",na=False)) & (ukb500k_ts_som_mca[\"CHR_COPY_CHANGE\"]!=\"NA\")  & (ukb500k_ts_som_mca[\"single_gene\"]==\"Control\"))]\n",
    "\n",
    "print(ukb500k_ts_som.shape,ukb500k_ts_som_mca.shape)\n",
    "\n",
    "ukb500k_ts_som_mca=ukb500k_ts_som_mca[((ukb500k_ts_som_mca[\"CELL_FRAC\"]!=\"unknown\") & ~(ukb500k_ts_som_mca[\"CELL_FRAC\"].str.contains(\",\"))) | (ukb500k_ts_som_mca[\"CELL_FRAC\"].isna())]\n",
    "print(ukb500k_ts_som_mca.shape)\n",
    "ukb500k_ts_som_mca[\"CELL_FRAC\"]=ukb500k_ts_som_mca[\"CELL_FRAC\"].astype(float)\n",
    "\n",
    "chromosomes=[str(i) for i in range(1,23)]+[\"X\",\"Y\"]\n",
    "possible_MCAs=[i+\":\"+\"loss\" for i in chromosomes]+[i+\":\"+\"neutral\" for i in chromosomes]+[i+\":\"+\"gain\" for i in chromosomes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89b411e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "########## Calculate prevalence of mCA #################           \n",
    "def check_autosome(row):\n",
    "    autosome=[i+\":\"+\"loss\" for i in chromosomes[0:22]]+[i+\":\"+\"neutral\" for i in chromosomes[0:22]]+[i+\":\"+\"gain\" for i in chromosomes[0:22]]#+[i+\":\"+\"unknown\" for i in chromosomes[0:22]]\n",
    "    for a in autosome:\n",
    "        if a in str(row).split(\",\"):\n",
    "            return \"autosomal\"\n",
    "    return row\n",
    "\n",
    "mca_prev_all=[]\n",
    "                                                                                     \n",
    "for m in [\"X:loss\",\"Y:loss\",\"autosomal\"]:\n",
    "    ukb_mca_age[\"single_gene\"]=\"UKB\"                                                                        \n",
    "    if m==\"autosomal\":\n",
    "        ukb_mca_age[\"single_gene\"]=ukb_mca_age[\"CHR_COPY_CHANGE\"].map(lambda row:check_autosome(row))\n",
    "        mca_prev=pd.concat([ukb_mca_age[\"age\"].value_counts().rename(\"t\"),ukb_mca_age.loc[ukb_mca_age[\"single_gene\"]==m,\"age\"].value_counts().rename(\"n\")],axis=1)\n",
    "        mca_prev[\"pct\"]=mca_prev[\"n\"]/mca_prev[\"t\"]*100\n",
    "        mca_prev[\"single_gene\"]=\"autosomal\"\n",
    "            \n",
    "    elif m==\"Y:loss\":\n",
    "        temp=ukb_mca_age.loc[(ukb_mca_age[\"sex\"]==\"Male\")].copy(deep=True)\n",
    "        temp[\"single_gene\"]=temp[\"CHR_COPY_CHANGE\"].map(lambda row:m if m in str(row).split(\",\") else row)\n",
    "        mca_prev=pd.concat([ukb_mca_age.loc[(ukb_mca_age[\"sex\"]==\"Male\"),\"age\"].value_counts().rename(\"t\"),temp.loc[temp[\"single_gene\"]==m,\"age\"].value_counts().rename(\"n\")],axis=1)\n",
    "        mca_prev[\"pct\"]=mca_prev[\"n\"]/mca_prev[\"t\"]*100\n",
    "        mca_prev[\"single_gene\"]=\"LOY\"\n",
    "        \n",
    "    else:\n",
    "        temp=ukb_mca_age.loc[(ukb_mca_age[\"sex\"]==\"Female\")].copy(deep=True)\n",
    "        temp[\"single_gene\"]=temp[\"CHR_COPY_CHANGE\"].map(lambda row:m if m in str(row).split(\",\") else row)\n",
    "        mca_prev=pd.concat([ukb_mca_age.loc[(ukb_mca_age[\"sex\"]==\"Female\"),\"age\"].value_counts().rename(\"t\"),temp.loc[temp[\"single_gene\"]==m,\"age\"].value_counts().rename(\"n\")],axis=1)\n",
    "        mca_prev[\"pct\"]=mca_prev[\"n\"]/mca_prev[\"t\"]*100\n",
    "        mca_prev[\"single_gene\"]=\"LOX\"\n",
    "\n",
    "    mca_prev_all.append(mca_prev)\n",
    "mca_prev_all=pd.concat(mca_prev_all) \n",
    "mca_prev_all=mca_prev_all.reset_index().rename(columns={\"index\":\"age\"})                                                                                                                            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc5e6e62",
   "metadata": {},
   "outputs": [],
   "source": [
    "#mca_prev_all.to_csv(\"mca_prevalence.tsv\",sep=\"\\t\",index=False,na_rep=\"NA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "222421a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "################ PRS for mCA types ###############\n",
    "temp=ukb500k_ts_som_mca[[\"total_PRS\",\"sex\", \"age\", \"smoking_status\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\",\"CHR_COPY_CHANGE\"]].copy(deep=True)\n",
    "print(temp.shape)\n",
    "temp=temp[~(temp[\"CHR_COPY_CHANGE\"].isin([\"Y:neutral\",\"X:neutral\",\"Y:gain\",\"X:gain\"]))]\n",
    "temp=temp.dropna()\n",
    "temp[\"MCA_type\"]=temp[\"CHR_COPY_CHANGE\"].map(lambda row:\"autosomal MCA\" if row not in [\"Y:loss\",\"X:loss\",\"Control\"] else row)\n",
    "y=\"total_PRS\"\n",
    "\n",
    "counts=temp[\"MCA_type\"].value_counts()\n",
    "print(counts.sum())\n",
    "gene_order=[\"autosomal MCA\",\"Y:loss\",\"X:loss\"]\n",
    "plt.figure(figsize=(6,4))\n",
    "b=sns.violinplot(data=temp,x=\"MCA_type\",y=y,order=[\"Control\"]+gene_order,linewidth=2,palette={\"Control\":gene_color[\"Control\"],\"autosomal MCA\":\"#999999\",\"Y:loss\":\"#E69F00\",\"X:loss\":\"#56B4E9\"})\n",
    "ax=plt.gca()\n",
    "xticks=ax.get_xticklabels()\n",
    "xticks_new=[]\n",
    "for i in xticks:\n",
    "    print(i.get_text(),counts[i.get_text()])\n",
    "    xticks_new.append(i.get_text()+\"\\n(\"+str(counts[i.get_text()])+\")\")\n",
    "#plt.ylim([-5,5])\n",
    "plt.xticks(np.arange(0,len(xticks),1),xticks_new)\n",
    "plt.axhline(y=temp.loc[temp[\"MCA_type\"]==\"Control\",y].median(), color='k', linestyle='--')\n",
    "plt.grid(axis='x')\n",
    "plt.xticks(fontsize=12)\n",
    "plt.yticks(fontsize=12)\n",
    "ax.xaxis.label.set_size(18)\n",
    "ax.yaxis.label.set_size(16)\n",
    "plt.ylabel(\"PRS\")\n",
    "plt.xlabel(\"\")\n",
    "plt.show()\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "127c667c",
   "metadata": {},
   "outputs": [],
   "source": [
    "################### Association analysis for mCA-PRS #####################\n",
    "plot_values=[]\n",
    "for mca_type in [\"autosomal MCA\",\"Y:loss\",\"X:loss\"]:\n",
    "    if mca_type==\"autosomal MCA\":\n",
    "        data=temp[temp[\"MCA_type\"].isin([\"Control\",mca_type])].copy(deep=True)\n",
    "        print(data.shape)\n",
    "        formula = 'MCA_type ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"total_PRS\"])\n",
    "    elif mca_type==\"Y:loss\":\n",
    "        data=temp[(temp[\"MCA_type\"].isin([\"Control\",mca_type])) & (temp[\"sex\"]==\"Male\")].copy(deep=True)\n",
    "        formula = 'MCA_type ~' + ' + '.join(['age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"total_PRS\"])\n",
    "    else:\n",
    "        data=temp[(temp[\"MCA_type\"].isin([\"Control\",mca_type])) & (temp[\"sex\"]==\"Female\")].copy(deep=True)\n",
    "        formula = 'MCA_type ~' + ' + '.join(['age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"total_PRS\"])\n",
    "    data[\"MCA_type\"]=data[\"MCA_type\"].map({mca_type:1,\"Control\":0})\n",
    "    data[\"MCA_type\"]=data[\"MCA_type\"].astype(int)\n",
    "    \n",
    "    log_model=smf.logit(formula=formula,data=data)\n",
    "    result=log_model.fit(max_iterations= 100000)\n",
    "    display(result.summary2())\n",
    "    params = np.exp(result.params)\n",
    "    conf = np.exp(result.conf_int())\n",
    "    print(result.pvalues)\n",
    "    plot_values.append([params[\"total_PRS\"],conf.loc[\"total_PRS\",0],conf.loc[\"total_PRS\",1],result.pvalues[\"total_PRS\"]])\n",
    "\n",
    "plot_values=np.array(plot_values)\n",
    "data_mca=temp.copy(deep=True)\n",
    "data_mca=data_mca.rename(columns={\"Z-adjusted T/S log\":\"tl\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "051acfe0",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.errorbar(np.arange(3),plot_values[:,0],yerr=[plot_values[:,0]-plot_values[:,1],plot_values[:,2]-plot_values[:,0]],fmt=\"o\",markersize=7)\n",
    "ticklabel=[\"autosomal MCA\",\"Y:loss\",\"X:loss\"]\n",
    "ticklabel=[ticklabel[i]+\"\\n*\" if plot_values[i,3]<0.05 else ticklabel[i] for i in range(3)]\n",
    "plt.xticks(np.arange(3),ticklabel,fontsize=13)\n",
    "plt.ylabel(\"OR\",fontsize=13)\n",
    "plt.axhline(y=1,linestyle=\"--\",lw=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbee56c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "########### Save data ################                                                                                                                                                                                                                                                                                                                                                       \n",
    "supptable_1_mca=pd.DataFrame(plot_values)\n",
    "supptable_1_mca[\"Gene/mCA\"]=[\"autosomal mCA\",\"LOY\",\"LOX\"]\n",
    "supptable_1_mca.columns=[\"OR\",\"OR lower CI (2.5%)\",\"OR upper CI (97.5%)\",\"P-value\",\"Gene/mCA\"]\n",
    "supptable_1_mca=supptable_1_mca[[\"Gene/mCA\",\"OR\",\"OR lower CI (2.5%)\",\"OR upper CI (97.5%)\",\"P-value\"]]\n",
    "supptable_1_mca[\"FDR\"]=multipletests(supptable_1_mca[\"P-value\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n",
    "#supptable_1_mca[[\"Gene/mCA\",\"Coefficient\",\"Lower CI (2.5%)\",\"Upper CI (97.5%)\",\"P-value\"]].to_csv(\"MCA_coef_supptable.csv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8e533e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "############# Association with all CH and PRS ###############\n",
    "y=\"total_PRS\"\n",
    "temp=ukb500k_ts_som.copy(deep=True)\n",
    "temp=temp.loc[(temp[y].notna())]\n",
    "\n",
    "formula = 'single_gene ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10','total_PRS'])\n",
    "\n",
    "or_table_pr=[]\n",
    "\n",
    "data=temp[[\"total_PRS\", \"sex\", \"age\", \"smoking_status\", \"PC1\", \"PC2\", \"PC3\", \"PC4\", \"PC5\", \"PC6\", \"PC7\", \"PC8\", \"PC9\", \"PC10\", \"single_gene\"]].copy(deep=True)\n",
    "data=data.dropna()\n",
    "print(data.shape)\n",
    "data[\"single_gene\"]=data[\"single_gene\"].map(lambda row:0 if row==\"Control\" else 1)\n",
    "data[\"single_gene\"]=data[\"single_gene\"].astype(int)\n",
    "print(data[\"single_gene\"].value_counts())\n",
    "log_model_allch=smf.logit(formula=formula,data=data)\n",
    "result_allch=log_model_allch.fit(max_iterations= 100000)\n",
    "display(result_allch.summary2())\n",
    "params_allch = np.exp(result_allch.params)\n",
    "conf_allch = np.exp(result_allch.conf_int())\n",
    "print(result_allch.pvalues)\n",
    "print([\"all CH\", params_allch[\"total_PRS\"], conf_allch.loc[\"total_PRS\",0], conf_allch.loc[\"total_PRS\",1], result_allch.pvalues[\"total_PRS\"]])\n",
    "or_table_pr.append([\"all CH\", params_allch[\"total_PRS\"], conf_allch.loc[\"total_PRS\",0], conf_allch.loc[\"total_PRS\",1], result_allch.pvalues[\"total_PRS\"]])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "593427b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "############# Association with CH and PRS ###############\n",
    "\n",
    "y=\"total_PRS\"\n",
    "temp=ukb500k_ts_som.copy(deep=True)\n",
    "temp=temp.loc[(temp[y].notna())]\n",
    "\n",
    "formula = 'single_gene ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10','total_PRS'])\n",
    "gene_list=list(temp[\"single_gene\"].unique())\n",
    "gene_list.remove(\"Control\")\n",
    "\n",
    "for gene in gene_list:\n",
    "    data=temp[[\"total_PRS\", \"sex\", \"age\", \"smoking_status\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\",\"single_gene\"]].copy(deep=True)\n",
    "    data=data.dropna()\n",
    "    print(data.shape)\n",
    "    data=data[(data[\"single_gene\"].isin([gene,\"Control\"]))]\n",
    "    data[\"single_gene\"]=data[\"single_gene\"].map({gene:1,\"Control\":0})\n",
    "    data[\"single_gene\"]=data[\"single_gene\"].astype(int)\n",
    "    print(data[\"single_gene\"].value_counts())\n",
    "    log_model=smf.logit(formula=formula,data=data)\n",
    "    result=log_model.fit(max_iterations= 100000)\n",
    "    display(result.summary2())\n",
    "    params = np.exp(result.params)\n",
    "    conf = np.exp(result.conf_int())\n",
    "    print(result.pvalues)\n",
    "    print([gene,params[\"total_PRS\"],conf.loc[\"total_PRS\",0],conf.loc[\"total_PRS\",1],result.pvalues[\"total_PRS\"]])\n",
    "    or_table_pr.append([gene,params[\"total_PRS\"],conf.loc[\"total_PRS\",0],conf.loc[\"total_PRS\",1],result.pvalues[\"total_PRS\"]])\n",
    "or_table_pr=pd.DataFrame(or_table_pr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a6cf445",
   "metadata": {},
   "outputs": [],
   "source": [
    "or_table_pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9d2583a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Plot Figure 1b\n",
    "\n",
    "y=\"total_PRS\"\n",
    "temp=ukb500k_ts_som[[\"total_PRS\", \"sex\", \"age\", \"smoking_status\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\",\"single_gene\"]].copy(deep=True)\n",
    "temp=temp.dropna()\n",
    "gene_count=Counter(temp[\"single_gene\"])\n",
    "\n",
    "left, width = 0.01, 15*0.05\n",
    "bottom_h, height_h=0.1,0.23\n",
    "bottom, height = 0.35, .60\n",
    "left_h = left+width+0.03\n",
    "rect_0 = [left, bottom, width, height]\n",
    "rect_1 = [left_h, bottom, 4*0.05, height]\n",
    "rect_2 = [left, bottom_h, width, height_h]\n",
    "rect_3 = [left_h, bottom_h, 4*0.05, height_h]\n",
    "\n",
    "plt.figure(figsize=(22,12))\n",
    "ax1=plt.axes(rect_0)\n",
    "ax2=plt.axes(rect_1)\n",
    "ax3=plt.axes(rect_2,sharex=ax1)\n",
    "ax = [ax1, ax2, ax3, plt.axes(rect_3,sharex=ax2)]\n",
    "\n",
    "gene_color_new={}\n",
    "for key in gene_color:\n",
    "    gene_color_new[key]=sns.set_hls_values(gene_color[key],s=0.5)\n",
    "black=sns.set_hls_values('#000000',s=0.5)\n",
    "\n",
    "gene_list=[]\n",
    "gene_groups=[]\n",
    "quartiles=[]\n",
    "\n",
    "gene_list.append(\"Control\")\n",
    "gene_groups.append(np.array(temp.loc[((temp[\"single_gene\"]==\"Control\") & (temp[y].notna())),y]))\n",
    "quartiles.append(np.percentile(gene_groups[-1],[25, 50, 75]))\n",
    "\n",
    "gene_list.append(\"all CH\")\n",
    "gene_groups.append(np.array(temp.loc[((temp[\"single_gene\"].isin([\"ASXL1\",\"CALR\",\"CBL\",\"DNMT3A_R882\",\"DNMT3A_other\",\"GNAS\", \"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"U2AF1\"])) & (temp[y].notna())),y]))\n",
    "quartiles.append(np.percentile(gene_groups[-1],[25, 50, 75]))\n",
    "\n",
    "for key in [\"ASXL1\",\"CALR\",\"CBL\",\"DNMT3A_R882\",\"DNMT3A_other\",\"GNAS\", \"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"U2AF1\"]: \n",
    "        if gene_count[key]>=20:\n",
    "            gene_list.append(key)\n",
    "            gene_groups.append(np.array(temp.loc[((temp[\"single_gene\"]==key) & (temp[y].notna())),y]))\n",
    "            quartiles.append(np.percentile(gene_groups[-1],[25, 50, 75]))\n",
    "quartiles=np.array(quartiles)           \n",
    "            \n",
    "if gene_list != []:\n",
    "        temp=temp.loc[((temp[\"single_gene\"].isin(gene_list)) & (temp[y].notna())),:]\n",
    "        counts=temp[\"single_gene\"].value_counts()\n",
    "        counts[\"all CH\"]=gene_groups[1].shape[0]\n",
    "        \n",
    "        parts = ax[0].violinplot(gene_groups, showmeans=False, showmedians=False, showextrema=False,widths=0.65,positions=np.arange(0,len(gene_list),1))\n",
    "\n",
    "        for i,pc in enumerate(parts['bodies']):\n",
    "            if gene_list[i]==\"Control\":\n",
    "                pc.set_facecolor(gene_color_new[\"Control\"])\n",
    "            else:\n",
    "                pc.set_facecolor(\"#999999\")\n",
    "            pc.set_edgecolor(black)\n",
    "            pc.set_alpha(1)\n",
    "            pc.set_linewidth(1.5)\n",
    "\n",
    "        ax[0].boxplot(gene_groups,positions=np.arange(0,len(gene_list),1),showcaps=False,widths=0.06, patch_artist=True, boxprops=dict(color=black, facecolor=black),whiskerprops=dict(color=black, linewidth=2), medianprops=dict(color=\"w\", linewidth=0 ),showfliers=False)\n",
    "        inds = np.arange(0, len(quartiles))\n",
    "        ax[0].scatter(inds, quartiles[:,1], marker='o', color='white', s=30, zorder=3)\n",
    "        \n",
    "        xticks_new=[]\n",
    "        for i in gene_list:\n",
    "            print(i,counts[i])\n",
    "            if i != \"Control\":\n",
    "                pval_prs=or_table_pr.loc[(or_table_pr[0]==i),4].reset_index()\n",
    "                pval_prs=pval_prs.iloc[0,1]\n",
    "            else:\n",
    "                pval_prs=1\n",
    "            if pval_prs <0.05:\n",
    "                xticks_new.append(i[0].upper()+i[1:]+\"\\n(\"+str(counts[i])+\")\\n*\")\n",
    "            else:\n",
    "                xticks_new.append(i[0].upper()+i[1:]+\"\\n(\"+str(counts[i])+\")\")\n",
    "        print(xticks_new)\n",
    "        ax[0].set_xticks(np.arange(0,len(gene_list),1))\n",
    "        #ax[0].set_xticklabels(xticks_new,rotation=45)\n",
    "        ax[0].set_xticklabels([])\n",
    "        ax[0].tick_params(axis='y',labelsize=15,size=7)\n",
    "        ax[0].tick_params(axis='x',labelsize=12,labelbottom=False)\n",
    "        ax[0].grid(axis='x')\n",
    "        ax[0].xaxis.label.set_size(20)\n",
    "        ax[0].yaxis.label.set_size(22)\n",
    "        ax[0].set_ylabel(\"PRS\")\n",
    "        ax[0].set_ylim([-4,4])\n",
    "        #ax[0].set_xlabel(\"Gene\")\n",
    "        ax[0].axhline(y=temp.loc[temp[\"single_gene\"]==\"Control\",y].median(), color='k', linestyle='--')\n",
    "        print(temp.loc[temp[\"single_gene\"]==\"Control\",y].median())\n",
    "                        \n",
    "temp=data_mca[[y,\"MCA_type\"]].copy(deep=True)\n",
    "temp=temp.dropna()\n",
    "temp[\"MCA_type\"]=temp[\"MCA_type\"].map(lambda row:\"autosomal\" if row not in [\"Y:loss\",\"X:loss\",\"Control\"] else row)\n",
    "y=\"total_PRS\"\n",
    "\n",
    "temp=temp.loc[(temp[y].notna()),:]\n",
    "counts=temp[\"MCA_type\"].value_counts()\n",
    "print(counts.sum())\n",
    "gene_groups=[]\n",
    "gene_order=[\"Control\",\"autosomal\",\"Y:loss\",\"X:loss\"]\n",
    "quartiles=[]\n",
    "for key in gene_order:\n",
    "    gene_groups.append(temp.loc[(temp[\"MCA_type\"]==key)& (temp[y].notna()),y])\n",
    "    quartiles.append(np.percentile(gene_groups[-1],[25, 50, 75]))\n",
    "quartiles=np.array(quartiles)           \n",
    "                       \n",
    "parts = ax[1].violinplot(gene_groups, showmeans=False, showmedians=False, showextrema=False,widths=0.65,positions=np.arange(0,len(gene_order),1))\n",
    "mca_color={\"Control\":gene_color[\"Control\"],\"autosomal\":\"#999999\",\"Y:loss\":\"#E69F00\",\"X:loss\":\"#56B4E9\"}           \n",
    "for key in mca_color:\n",
    "    mca_color[key]=sns.set_hls_values(mca_color[key],s=0.5)\n",
    "    \n",
    "for i,pc in enumerate(parts['bodies']):\n",
    "            if gene_order[i]==\"Control\":\n",
    "                pc.set_facecolor(mca_color[gene_order[i]])\n",
    "            else:\n",
    "                pc.set_facecolor(\"#8585ad\")\n",
    "            pc.set_edgecolor(black)\n",
    "            pc.set_alpha(1)\n",
    "            pc.set_linewidth(1.5)\n",
    "\n",
    "ax[1].boxplot(gene_groups,positions=np.arange(0,len(gene_order),1),showcaps=False,widths=0.06, patch_artist=True,boxprops=dict(color=black, facecolor=black),\n",
    "            whiskerprops=dict(color=black, linewidth=2),\n",
    "            medianprops=dict(color=\"w\", linewidth=0 ),showfliers=False)                       \n",
    "inds = np.arange(0, len(quartiles))\n",
    "ax[1].scatter(inds, quartiles[:,1], marker='o', color='white', s=30, zorder=3)                       \n",
    "xticks_new_mca=[]\n",
    "gene_rename={\"autosomal\":\"autosomal\",\"Y:loss\":\"LOY\",\"X:loss\":\"LOX\"}\n",
    "for j,i in enumerate(gene_order):\n",
    "    print(i,counts[i])\n",
    "    if i!=\"Control\":\n",
    "        if plot_values[j-1,3]<0.05:\n",
    "            xticks_new_mca.append(gene_rename[i]+\"\\n(\"+str(counts[i])+\")\"+\"\\n*\")\n",
    "        else:\n",
    "            xticks_new_mca.append(gene_rename[i]+\"\\n(\"+str(counts[i])+\")\")\n",
    "    else:\n",
    "        xticks_new_mca.append(i+\"\\n(\"+str(counts[i])+\")\")\n",
    "\n",
    "ax[1].set_xticks(np.arange(0,len(xticks),1))\n",
    "ax[1].set_xticklabels([])\n",
    "ax[1].grid(axis='x')\n",
    "ax[1].tick_params(axis='x',labelsize=12,labelbottom=False)\n",
    "ax[1].tick_params(axis='y',labelsize=15)\n",
    "ax[1].xaxis.label.set_size(25)\n",
    "ax[1].yaxis.label.set_size(25)\n",
    "ax[1].set_ylim([-4,4])\n",
    "\n",
    "ax[1].axhline(y=temp.loc[temp[\"MCA_type\"]==\"Control\",y].median(), color='k', linestyle='--')\n",
    "plt.subplots_adjust(wspace=0.05)\n",
    "\n",
    "#or_table_pr=or_table_pr.set_index(0).loc[[x for x in gene_list if x!=\"Control\"],:]\n",
    "#or_table_pr=or_table_pr.reset_index()\n",
    "or_table_pr=or_table_pr.set_index(0).loc[[x for x in gene_list if x!=\"Control\"],:]\n",
    "or_table_pr=or_table_pr.reset_index()\n",
    "or_significant=or_table_pr.loc[(or_table_pr[4]>=0.05),:]\n",
    "ax[2].errorbar(or_significant.index+1,or_significant.loc[:,1],yerr=[or_significant.loc[:,1]-or_significant.loc[:,2],or_significant.loc[:,3]-or_significant.loc[:,1]],fmt=\"o\",markersize=20,elinewidth=2,c=\"#b3b3b3\",label=\"P\"+r'$\\mathrm{\\geq}0.05$')\n",
    "or_significant=or_table_pr.loc[(or_table_pr[4]<0.05),:]\n",
    "ax[2].errorbar(or_significant.index+1,or_significant.loc[:,1],yerr=[or_significant.loc[:,1]-or_significant.loc[:,2],or_significant.loc[:,3]-or_significant.loc[:,1]],fmt=\"o\",markersize=20,elinewidth=2,c=\"#4d4d4d\",label=\"P\"+r'$\\mathrm{<}0.05$')\n",
    "\n",
    "ax[2].set_xticks(np.arange(0,16))\n",
    "ax[2].set_xticklabels(xticks_new,fontsize=15,rotation=35)\n",
    "ax[2].set_ylabel(\"OR\",fontsize=22)\n",
    "ax[2].set_xlabel(\"CH driver gene\",fontsize=22)\n",
    "ax[2].tick_params(axis='y',labelsize=15,size=7)\n",
    "ax[2].axhline(y=1,linestyle=\"--\",lw=1)\n",
    "ax[2].legend()\n",
    "\n",
    "\n",
    "plot_values_s=plot_values[(plot_values[:,3]<0.05),:]\n",
    "ax[3].errorbar(np.where(plot_values[:,3]<0.05)[0]+1,plot_values_s[:,0],yerr=[plot_values_s[:,0]-plot_values_s[:,1],plot_values_s[:,2]-plot_values_s[:,0]],fmt=\"o\",markersize=20,elinewidth=2,c=\"#575782\",label=\"P\"+r'$\\mathrm{<}0.05$')\n",
    "\n",
    "plot_values_n=plot_values[plot_values[:,3]>=0.05,:]\n",
    "ax[3].errorbar(np.where(plot_values[:,3]>=0.05)[0]+1,plot_values_n[:,0],yerr=[plot_values_n[:,0]-plot_values_n[:,1],plot_values_n[:,2]-plot_values_n[:,0]],fmt=\"o\",markersize=20,elinewidth=2,c=\"#b1b1cb\",label=\"P\"+r'$\\mathrm{\\geq}0.05$')\n",
    "ax[3].set_xticks(np.arange(0,4))\n",
    "ax[3].set_xticklabels(xticks_new_mca,fontsize=15,rotation=45)\n",
    "ax[3].tick_params(axis='y',labelsize=15)\n",
    "ax[3].axhline(y=1,linestyle=\"--\",lw=1)\n",
    "ax[3].set_xlabel(\"mCA type\",fontsize=22)\n",
    "ax[3].legend()\n",
    "\n",
    "for a in ax:\n",
    "    a.spines[['right', 'top']].set_visible(False)\n",
    "\n",
    "#plt.savefig(\"../figures_revision1/prs_ch_or_mca_singlegene.png\",dpi=600,bbox_inches=\"tight\")\n",
    "#plt.savefig(\"../figures_revision1/prs_ch_or_mca_singlegene.pdf\",bbox_inches=\"tight\")\n",
    "#plt.savefig(\"../figures_revision1/prs_ch_or_mca_singlegene.svg\",bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdbf01b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "supptable_1_gene=or_table_pr.copy(deep=True)\n",
    "supptable_1_gene.columns=[\"Gene/mCA\",\"OR\",\"OR lower CI (2.5%)\",\"OR upper CI (97.5%)\",\"P-value\"]\n",
    "supptable_1_gene=supptable_1_gene.sort_values(by=\"Gene/mCA\",key=lambda col: col.str.lower())\n",
    "supptable_1_gene[\"FDR\"]=multipletests(supptable_1_gene[\"P-value\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n",
    "\n",
    "supptable_1=pd.concat([supptable_1_gene[[\"Gene/mCA\",\"OR\",\"OR lower CI (2.5%)\",\"OR upper CI (97.5%)\",\"P-value\",\"FDR\"]],supptable_1_mca[[\"Gene/mCA\",\"OR\",\"OR lower CI (2.5%)\",\"OR upper CI (97.5%)\",\"P-value\",\"FDR\"]]])\n",
    "supptable_1[\"FDR_all\"]=multipletests(supptable_1[\"P-value\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n",
    "#supptable_1.to_csv(\"prs_coef_supptable_r1.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23dd422f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#supptable_1.to_csv(\"prs_coef_supptable_singlegene_r1_singlegene.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41577fbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "supptable_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9adf87d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "############### Association between mCA and measured telomere length ####################\n",
    "temp=ukb500k_ts_som_mca[[\"Z-adjusted T/S log\",\"sex\", \"age\", \"smoking_status\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\",\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",\"CHR_COPY_CHANGE\",\"CELL_FRAC\",\"HM\"]].copy(deep=True)\n",
    "temp=temp[~(temp[\"CHR_COPY_CHANGE\"].isin([\"Y:neutral\",\"X:neutral\",\"Y:gain\",\"X:gain\"]))]\n",
    "temp.loc[temp[\"CHR_COPY_CHANGE\"]==\"Control\",\"CELL_FRAC\"]=0\n",
    "temp=temp.dropna()\n",
    "temp=temp[temp[\"HM\"]==0]\n",
    "temp[\"MCA_type\"]=temp[\"CHR_COPY_CHANGE\"].map(lambda row:\"autosomal MCA\" if row not in [\"Y:loss\",\"X:loss\",\"Control\"] else row)\n",
    "y=\"Z-adjusted T/S log\"\n",
    "\n",
    "counts=temp[\"MCA_type\"].value_counts()\n",
    "print(counts.sum())\n",
    "gene_order=[\"autosomal MCA\",\"Y:loss\",\"X:loss\"]\n",
    "plt.figure(figsize=(6,4))\n",
    "b=sns.violinplot(data=temp,x=\"MCA_type\",y=y,order=[\"Control\"]+gene_order,linewidth=2,palette={\"Control\":gene_color[\"Control\"],\"autosomal MCA\":\"#999999\",\"Y:loss\":\"#E69F00\",\"X:loss\":\"#56B4E9\"})\n",
    "ax=plt.gca()\n",
    "xticks=ax.get_xticklabels()\n",
    "xticks_new=[]\n",
    "for i in xticks:\n",
    "    print(i.get_text(),counts[i.get_text()])\n",
    "    xticks_new.append(i.get_text()+\"\\n(\"+str(counts[i.get_text()])+\")\")\n",
    "plt.ylim([-5,5])\n",
    "plt.xticks(np.arange(0,len(xticks),1),xticks_new)\n",
    "plt.axhline(y=temp.loc[temp[\"MCA_type\"]==\"Control\",y].median(), color='k', linestyle='--')\n",
    "plt.grid(axis='x')\n",
    "plt.xticks(fontsize=12)\n",
    "plt.yticks(fontsize=12)\n",
    "ax.xaxis.label.set_size(18)\n",
    "ax.yaxis.label.set_size(16)\n",
    "plt.ylabel(\"Telomere length\")\n",
    "plt.xlabel(\"\")\n",
    "#plt.savefig(\"../figures_new/tl_mca.png\",dpi=600,bbox_inches=\"tight\")\n",
    "#plt.savefig(\"../figures_new/tl_mca.pdf\",bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "            \n",
    "\n",
    "\n",
    "plot_values=[]\n",
    "for mca_type in [\"autosomal MCA\",\"Y:loss\",\"X:loss\"]:\n",
    "    if mca_type==\"autosomal MCA\":\n",
    "        data=temp[temp[\"MCA_type\"].isin([\"Control\",mca_type])].copy(deep=True)\n",
    "        formula = 'tl ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",'C(MCA_type,Treatment(reference=\"Control\"))'])\n",
    "    elif mca_type==\"Y:loss\":\n",
    "        data=temp[(temp[\"MCA_type\"].isin([\"Control\",mca_type])) & (temp[\"sex\"]==\"Male\")].copy(deep=True)\n",
    "        formula = 'tl ~' + ' + '.join(['age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",'C(MCA_type,Treatment(reference=\"Control\"))'])\n",
    "    else:\n",
    "        data=temp[(temp[\"MCA_type\"].isin([\"Control\",mca_type])) & (temp[\"sex\"]==\"Female\")].copy(deep=True)\n",
    "        formula = 'tl ~' + ' + '.join(['age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",'C(MCA_type,Treatment(reference=\"Control\"))'])\n",
    "    print(data[\"MCA_type\"].value_counts())\n",
    "    data=data.rename(columns={\"Z-adjusted T/S log\":\"tl\"})\n",
    "    lin_model=smf.ols(formula=formula,data=data)\n",
    "    result=lin_model.fit(max_iterations= 100000)\n",
    "    display(result.summary2())\n",
    "    params = result.params\n",
    "    conf = result.conf_int()\n",
    "    print(result.pvalues)\n",
    "    key='C(MCA_type, Treatment(reference=\"Control\"))[T.'+mca_type+']'\n",
    "    plot_values.append([params[key],conf.loc[key,0],conf.loc[key,1],result.pvalues[key]])\n",
    "    \n",
    "plot_values_mca=np.array(plot_values)\n",
    "\n",
    "\n",
    "supptable_2_mca=pd.DataFrame(plot_values_mca)\n",
    "supptable_2_mca[\"Gene/mCA\"]=[\"autosomal mCA\",\"LOY\",\"LOX\"]\n",
    "supptable_2_mca.columns=[\"Coefficient\",\"Coefficient lower CI (2.5%)\",\"Coefficient upper CI (97.5%)\",\"P-value\",\"Gene/mCA\"]\n",
    "supptable_2_mca=supptable_2_mca[[\"Gene/mCA\",\"Coefficient\",\"Coefficient lower CI (2.5%)\",\"Coefficient upper CI (97.5%)\",\"P-value\"]]\n",
    "supptable_2_mca[\"FDR\"]=multipletests(supptable_2_mca[\"P-value\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n",
    "\n",
    "plot_values=[]\n",
    "for mca_type in [\"autosomal MCA\",\"Y:loss\",\"X:loss\"]:\n",
    "    if mca_type==\"autosomal MCA\":\n",
    "        data=temp[temp[\"MCA_type\"].isin([\"Control\",mca_type])].copy(deep=True)\n",
    "\n",
    "        formula = 'tl ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",'C(MCA_type,Treatment(reference=\"Control\"))','CELL_FRAC'])\n",
    "    elif mca_type==\"Y:loss\":\n",
    "        data=temp[(temp[\"MCA_type\"].isin([\"Control\",mca_type])) & (temp[\"sex\"]==\"Male\")].copy(deep=True)\n",
    "        formula = 'tl ~' + ' + '.join(['age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",'C(MCA_type,Treatment(reference=\"Control\"))','CELL_FRAC'])\n",
    "    else:\n",
    "        data=temp[(temp[\"MCA_type\"].isin([\"Control\",mca_type])) & (temp[\"sex\"]==\"Female\")].copy(deep=True)\n",
    "        formula = 'tl ~' + ' + '.join(['age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",'C(MCA_type,Treatment(reference=\"Control\"))','CELL_FRAC'])\n",
    "    print(data[\"MCA_type\"].value_counts())\n",
    "    data=data.rename(columns={\"Z-adjusted T/S log\":\"tl\"})\n",
    "    lin_model=smf.ols(formula=formula,data=data)\n",
    "    result=lin_model.fit(max_iterations= 100000)\n",
    "    display(result.summary2())\n",
    "    params = result.params\n",
    "    conf = result.conf_int()\n",
    "    print(result.pvalues)\n",
    "    #key='C(MCA_type, Treatment(reference=\"Control\"))[T.'+mca_type+']'\n",
    "    key=\"CELL_FRAC\"\n",
    "    plot_values.append([params[key],conf.loc[key,0],conf.loc[key,1],result.pvalues[key]])\n",
    "    \n",
    "plot_values=np.array(plot_values)\n",
    "data_mca=temp.copy(deep=True)\n",
    "data_mca=data_mca.rename(columns={\"Z-adjusted T/S log\":\"tl\"})\n",
    "\n",
    "plt.bar(np.arange(3),plot_values_mca[:,0],yerr=[plot_values_mca[:,0]-plot_values_mca[:,1],plot_values_mca[:,2]-plot_values_mca[:,0]],width=0.5)\n",
    "ticklabel=[\"autosomal MCA\",\"Y:loss\",\"X:loss\"]\n",
    "ticklabel=[ticklabel[i]+\"\\n*\" if plot_values_mca[i,3]<0.05 else ticklabel[i] for i in range(3)]\n",
    "plt.xticks(np.arange(3),ticklabel,fontsize=13)\n",
    "plt.ylabel(\"Coefficient\",fontsize=13)\n",
    "#plt.yticks([-0.15,-0.10,-0.05])\n",
    "plt.axhline(y=0,linestyle=\"--\",lw=1)\n",
    "#plt.savefig(\"../figures_new/tl_mca_coef.png\",dpi=600,bbox_inches=\"tight\")\n",
    "#plt.savefig(\"../figures_new/tl_mca_coef.pdf\",bbox_inches=\"tight\")\n",
    "\n",
    "supptable_3_mca=pd.DataFrame(plot_values)\n",
    "supptable_3_mca[\"Gene/mCA\"]=[\"autosomal mCA\",\"LOY\",\"LOX\"]\n",
    "supptable_3_mca.columns=[\"Coefficient\",\"Coefficient lower CI (2.5%)\",\"Coefficient upper CI (97.5%)\",\"P-value\",\"Gene/mCA\"]\n",
    "supptable_3_mca=supptable_3_mca[[\"Gene/mCA\",\"Coefficient\",\"Coefficient lower CI (2.5%)\",\"Coefficient upper CI (97.5%)\",\"P-value\"]]\n",
    "supptable_3_mca[\"FDR\"]=multipletests(supptable_3_mca[\"P-value\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n",
    "#supptable_1_mca[[\"mCA type\",\"Coefficient\",\"Lower CI (2.5%)\",\"Upper CI (97.5%)\",\"P-value\"]].to_csv(\"MCA_cellfrac_coef_supptable.csv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4f3a5ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "############## Association with telomere length and individual gene+vaf group ##############\n",
    "\n",
    "data=ukb500k_ts_som[[\"Z-adjusted T/S log\", \"sex\", \"age\", \"smoking_status\", \"single_gene\", \"single_gene_vaf\", \"single_gene_vaf_0.10\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\", \"WBC_wins\",\"MO_perc_wins\", \"LY_perc_wins\", \"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",\"eid\",\"HM\"]].copy(deep=True)\n",
    "data=data.dropna()\n",
    "data=data[(data[\"HM\"]==0)]\n",
    "data[\"single_gene_vaf_group\"]=data[\"single_gene\"]+\"_\"+data[\"single_gene_vaf_0.10\"].astype(\"str\")\n",
    "print(data.shape)\n",
    "\n",
    "data=data.rename(columns={\"Z-adjusted T/S log\":\"tl\"})\n",
    "formula = 'tl ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','C(single_gene_vaf_group,Treatment(reference=\"Control_NA\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\"])\n",
    "\n",
    "lin_model_fullreg_gene=smf.ols(formula=formula,data=data)\n",
    "result_fullreg_gene=lin_model_fullreg_gene.fit(max_iterations= 100000)\n",
    "display(result_fullreg_gene.summary2())\n",
    "params_fullreg_gene = result_fullreg_gene.params\n",
    "conf_fullreg_gene = result_fullreg_gene.conf_int() \n",
    "\n",
    "print(result_fullreg_gene.pvalues)\n",
    "\n",
    "\n",
    "############## Association with telomere length and individual gene ##############\n",
    "\n",
    "data=ukb500k_ts_som[[\"Z-adjusted T/S log\", \"sex\", \"age\", \"smoking_status\", \"single_gene\", \"single_gene_vaf\", \"single_gene_vaf_0.10\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\", \"WBC_wins\",\"MO_perc_wins\", \"LY_perc_wins\", \"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",\"eid\",\"HM\"]].copy(deep=True)\n",
    "data=data.dropna()\n",
    "data=data[(data[\"HM\"]==0)]\n",
    "\n",
    "\n",
    "print(data.shape)\n",
    "\n",
    "data=data.rename(columns={\"Z-adjusted T/S log\":\"tl\"})\n",
    "formula = 'tl ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','C(single_gene,Treatment(reference=\"Control\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\"])\n",
    "\n",
    "lin_model=smf.ols(formula=formula,data=data)\n",
    "result_fullreg_gene=lin_model.fit(max_iterations= 100000)\n",
    "display(result_fullreg_gene.summary2())\n",
    "params_fullreg_gene = result_fullreg_gene.params\n",
    "conf_fullreg_gene = result_fullreg_gene.conf_int()\n",
    "\n",
    "print(result_fullreg_gene.pvalues)\n",
    "\n",
    "\n",
    "\n",
    "############## Association with telomere length and all CH ################\n",
    "\n",
    "data=ukb500k_ts_som[[\"Z-adjusted T/S log\", \"sex\", \"age\", \"smoking_status\", \"single_gene\", \"single_gene_vaf\", \"single_gene_vaf_0.10\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\", \"WBC_wins\",\"MO_perc_wins\", \"LY_perc_wins\", \"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",\"eid\",\"HM\"]].copy(deep=True)\n",
    "data=data.dropna()\n",
    "data=data[(data[\"HM\"]==0)]\n",
    "\n",
    "print(data.shape)\n",
    "\n",
    "data=data.rename(columns={\"Z-adjusted T/S log\":\"tl\"})\n",
    "data[\"single_gene\"]=data[\"single_gene\"].map(lambda row:row if row==\"Control\" else \"all CH\")\n",
    "formula = 'tl ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','C(single_gene,Treatment(reference=\"Control\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\"])\n",
    "\n",
    "lin_model=smf.ols(formula=formula,data=data)\n",
    "result_fullreg_allch=lin_model.fit(max_iterations= 100000)\n",
    "display(result_fullreg_allch.summary2())\n",
    "params_fullreg_allch = result_fullreg_allch.params\n",
    "conf_fullreg_allch = result_fullreg_allch.conf_int()\n",
    "\n",
    "print(result_fullreg_allch.pvalues)\n",
    "\n",
    "\n",
    "data[\"single_gene\"].value_counts()\n",
    "\n",
    "############## Association with telomere length and all CH vaf ################\n",
    "\n",
    "data=ukb500k_ts_som[[\"Z-adjusted T/S log\", \"sex\", \"age\", \"smoking_status\", \"single_gene\", \"single_gene_vaf\", \"single_gene_vaf_0.10\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\", \"WBC_wins\",\"MO_perc_wins\", \"LY_perc_wins\", \"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",\"eid\",\"HM\"]].copy(deep=True)\n",
    "data=data.dropna()\n",
    "data=data[(data[\"HM\"]==0)]\n",
    "\n",
    "print(data.shape)\n",
    "\n",
    "data=data.rename(columns={\"Z-adjusted T/S log\":\"tl\"})\n",
    "data[\"single_gene\"]=data[\"single_gene\"].map(lambda row:row if row==\"Control\" else \"all CH\")\n",
    "data.loc[(data[\"single_gene_vaf\"]==\"NA\"),\"single_gene_vaf\"]=0\n",
    "data[\"single_gene_vaf\"]=data[\"single_gene_vaf\"].astype(float)\n",
    "\n",
    "formula = 'tl ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','C(single_gene,Treatment(reference=\"Control\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",\"single_gene_vaf\"])\n",
    "\n",
    "lin_model=smf.ols(formula=formula,data=data)\n",
    "result_fullreg_allch_vaf=lin_model.fit(max_iterations= 100000)\n",
    "display(result_fullreg_allch_vaf.summary2())\n",
    "params_allch_vaf = result_fullreg_allch_vaf.params\n",
    "conf_fullreg_allch_vaf = result_fullreg_allch_vaf.conf_int()\n",
    "\n",
    "print(result_fullreg_allch_vaf.pvalues)\n",
    "\n",
    "\n",
    "############## Association with telomere length and individual gene CH vaf ################\n",
    "\n",
    "data=ukb500k_ts_som[[\"Z-adjusted T/S log\", \"sex\", \"age\", \"smoking_status\", \"single_gene\", \"single_gene_vaf\", \"single_gene_vaf_0.10\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\", \"WBC_wins\", \"MO_perc_wins\", \"LY_perc_wins\", \"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",\"eid\",\"HM\"]].copy(deep=True)\n",
    "data=data.dropna()\n",
    "data=data[(data[\"HM\"]==0)]\n",
    "\n",
    "print(data.shape)\n",
    "\n",
    "gene_vaf=[]\n",
    "for gene in data.single_gene.unique():\n",
    "    if gene != \"Control\":\n",
    "        data[gene+\"_vaf\"]=data.apply(lambda row:row[\"single_gene_vaf\"] if row[\"single_gene\"]==gene else 0,axis=1)\n",
    "        gene_vaf.append(gene+\"_vaf\")\n",
    "\n",
    "data=data.rename(columns={\"Z-adjusted T/S log\":\"tl\"})\n",
    "formula = 'tl ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','C(single_gene,Treatment(reference=\"Control\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\"]+gene_vaf)\n",
    "\n",
    "lin_model=smf.ols(formula=formula,data=data)\n",
    "result_fullreg=lin_model.fit(max_iterations= 100000)\n",
    "display(result_fullreg.summary2())\n",
    "params_fullreg = result_fullreg.params\n",
    "conf_fullreg = result_fullreg.conf_int()\n",
    "\n",
    "print(result_fullreg.pvalues)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "supptable_2_gene=pd.DataFrame(pd.concat([params_fullreg_gene,conf_fullreg_gene,result_fullreg_gene.pvalues],axis=1))\n",
    "supptable_2_gene=(supptable_2_gene.reset_index().iloc[4:18,:]).reset_index(drop=True)\n",
    "supptable_2_gene=supptable_2_gene.append((pd.concat([params_fullreg_allch,conf_fullreg_allch,result_fullreg_allch.pvalues],axis=1).loc['C(single_gene, Treatment(reference=\"Control\"))[T.all CH]',:]).to_frame().T.reset_index())\n",
    "display(supptable_2_gene)\n",
    "supptable_2_gene[\"Gene/mCA\"]=supptable_2_gene[\"index\"].map(lambda row:row.split(\"[\")[-1].strip(\"]\")[2:])\n",
    "supptable_2_gene.columns=[\"Index\",\"Coefficient\",\"Coefficient lower CI (2.5%)\",\"Coefficient upper CI (97.5%)\",\"P-value\",\"Gene/mCA\"]\n",
    "supptable_2_gene=supptable_2_gene.sort_values(by=\"Gene/mCA\",key=lambda col: col.str.lower())\n",
    "supptable_2_gene[\"FDR\"]=multipletests(supptable_2_gene[\"P-value\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n",
    "\n",
    "supptable_2=pd.concat([supptable_2_gene[[\"Gene/mCA\",\"Coefficient\",\"Coefficient lower CI (2.5%)\",\"Coefficient upper CI (97.5%)\",\"P-value\",\"FDR\",\"FDR_selected\"]],supptable_2_mca[[\"Gene/mCA\",\"Coefficient\",\"Coefficient lower CI (2.5%)\",\"Coefficient upper CI (97.5%)\",\"P-value\",\"FDR\"]]])\n",
    "supptable_2[\"FDR_all\"]=multipletests(supptable_2[\"P-value\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n",
    "\n",
    "#supptable_2.to_csv(\"coef_supptable_r1.tsv\",sep=\"\\t\",index=False)\n",
    "\n",
    "#supptable_2.to_csv(\"coef_supptable_singlegene_r1_n.tsv\",sep=\"\\t\",index=False)\n",
    "\n",
    "supptable_2\n",
    "\n",
    "supptable_3_gene=pd.DataFrame(pd.concat([params_fullreg,conf_fullreg,result_fullreg.pvalues],axis=1))\n",
    "supptable_3_gene=(supptable_3_gene.reset_index().iloc[35:49,:]).reset_index(drop=True)\n",
    "supptable_3_gene=supptable_3_gene.append((pd.concat([params_allch_vaf,conf_fullreg_allch_vaf,result_fullreg_allch_vaf.pvalues],axis=1).loc['single_gene_vaf',:]).to_frame().T.reset_index())\n",
    "supptable_3_gene[\"Gene/mCA\"]=supptable_3_gene[\"index\"].map(lambda row:row[:-4])\n",
    "supptable_3_gene.columns=[\"Index\",\"Coefficient\",\"Coefficient lower CI (2.5%)\",\"Coefficient upper CI (97.5%)\",\"P-value\",\"Gene/mCA\"]\n",
    "supptable_3_gene=supptable_3_gene.sort_values(by=\"Gene/mCA\",key=lambda col: col.str.lower())\n",
    "supptable_3_gene[\"FDR\"]=multipletests(supptable_3_gene[\"P-value\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n",
    "supptable_3_gene.loc[(supptable_3_gene[\"Gene/mCA\"].isin([\"ASXL1\",\"CBL\",\"GNAS\", \"GNB1\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TP53\",\"U2AF1\"])),\"FDR_selected\"]=multipletests(supptable_3_gene.loc[(supptable_3_gene[\"Gene/mCA\"].isin([\"ASXL1\",\"CBL\",\"GNAS\", \"GNB1\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TP53\",\"U2AF1\"])),\"P-value\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n",
    "\n",
    "supptable_3=pd.concat([supptable_3_gene[[\"Gene/mCA\",\"Coefficient\",\"Coefficient lower CI (2.5%)\",\"Coefficient upper CI (97.5%)\",\"P-value\",\"FDR\",\"FDR_selected\"]],supptable_3_mca[[\"Gene/mCA\",\"Coefficient\",\"Coefficient lower CI (2.5%)\",\"Coefficient upper CI (97.5%)\",\"P-value\",\"FDR\"]]])\n",
    "supptable_3[\"FDR_all\"]=multipletests(supptable_3[\"P-value\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n",
    "\n",
    "#supptable_3.to_csv(\"vaf_coef_supptable_r1.tsv\",sep=\"\\t\",index=False)\n",
    "display(supptable_3)\n",
    "\n",
    "\n",
    "supptable_3_gene\n",
    "\n",
    "#supptable_3.to_csv(\"vaf_coef_supptable_singlegene_r1_check.tsv\",sep=\"\\t\",index=False)\n",
    "\n",
    "\n",
    "supptable_3\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "left_0, width_0 = 0.04, 14*0.05\n",
    "left_1, width_1 = left_0+width_0+0.02,4*0.05\n",
    "\n",
    "\n",
    "bottom_2, height_2 = 0.1,0.076*1.5\n",
    "gap_0_1=0.01\n",
    "bottom_1, height_1 = bottom_2+height_2+gap_0_1,0.076*5\n",
    "gap_1_2=0.03\n",
    "bottom_0, height_0 = bottom_1+height_1+gap_1_2,0.076*6\n",
    "\n",
    "plt.figure(figsize=(26,12))\n",
    "axes=np.array([np.array([plt.axes([left_0, bottom_0, width_0, height_0]),plt.axes([left_1, bottom_0, width_1, height_0])]),np.array([plt.axes([left_0, bottom_1, width_0, height_1]),plt.axes([left_1, bottom_1, width_1, height_1])]),np.array([plt.axes([left_0, bottom_2, width_0, height_2]),plt.axes([left_1, bottom_2, width_1, height_2])])])\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "fig, axes = plt.subplots(3, 2,sharex=False,sharey=False,figsize=(28,12),gridspec_kw={'height_ratios':[5,5,1.5],'width_ratios':[15,4]})\n",
    "y=\"tl\"\n",
    "temp=data[[\"single_gene\",y]].copy(deep=True)\n",
    "gene_count=Counter(temp[\"single_gene\"])\n",
    "\n",
    "gene_color_new={}\n",
    "for key in gene_color:\n",
    "    gene_color_new[key]=sns.set_hls_values(gene_color[key],s=0.5)\n",
    "black=sns.set_hls_values('#000000',s=0.5)\n",
    "\n",
    "gene_list=[]\n",
    "gene_groups=[]\n",
    "quartiles=[]\n",
    "\n",
    "gene_list.append(\"Control\")\n",
    "gene_groups.append(np.array(temp.loc[((temp[\"single_gene\"]==\"Control\") & (temp[y].notna())),y]))\n",
    "quartiles.append(np.percentile(gene_groups[-1],[25, 50, 75]))\n",
    "\n",
    "gene_list.append(\"all CH\")\n",
    "gene_groups.append(np.array(temp.loc[((temp[\"single_gene\"].isin([\"ASXL1\",\"CALR\",\"CBL\",\"DNMT3A_R882\",\"DNMT3A_other\",\"GNAS\",\"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"U2AF1\"])) & (temp[y].notna())),y]))\n",
    "quartiles.append(np.percentile(gene_groups[-1],[25, 50, 75]))\n",
    "\n",
    "for key in [\"ASXL1\",\"CALR\",\"CBL\",\"DNMT3A_R882\",\"DNMT3A_other\",\"GNAS\",\"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"U2AF1\"]:\n",
    "        if gene_count[key]>=20:\n",
    "            gene_list.append(key)\n",
    "            gene_groups.append(np.array(temp.loc[((temp[\"single_gene\"]==key) & (temp[y].notna())),y]))\n",
    "            quartiles.append(np.percentile(gene_groups[-1],[25, 50, 75]))\n",
    "            print(gene_groups[-1].shape,gene_count[key])\n",
    "quartiles=np.array(quartiles)           \n",
    "            \n",
    "if gene_list != []:\n",
    "        temp=temp.loc[((temp[\"single_gene\"].isin(gene_list)) & (temp[y].notna())),:]\n",
    "        counts=temp[\"single_gene\"].value_counts()\n",
    "        \n",
    "        parts = axes[0,0].violinplot(gene_groups, showmeans=False, showmedians=False, showextrema=False,widths=0.9,positions=np.arange(-2,2*len(gene_list)-2,2))\n",
    "\n",
    "        for i,pc in enumerate(parts['bodies']):\n",
    "            if gene_list[i]==\"Control\":\n",
    "                pc.set_facecolor(gene_color_new[gene_list[i]])\n",
    "            else:\n",
    "                pc.set_facecolor(\"#999999\")\n",
    "            pc.set_edgecolor(black)\n",
    "            pc.set_alpha(1)\n",
    "            pc.set_linewidth(1.5)\n",
    "\n",
    "        axes[0,0].boxplot(gene_groups,positions=np.arange(-2,2*len(gene_list)-2,2),showcaps=False,widths=0.12, patch_artist=True, boxprops=dict(color=black, facecolor=black),whiskerprops=dict(color=black, linewidth=2), medianprops=dict(color=\"w\", linewidth=0 ),showfliers=False)\n",
    "        inds = np.arange(0, len(quartiles))\n",
    "        axes[0,0].scatter(np.arange(-2,2*len(gene_list)-2,2), quartiles[:,1], marker='o', color='white', s=30, zorder=3)\n",
    "        \n",
    "        axes[0,0].tick_params(axis='y',labelsize=15)\n",
    "        axes[0,0].tick_params(axis='x',labelsize=15)\n",
    "        axes[0,0].grid(axis='x')\n",
    "        axes[0,0].xaxis.label.set_size(20)\n",
    "        axes[0,0].yaxis.label.set_size(20)\n",
    "        axes[0,0].set_ylabel(\"LTL\")\n",
    "        axes[0,0].set_xlabel(\"CH driver gene\",labelpad=8)\n",
    "        axes[0,0].axhline(y=temp.loc[temp[\"single_gene\"]==\"Control\",y].median(), color='k', linestyle='--')\n",
    "        axes[0,0].set_ylim([-5,5])\n",
    "        axes[0,0].xaxis.set_tick_params(which='both', labelbottom=True)\n",
    "        print(temp.loc[temp[\"single_gene\"]==\"Control\",y].median())  \n",
    "        ticklabels=[\"Control\"]+[\"All CH\\n*\" if result_fullreg_allch.pvalues['C(single_gene, Treatment(reference=\"Control\"))[T.all CH]']<0.05 else \"All CH\"]+[x+\"\\n*\" if result_fullreg_gene.pvalues['C(single_gene, Treatment(reference=\"Control\"))[T.'+x+']']<0.05 else x for x in gene_list[2:]]\n",
    "        print(ticklabels)\n",
    "        axes[0,0].set_xticks(np.arange(-2,30,2))\n",
    "        axes[0,0].set_xticklabels(ticklabels,rotation=15)\n",
    "        axes[0,0].spines[['right', 'top']].set_visible(False)\n",
    "                       \n",
    "temp=data_mca[[y,\"MCA_type\",\"CELL_FRAC\"]].copy(deep=True)\n",
    "temp=temp.loc[(temp[y].notna()),:]\n",
    "temp[\"MCA_type\"]=temp[\"MCA_type\"].map(lambda row:\"autosomal\" if row not in [\"Y:loss\",\"X:loss\",\"Control\"] else row)\n",
    "counts=temp[\"MCA_type\"].value_counts()\n",
    "print(counts)\n",
    "gene_groups=[]\n",
    "gene_order=[\"Control\",\"autosomal\",\"Y:loss\",\"X:loss\"]\n",
    "quartiles=[]\n",
    "for key in gene_order:\n",
    "    gene_groups.append(np.array(temp.loc[(temp[\"MCA_type\"]==key)& (temp[y].notna()),y]))\n",
    "    quartiles.append(np.percentile(gene_groups[-1],[25, 50, 75]))\n",
    "quartiles=np.array(quartiles)           \n",
    "                       \n",
    "parts = axes[0,1].violinplot(gene_groups, showmeans=False, showmedians=False, showextrema=False,widths=0.9,positions=np.arange(-2,6,2))\n",
    "mca_color={\"Control\":gene_color[\"Control\"],\"autosomal\":\"#999999\",\"Y:loss\":\"#E69F00\",\"X:loss\":\"#56B4E9\"}           \n",
    "for key in mca_color:\n",
    "    mca_color[key]=sns.set_hls_values(mca_color[key],s=0.5)\n",
    "    \n",
    "for i,pc in enumerate(parts['bodies']):\n",
    "            if gene_order[i]==\"Control\":\n",
    "                pc.set_facecolor(mca_color[gene_order[i]])\n",
    "            else:\n",
    "                pc.set_facecolor(\"#8585ad\")\n",
    "            pc.set_edgecolor(black)\n",
    "            pc.set_alpha(1)\n",
    "            pc.set_linewidth(1.5)\n",
    "\n",
    "axes[0,1].boxplot(gene_groups,positions=np.arange(-2,6,2),showcaps=False,widths=0.12, patch_artist=True,boxprops=dict(color=black, facecolor=black),\n",
    "            whiskerprops=dict(color=black, linewidth=2),\n",
    "            medianprops=dict(color=\"w\", linewidth=0 ),showfliers=False)                       \n",
    "inds = np.arange(0, len(quartiles))\n",
    "axes[0,1].scatter(np.arange(-2,6,2), quartiles[:,1], marker='o', color='white', s=30, zorder=3)\n",
    "\n",
    "axes[0,1].set_xticks(np.arange(-2,6,2))\n",
    "axes[0,1].set_xticklabels(gene_order,rotation=35)\n",
    "axes[0,1].spines[['right', 'top']].set_visible(False)\n",
    "#axes[0,1].set_yticklabels([])\n",
    "axes[0,1].grid(axis='x')\n",
    "axes[0,1].tick_params(axis='x',labelsize=15)\n",
    "axes[0,1].tick_params(axis='y',labelsize=15)\n",
    "axes[0,1].xaxis.label.set_size(20)\n",
    "axes[0,1].yaxis.label.set_size(20)\n",
    "axes[0,1].set_xlabel(\"mCA type\",labelpad=20)\n",
    "axes[0,1].set_ylim([-5,5])\n",
    "axes[0,1].axhline(y=temp.loc[temp[\"MCA_type\"]==\"Control\",y].median(), color='k', linestyle='--')\n",
    "axes[0,1].xaxis.set_tick_params(which='both', labelbottom=True)\n",
    "ticklabels=[\"Control\"]+[x+\"\\n*\" if plot_values_mca[i,3]<0.05 else x for i,x in enumerate([\"autosomal\",\"LOY\",\"LOX\"])]\n",
    "axes[0,1].set_xticklabels(ticklabels,rotation=15)\n",
    "\n",
    "###################################################################################################\n",
    "\n",
    "\n",
    "y=\"tl\"#\"Z-adjusted T/S log\"\n",
    "temp=data[[\"single_gene\",\"single_gene_vaf\",\"single_gene_vaf_0.10\",y]].copy(deep=True)\n",
    "temp=temp[(temp[\"single_gene\"]!=\"NA\") & (temp[y].notna())]\n",
    "gene_order_plot=[\"all CH\", \"ASXL1\", \"CALR\", \"CBL\", \"DNMT3A_R882\", \"DNMT3A_other\", \"GNAS\", \"GNB1\", \"JAK2\", \"PPM1D\", \"SF3B1\",\"SRSF2\", \"TET2\",\"TP53\",\"U2AF1\"]\n",
    "black=sns.set_hls_values('#000000',s=0.5)\n",
    "\n",
    "\n",
    "for vaf in [\"0.10\"]:\n",
    "        gene_vaf_groups=temp.groupby([\"single_gene\",\"single_gene_vaf_0.10\"])\n",
    "        gene_vaf_groups_dict={}\n",
    "        \n",
    "        gene_vaf_groups_dict[\"all CH_0\"]=np.array(temp.loc[((temp[\"single_gene\"].isin(gene_order_plot)) & (temp[\"single_gene_vaf_0.10\"]==0)),y].dropna())\n",
    "        gene_vaf_groups_dict[\"all CH_1\"]=np.array(temp.loc[((temp[\"single_gene\"].isin(gene_order_plot)) & (temp[\"single_gene_vaf_0.10\"]==1)),y].dropna())\n",
    "        \n",
    "        for group_name, df_group in gene_vaf_groups:\n",
    "            gene_vaf_groups_dict[group_name[0]+\"_\"+str(group_name[1])]=np.array(df_group[y].dropna())\n",
    "        \n",
    "        gene_vaf_group_list=[gene_vaf_groups_dict[\"Control_NA\"]]\n",
    "        pos=[-2]\n",
    "        colors_violin=[sns.set_hls_values(gene_color[\"Control\"],s=0.5)]\n",
    "        percentiles=[np.percentile(gene_vaf_groups_dict[\"Control_NA\"],[25, 50, 75])]\n",
    "        \n",
    "        for i,gene in enumerate(gene_order_plot):\n",
    "            gene_vaf_group_list.append(gene_vaf_groups_dict[gene+\"_0\"])\n",
    "            gene_vaf_group_list.append(gene_vaf_groups_dict[gene+\"_1\"])\n",
    "            pos+=[i*2-0.45,i*2+0.45]\n",
    "            #colors_violin+=[sns.set_hls_values('#70db70',s=0.5),sns.set_hls_values(\"#ff99e6\",s=0.5)]\n",
    "            colors_violin+=['#737373',\"#bfbfbf\"]#808080\n",
    "            percentiles.append(np.percentile(gene_vaf_groups_dict[gene+\"_0\"],[25, 50, 75]))\n",
    "            percentiles.append(np.percentile(gene_vaf_groups_dict[gene+\"_1\"],[25, 50, 75]))\n",
    "        percentiles=np.array(percentiles)\n",
    "        axes[1,0].set_ylim([-5,5])\n",
    "        counts=temp[[\"single_gene\",\"single_gene_vaf_\"+str(vaf)]].value_counts()\n",
    "        counts.loc[\"all CH\",0]=gene_vaf_groups_dict[\"all CH_0\"].shape[0]\n",
    "        counts.loc[\"all CH\",1]=gene_vaf_groups_dict[\"all CH_1\"].shape[0]\n",
    "        display (counts)\n",
    "        parts=axes[1,0].violinplot(gene_vaf_group_list, pos, widths=0.6,\n",
    "                     showmedians=False,showextrema=False)\n",
    "        for i,pc in enumerate(parts['bodies']):\n",
    "            pc.set_facecolor(colors_violin[i])\n",
    "            pc.set_edgecolor(black)\n",
    "            pc.set_alpha(1)\n",
    "            pc.set_linewidth(1.5)\n",
    "        axes[1,0].boxplot(gene_vaf_group_list,positions=pos,showcaps=False,widths=0.08, patch_artist=True, boxprops=dict(color=black, facecolor=black),whiskerprops=dict(color=black, linewidth=2), medianprops=dict(color=\"w\", linewidth=0 ),showfliers=False)\n",
    "        \n",
    "        quartile1, medians, quartile3 = percentiles[:,0],percentiles[:,1],percentiles[:,2]\n",
    "\n",
    "        axes[1,0].scatter(pos, medians, marker='o', color='white', s=25, zorder=3)\n",
    "        \n",
    "        xticks_new=[\"Control\\n(\"+str(counts[\"Control\",\"NA\"])+\")\"]\n",
    "        result_fullreg_pvalues=result_fullreg.pvalues\n",
    "        result_fullreg_pvalues[\"all CH_vaf\"]=result_fullreg_allch_vaf.pvalues[\"single_gene_vaf\"]\n",
    "        for i in gene_order_plot:\n",
    "            print(i,counts[i])\n",
    "            try:\n",
    "                low_vaf=str(counts[i,0])\n",
    "            except KeyError:\n",
    "                low_vaf=\"0\"\n",
    "            try:\n",
    "                high_vaf=str(counts[i,1])\n",
    "            except KeyError:\n",
    "                high_vaf=\"0\"\n",
    "            if result_fullreg_pvalues[i+\"_vaf\"]<0.05:\n",
    "                xticks_new.append(i[0].upper()+i[1:]+\"\\n(\"+low_vaf+\",\"+high_vaf+\")\\n*\")\n",
    "            else:\n",
    "                xticks_new.append(i[0].upper()+i[1:]+\"\\n(\"+low_vaf+\",\"+high_vaf+\")\")\n",
    "\n",
    "        print(xticks_new)\n",
    "        \n",
    "        axes[1,0].axhline(y=temp.loc[temp[\"single_gene\"]==\"Control\",y].median(), color='k', linestyle='--')\n",
    "        axes[1,0].set_ylabel(\"LTL\",fontsize=20)\n",
    "        axes[1,0].set_xlabel(\"\")\n",
    "        axes[1,0].set_xticks([])\n",
    "        axes[1,0].spines[['right', 'top']].set_visible(False)\n",
    "        axes[1,0].tick_params(labelsize=15)\n",
    "        legend_handles = [mpatches.Patch(color=sns.set_hls_values(gene_color[\"Control\"],s=0.5), label='Control'),mpatches.Patch(color='#737373',label=\"VAF \"+r'$\\mathrm{<} $'+vaf),mpatches.Patch(color='#bfbfbf',label=\"VAF \"+r'$\\mathrm{\\geq} $'+vaf)]\n",
    "        axes[1,0].legend(handles=legend_handles,fontsize=15,loc=\"upper left\")\n",
    "        for i in range(1,len(pos),2): \n",
    "            if result_fullreg_pvalues[gene_order_plot[int((i-1)/2)]+\"_vaf\"]<0.05:\n",
    "                if medians[i+1]-medians[i]>=0:\n",
    "                    axes[1,0].arrow(pos[i],medians[i],pos[i+1]-pos[i],medians[i+1]-medians[i],width=0.05,length_includes_head=True,head_width=0.45,head_length=0.4,linestyle=(0, (5,10)),color='red')\n",
    "                else:\n",
    "                    axes[1,0].arrow(pos[i],medians[i],pos[i+1]-pos[i],medians[i+1]-medians[i],width=0.05,length_includes_head=True,head_width=0.45,head_length=0.4,linestyle=(0, (5,10)),color='b')\n",
    "        #axes[1,0].arrow(pos[1],medians[1],pos[2]-pos[1],medians[2]-medians[1],width=0.05,length_includes_head=True,head_width=0.45,head_length=0.4,linestyle=(0, (5,10)),color='b')\n",
    "                \n",
    "gene_vaf_order=[x+\"_vaf\" for x in gene_order_plot]\n",
    "params_fullreg[\"all CH_vaf\"]=result_fullreg_allch_vaf.params[\"single_gene_vaf\"]\n",
    "conf_fullreg.loc[\"all CH_vaf\"]=conf_fullreg_allch_vaf.loc[\"single_gene_vaf\"]\n",
    "\n",
    "axes[2,0].bar(np.arange(0,30,2),params_fullreg[gene_vaf_order],yerr=[params_fullreg[gene_vaf_order]-conf_fullreg.loc[gene_vaf_order,0],conf_fullreg.loc[gene_vaf_order,1]-params_fullreg[gene_vaf_order]],width=1.2)\n",
    "axes[2,0].set_ylabel(\"Coefficient\",fontsize=20)\n",
    "axes[2,0].spines[['right', 'top']].set_visible(False)\n",
    "axes[2,0].axhline(y=0, color='k', linestyle='--', lw=1)\n",
    "axes[2,0].tick_params(labelsize=15)\n",
    "axes[2,0].set_xticks(np.arange(-2,30,2))\n",
    "print(xticks_new)\n",
    "axes[2,0].set_xticklabels(xticks_new,fontsize=15,rotation=35)\n",
    "\n",
    "\n",
    "#gene_order_plot=[x+\"\\n*\" if result_fullreg.pvalues[x+\"_vaf\"]<0.05 else x for x in gene_order_plot ]\n",
    "#axes[2,0].set_xticklabels([\"Control\"]+gene_order_plot,fontsize=15,rotation=35)\n",
    "\n",
    "print([\"Control\"]+gene_order_plot)\n",
    "plt.subplots_adjust(hspace=0.05)\n",
    "plt.subplots_adjust(wspace=0.05)\n",
    "\n",
    "y=\"tl\"\n",
    "temp=data_mca[[y,\"MCA_type\",\"CELL_FRAC\"]].copy(deep=True)\n",
    "temp[\"mca_0.2\"]=temp[\"CELL_FRAC\"].map(lambda row:\"1\" if row>=0.2 else \"0\",na_action='ignore')\n",
    "temp[\"MCA_type\"]=temp[\"MCA_type\"].map(lambda row:\"autosomal\" if row not in [\"Y:loss\",\"X:loss\",\"Control\"] else row)\n",
    "temp=temp.loc[(temp[y].notna()),:]\n",
    "print(temp[[\"mca_0.2\",\"MCA_type\"]].value_counts())\n",
    "temp.loc[temp[\"MCA_type\"]==\"Control\",\"mca_0.2\"]=\"NA\"\n",
    "\n",
    "gene_vaf_groups=temp.groupby([\"MCA_type\",\"mca_0.2\"])\n",
    "gene_vaf_groups_dict={}\n",
    "for group_name, df_group in gene_vaf_groups:\n",
    "    gene_vaf_groups_dict[group_name[0]+\"_\"+str(group_name[1])]=np.array(df_group[y].dropna())\n",
    "    \n",
    "gene_vaf_group_list=[gene_vaf_groups_dict[\"Control_NA\"]]\n",
    "pos=[-2]\n",
    "colors_violin=[sns.set_hls_values(gene_color[\"Control\"],s=0.5)]\n",
    "percentiles=[np.percentile(gene_vaf_groups_dict[\"Control_NA\"],[25, 50, 75])]\n",
    "\n",
    "for i,gene in enumerate([\"autosomal\",\"Y:loss\",\"X:loss\"]):\n",
    "    gene_vaf_group_list.append(gene_vaf_groups_dict[gene+\"_0\"])\n",
    "    gene_vaf_group_list.append(gene_vaf_groups_dict[gene+\"_1\"])\n",
    "    pos+=[i*2-0.45,i*2+0.45]\n",
    "    colors_violin+=['#5c5c8a',\"#b3b3cc\"]\n",
    "    percentiles.append(np.percentile(gene_vaf_groups_dict[gene+\"_0\"],[25, 50, 75]))\n",
    "    percentiles.append(np.percentile(gene_vaf_groups_dict[gene+\"_1\"],[25, 50, 75]))\n",
    "percentiles=np.array(percentiles)\n",
    "      \n",
    "axes[1,1].set_ylim([-5,5])\n",
    "counts=temp[[\"MCA_type\",\"mca_0.2\"]].value_counts()\n",
    "display (counts)\n",
    "parts=axes[1,1].violinplot(gene_vaf_group_list, pos, widths=0.6,\n",
    "                     showmedians=False,showextrema=False)\n",
    "for i,pc in enumerate(parts['bodies']):\n",
    "    pc.set_facecolor(colors_violin[i])\n",
    "    pc.set_edgecolor(black)\n",
    "    pc.set_alpha(1)\n",
    "    pc.set_linewidth(1.5)\n",
    "axes[1,1].boxplot(gene_vaf_group_list,positions=pos,showcaps=False,widths=0.08, patch_artist=True, boxprops=dict(color=black, facecolor=black),whiskerprops=dict(color=black, linewidth=2), medianprops=dict(color=\"w\", linewidth=0 ),showfliers=False)\n",
    "        \n",
    "quartile1, medians, quartile3 = percentiles[:,0],percentiles[:,1],percentiles[:,2]\n",
    "\n",
    "axes[1,1].scatter(pos, medians, marker='o', color='white', s=25, zorder=3)\n",
    "\n",
    "xticks_new=[\"Control\\n(\"+str(counts[\"Control\",\"NA\"])+\")\"]\n",
    "gene_rename={\"autosomal\":\"autosomal\",\"Y:loss\":\"LOY\",\"X:loss\":\"LOX\"}\n",
    "for i,gene in enumerate([\"autosomal\",\"Y:loss\",\"X:loss\"]):\n",
    "    try:\n",
    "        low_vaf=str(counts[gene,\"0\"])\n",
    "    except KeyError:\n",
    "        low_vaf=\"0\"\n",
    "    try:\n",
    "        high_vaf=str(counts[gene,\"1\"])\n",
    "    except KeyError:\n",
    "        high_vaf=\"0\"\n",
    "    if plot_values[i,3]<0.05:\n",
    "        xticks_new.append(gene_rename[gene]+\"\\n(\"+low_vaf+\",\"+high_vaf+\")\\n*\")\n",
    "    else:\n",
    "        xticks_new.append(gene_rename[gene]+\"\\n(\"+low_vaf+\",\"+high_vaf+\")\")\n",
    "        \n",
    "axes[1,1].axhline(y=temp.loc[temp[\"MCA_type\"]==\"Control\",y].median(), color='k', linestyle='--')\n",
    "        \n",
    "\n",
    "axes[1,1].set_xlabel(\"\")\n",
    "axes[1,1].set_xticks([])\n",
    "legend_handles = [mpatches.Patch(color=sns.set_hls_values(gene_color[\"Control\"],s=0.5),label='Control'),mpatches.Patch(color='#5c5c8a',label=\"Cell fraction\"+r'$\\mathrm{<}$'+\" 0.2\"),mpatches.Patch(color=\"#bfbfbf\",label=\"Cell fraction\"+r'$\\mathrm{\\geq}$'+\" 0.2\")]\n",
    "axes[1,1].legend(handles=legend_handles,fontsize=15)\n",
    "for i in range(1,len(pos),2):\n",
    "            if plot_values[int((i-1)/2),3]<0.05:\n",
    "                if medians[i+1]-medians[i]>=0:\n",
    "                    axes[1,1].arrow(pos[i],medians[i],pos[i+1]-pos[i],medians[i+1]-medians[i],width=0.05,length_includes_head=True,head_width=0.45,head_length=0.4,linestyle=(0, (5,10)),color='red')\n",
    "                else:\n",
    "                    axes[1,1].arrow(pos[i],medians[i],pos[i+1]-pos[i],medians[i+1]-medians[i],width=0.05,length_includes_head=True,head_width=0.45,head_length=0.4,linestyle=(0, (5,10)),color='b')\n",
    "\n",
    "\n",
    "axes[1,1].spines[['right', 'top']].set_visible(False)\n",
    "axes[1,1].tick_params(axis='y',labelsize=15)\n",
    "\n",
    "\n",
    "axes[2,1].bar(np.arange(0,6,2),plot_values[:,0],yerr=[plot_values[:,0]-plot_values[:,1],plot_values[:,2]-plot_values[:,0]],width=1.2)\n",
    "ticklabel=[\"autosomal\",\"Y:loss\",\"X:loss\"]\n",
    "ticklabel=[\"Control\"]+[ticklabel[i]+\"\\n*\" if plot_values[i,3]<0.05 else ticklabel[i] for i in range(3)]\n",
    "axes[2,1].set_xticks(np.arange(-2,6,2))\n",
    "axes[2,1].set_xticklabels(xticks_new,rotation=45)\n",
    "#plt.yticks([-0.15,-0.10,-0.05])\n",
    "axes[2,1].axhline(y=0,linestyle=\"--\",lw=1)\n",
    "axes[2,1].tick_params('both',labelsize=15)\n",
    "axes[2,1].spines[['right', 'top']].set_visible(False)\n",
    "\n",
    "\n",
    "axes[0,0].set_xlim([-2.5, 28.95])\n",
    "axes[1,0].set_xlim([-2.5, 28.95])\n",
    "axes[2,0].set_xlim([-2.5, 28.95])\n",
    "axes[0,1].set_xlim([-2.5, 4.95])\n",
    "axes[1,1].set_xlim([-2.5, 4.95])\n",
    "axes[2,1].set_xlim([-2.5, 4.95])\n",
    "plt.subplots_adjust(wspace=0.05,hspace=0.1)\n",
    "\"\"\"\n",
    "plt.savefig(\"../figures_revision1/tl_ch_mca_singlegene_noHM.png\",dpi=600,bbox_inches=\"tight\")\n",
    "plt.savefig(\"../figures_revision1/tl_ch_mca_singlegene_noHM.pdf\",bbox_inches=\"tight\")\n",
    "plt.savefig(\"../figures_revision1/tl_ch_mca_singlegene_noHM.svg\",bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bd6fe68",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get all relevant data (OR, frequencies, pvalues and counts)\n",
    "#Include VAF<0.1\n",
    "######## Figure 2C##############\n",
    "\n",
    "n=2\n",
    "freq_all,prs_freq_all,tl_freq_all,contigency_table_all,prs_counts,tl_counts,or_all,pv_all,prs_or_all,prs_pv_all,tl_or_all,tl_pv_all=[[] for i in range(n)],[[] for i in range(n)],[[] for i in range(n)],[[] for i in range(n)],[[] for i in range(n)],[[] for i in range(n)],[[] for i in range(n)],[[] for i in range(n)],[[] for i in range(n)],[[] for i in range(n)],[[] for i in range(n)],[[] for i in range(n)]\n",
    "\n",
    "for i,n in enumerate([2]):\n",
    "    temp=ukb500k_ts_som.loc[:,[\"total_PRS\",\"Z-adjusted T/S log\",\"single_gene\", \"single_gene_vaf\",\"HM\"] ].copy(deep=True)\n",
    "    temp=temp[(temp[\"total_PRS\"].notna()) & (temp[\"Z-adjusted T/S log\"].notna()) & (temp[\"HM\"]==0)]\n",
    "    temp[\"prs_q\"]=pd.qcut(temp[\"total_PRS\"],q = n, labels = False)\n",
    "    temp[\"tl_q\"]=pd.qcut(temp[\"Z-adjusted T/S log\"],q = n, labels = False)\n",
    "    ref_freq=pd.crosstab(temp[\"prs_q\"],temp[\"tl_q\"])\n",
    "    freq_row,prs_freq_row,tl_freq_row,contigency_table_row,prs_counts_row,tl_counts_row,or_row,pv_row=[],[],[],[],[],[],[],[]\n",
    "    for j,gene in enumerate([\"DNMT3A_other\",\"TET2\",\"SRSF2\",\"PPM1D\"]):\n",
    "        temp_1=temp.loc[(temp[\"single_gene\"]==gene),[\"prs_q\",\"tl_q\",\"single_gene_vaf\"]]\n",
    "        temp_1=temp_1[(temp_1[\"single_gene_vaf\"].astype(float)<0.1)]\n",
    "        s=temp_1.shape[0]\n",
    "        #print(pd.crosstab(temp_1[\"prs_q\"],temp_1[\"tl_q\"])*temp.shape[0],(ref_freq*s))\n",
    "        freq=pd.crosstab(temp_1[\"prs_q\"],temp_1[\"tl_q\"])*temp.shape[0]/(ref_freq*s)\n",
    "        freq_all[i].append(freq)\n",
    "        prs_freq_all[i].append(temp_1[\"prs_q\"].value_counts().reindex(np.arange(n))*temp.shape[0]/(temp[\"prs_q\"].value_counts().reindex(np.arange(n))*s))\n",
    "        tl_freq_all[i].append(temp_1[\"tl_q\"].value_counts().reindex(np.arange(n))*temp.shape[0]/(temp[\"tl_q\"].value_counts().reindex(np.arange(n))*s))\n",
    "        contigency_table_all[i].append(pd.crosstab(temp_1[\"prs_q\"],temp_1[\"tl_q\"]))\n",
    "        prs_counts[i].append(temp_1[\"prs_q\"].value_counts().reindex(np.arange(n)))\n",
    "        tl_counts[i].append(temp_1[\"tl_q\"].value_counts().reindex(np.arange(n)))\n",
    "        or_table_all,pvalue_table_all=[[] for x in range(n)],[[] for x in range(n)]\n",
    "        or_table_prs,pvalue_table_prs,or_table_tl,pvalue_table_tl=[],[],[],[]\n",
    "        \n",
    "        temp_1=temp.copy(deep=True)\n",
    "        temp_1[gene]=0\n",
    "        temp_1.loc[(temp_1[\"single_gene_vaf\"]==\"NA\"),\"single_gene_vaf\"]=0\n",
    "        temp_1.loc[(temp_1[\"single_gene\"]==gene) & (temp_1[\"single_gene_vaf\"].astype(float)<0.1),gene]=1\n",
    "        for y in range(n):\n",
    "            for z in range(n):\n",
    "                temp_1[\"prs_tl\"]=0\n",
    "                temp_1.loc[(temp_1[\"prs_q\"]==y) & (temp_1[\"tl_q\"]==z),\"prs_tl\"]=1\n",
    "                fe_or,fe_pv=fisher_exact(pd.crosstab(temp_1[gene],temp_1[\"prs_tl\"]))\n",
    "                or_table_all[y].append(fe_or)\n",
    "                pvalue_table_all[y].append(fe_pv)\n",
    "                \n",
    "        or_all[i].append(np.array([np.array(row,dtype=float) for row in or_table_all],dtype=float))\n",
    "        pv_all[i].append(np.array([np.array(row,dtype=float) for row in pvalue_table_all],dtype=float))\n",
    "                \n",
    "        for y in range(n):\n",
    "            temp_1[\"prs_group\"]=0\n",
    "            temp_1.loc[(temp_1[\"prs_q\"]==y),\"prs_group\"]=1\n",
    "            #print(pd.crosstab(temp_1[gene],temp_1[\"prs_group\"]))\n",
    "            fe_or,fe_pv=fisher_exact(pd.crosstab(temp_1[gene],temp_1[\"prs_group\"]))\n",
    "            or_table_prs.append(fe_or)\n",
    "            pvalue_table_prs.append(fe_pv)\n",
    "            \n",
    "            temp_1[\"tl_group\"]=0\n",
    "            temp_1.loc[(temp_1[\"tl_q\"]==y),\"tl_group\"]=1\n",
    "            fe_or,fe_pv=fisher_exact(pd.crosstab(temp_1[gene],temp_1[\"tl_group\"]))\n",
    "            or_table_tl.append(fe_or)\n",
    "            pvalue_table_tl.append(fe_pv)\n",
    "        \n",
    "        prs_or_all[i].append(np.array(or_table_prs))\n",
    "        prs_pv_all[i].append(np.array(pvalue_table_prs))\n",
    "        tl_or_all[i].append(np.array(or_table_tl))\n",
    "        tl_pv_all[i].append(np.array(pvalue_table_tl))\n",
    "        \n",
    "        \n",
    "    i=1\n",
    "    temp=ukb500k_ts_som.loc[:,[\"total_PRS\",\"Z-adjusted T/S log\",\"single_gene\", \"single_gene_vaf\"] ].copy(deep=True)\n",
    "    temp=temp[(temp[\"total_PRS\"].notna()) & (temp[\"Z-adjusted T/S log\"].notna())]\n",
    "    temp[\"prs_q\"]=pd.qcut(temp[\"total_PRS\"],q = n, labels = False)\n",
    "    temp[\"tl_q\"]=pd.qcut(temp[\"Z-adjusted T/S log\"],q = n, labels = False)\n",
    "    ref_freq=pd.crosstab(temp[\"prs_q\"],temp[\"tl_q\"])\n",
    "    freq_row,prs_freq_row,tl_freq_row,contigency_table_row,prs_counts_row,tl_counts_row,or_row,pv_row=[],[],[],[],[],[],[],[]\n",
    "    for j,gene in enumerate([\"DNMT3A_other\",\"TET2\",\"SRSF2\",\"PPM1D\"]):\n",
    "        temp_1=temp.loc[(temp[\"single_gene\"]==gene),[\"prs_q\",\"tl_q\",\"single_gene_vaf\"]]\n",
    "        temp_1=temp_1[(temp_1[\"single_gene_vaf\"].astype(float)>=0.1)]\n",
    "        s=temp_1.shape[0]\n",
    "        #print(pd.crosstab(temp_1[\"prs_q\"],temp_1[\"tl_q\"])*temp.shape[0],(ref_freq*s))\n",
    "        freq=pd.crosstab(temp_1[\"prs_q\"],temp_1[\"tl_q\"])*temp.shape[0]/(ref_freq*s)\n",
    "        freq_all[i].append(freq)\n",
    "        prs_freq_all[i].append(temp_1[\"prs_q\"].value_counts().reindex(np.arange(n))*temp.shape[0]/(temp[\"prs_q\"].value_counts().reindex(np.arange(n))*s))\n",
    "        tl_freq_all[i].append(temp_1[\"tl_q\"].value_counts().reindex(np.arange(n))*temp.shape[0]/(temp[\"tl_q\"].value_counts().reindex(np.arange(n))*s))\n",
    "        contigency_table_all[i].append(pd.crosstab(temp_1[\"prs_q\"],temp_1[\"tl_q\"]))\n",
    "        prs_counts[i].append(temp_1[\"prs_q\"].value_counts().reindex(np.arange(n)))\n",
    "        tl_counts[i].append(temp_1[\"tl_q\"].value_counts().reindex(np.arange(n)))\n",
    "        or_table_all,pvalue_table_all=[[] for x in range(n)],[[] for x in range(n)]\n",
    "        or_table_prs,pvalue_table_prs,or_table_tl,pvalue_table_tl=[],[],[],[]\n",
    "        \n",
    "        temp_1=temp.copy(deep=True)\n",
    "        temp_1[gene]=0\n",
    "        temp_1.loc[(temp_1[\"single_gene_vaf\"]==\"NA\"),\"single_gene_vaf\"]=0\n",
    "        temp_1.loc[(temp_1[\"single_gene\"]==gene) & (temp_1[\"single_gene_vaf\"].astype(float)>=0.1),gene]=1\n",
    "        for y in range(n):\n",
    "            for z in range(n):\n",
    "                temp_1[\"prs_tl\"]=0\n",
    "                temp_1.loc[(temp_1[\"prs_q\"]==y) & (temp_1[\"tl_q\"]==z),\"prs_tl\"]=1\n",
    "                fe_or,fe_pv=fisher_exact(pd.crosstab(temp_1[gene],temp_1[\"prs_tl\"]))\n",
    "                or_table_all[y].append(fe_or)\n",
    "                pvalue_table_all[y].append(fe_pv)\n",
    "                \n",
    "        or_all[i].append(np.array([np.array(row,dtype=float) for row in or_table_all],dtype=float))\n",
    "        pv_all[i].append(np.array([np.array(row,dtype=float) for row in pvalue_table_all],dtype=float))\n",
    "                \n",
    "        for y in range(n):\n",
    "            temp_1[\"prs_group\"]=0\n",
    "            temp_1.loc[(temp_1[\"prs_q\"]==y),\"prs_group\"]=1\n",
    "            #print(pd.crosstab(temp_1[gene],temp_1[\"prs_group\"]))\n",
    "            fe_or,fe_pv=fisher_exact(pd.crosstab(temp_1[gene],temp_1[\"prs_group\"]))\n",
    "            or_table_prs.append(fe_or)\n",
    "            pvalue_table_prs.append(fe_pv)\n",
    "            \n",
    "            temp_1[\"tl_group\"]=0\n",
    "            temp_1.loc[(temp_1[\"tl_q\"]==y),\"tl_group\"]=1\n",
    "            fe_or,fe_pv=fisher_exact(pd.crosstab(temp_1[gene],temp_1[\"tl_group\"]))\n",
    "            or_table_tl.append(fe_or)\n",
    "            pvalue_table_tl.append(fe_pv)\n",
    "        \n",
    "        prs_or_all[i].append(np.array(or_table_prs))\n",
    "        prs_pv_all[i].append(np.array(pvalue_table_prs))\n",
    "        tl_or_all[i].append(np.array(or_table_tl))\n",
    "        tl_pv_all[i].append(np.array(pvalue_table_tl))   \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f43829d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "n=2\n",
    "significance=[[] for i in range(n)]\n",
    "significance_prs=[[] for i in range(n)]\n",
    "significance_tl=[[] for i in range(n)]\n",
    "for i in range(n):\n",
    "    for j in range(4):\n",
    "        print(i,j)\n",
    "        s=np.full((2,2),\"\",dtype=\"str\")\n",
    "        s[pv_all[i][j]<0.05]=\"*\"\n",
    "        significance[i].append(s)\n",
    "        \n",
    "        s=np.full(2,\"\",dtype=\"str\")\n",
    "        s[prs_pv_all[i][j]<0.05]=\"*\"\n",
    "        significance_prs[i].append(s)\n",
    "\n",
    "        s=np.full(2,\"\",dtype=\"str\")\n",
    "        s[tl_pv_all[i][j]<0.05]=\"*\"\n",
    "        significance_tl[i].append(s)\n",
    "        \n",
    "annot=[[] for i in range(n)]\n",
    "annot_prs=[[] for i in range(n)]\n",
    "annot_tl=[[] for i in range(n)]\n",
    "\n",
    "for i in range(n):\n",
    "    for j in range(4):\n",
    "        annot[i].append(contigency_table_all[i][j].astype(\"str\")+\"\\n\"+significance[i][j])\n",
    "        annot_prs[i].append(prs_counts[i][j].astype(\"str\")+\"\\n\"+significance_prs[i][j])\n",
    "        annot_tl[i].append(tl_counts[i][j].astype(\"str\")+\"\\n\"+significance_tl[i][j])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96ab6955",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "log_freq_all=[]\n",
    "for x in freq_all:\n",
    "    log_freq_all.append([np.log2(f) for f in x])\n",
    "\n",
    "log_prs_freq_all=[[np.log2(x) for x in xs] for xs in prs_freq_all]\n",
    "log_tl_freq_all=[[np.log2(x) for x in xs] for xs in tl_freq_all]\n",
    "\n",
    "max_or=[x.max().max() for xs in or_all for x in xs]+[x.max() for xs in prs_or_all for x in xs]+[x.max() for xs in tl_or_all for x in xs]\n",
    "min_or=[x.min().min() for xs in or_all for x in xs]+[x.min() for xs in prs_or_all for x in xs]+[x.min() for xs in tl_or_all for x in xs]\n",
    "\n",
    "fig = plt.figure(figsize=(12,7))\n",
    "gs0 = fig.add_gridspec(2,4, wspace=0.15, hspace=0.43)\n",
    "ax=[[[] for i in range(4)] for j in range(2)]\n",
    "for i in range(2):\n",
    "    for j in range(4):\n",
    "        spec=gs0[i,j].subgridspec(ncols=2, nrows=2,width_ratios=[1, 2], wspace=0.12, hspace=0.07, height_ratios=[ 2, 1])\n",
    "        ax[i][j].append(fig.add_subplot(spec[0]))\n",
    "        ax[i][j].append(fig.add_subplot(spec[1],sharey=ax[i][j][0]))\n",
    "        ax[i][j].append(fig.add_subplot(spec[3],sharex=ax[i][j][1]))\n",
    "        for k in [1,2]:\n",
    "            ax[i][j][k].tick_params(labelleft=False)\n",
    "        if j!=0:\n",
    "            ax[i][j][0].tick_params(labelleft=False)\n",
    "        for k in [0,1]:\n",
    "            ax[i][j][k].tick_params(labelbottom=False)\n",
    "        \n",
    "\n",
    "#divnorm = colors.TwoSlopeNorm(vmin=min(min_or), vcenter=1,vmax= max(max_or))\n",
    "divnorm = matplotlib.colors.TwoSlopeNorm(vmin=0.2, vcenter=1,vmax= 4.3)\n",
    "\n",
    "current_cmap = matplotlib.colormaps['bwr'].copy()\n",
    "cbar_ax = fig.add_axes([0.93, 0.15, 0.02, 0.7])\n",
    "\n",
    "for i in range(2):\n",
    "    for j,gene in enumerate([\"DNMT3A_other\",\"TET2\",\"SRSF2\",\"PPM1D\"]):\n",
    "        if i==1 and j==3:\n",
    "            cbar=sns.heatmap(or_all[i][j][::-1], ax=ax[i][j][1], annot=significance[i][j][::-1], norm=divnorm, cmap=current_cmap, square=True,linewidths=0.18,linecolor = \"gray\",clip_on=False, cbar_ax=cbar_ax, fmt=\"s\", annot_kws={'color':'k'},cbar_kws={'ticks':[0.25,0.5,0.75,1,2,3,4]})\n",
    "        else:\n",
    "            cbar=sns.heatmap(or_all[i][j][::-1], ax=ax[i][j][1], annot=significance[i][j][::-1], norm=divnorm, cmap=current_cmap, square=True,linewidths=0.18,linecolor = \"gray\",clip_on=False,cbar=False,fmt=\"s\",annot_kws={'color':'k'})\n",
    "        \n",
    "        cbar=sns.heatmap(np.asarray(prs_or_all[i][j][::-1]).reshape(2,1), ax=ax[i][j][0], annot=np.asarray(significance_prs[i][j][::-1]).reshape(2,1), norm=divnorm, cmap=current_cmap, square=True,linewidths=0.18,linecolor = \"gray\",clip_on=False,cbar=False,fmt=\"s\",annot_kws={'color':'k'})\n",
    "        cbar=sns.heatmap(np.asarray(tl_or_all[i][j]).reshape(1,2), ax=ax[i][j][2], annot=np.asarray(significance_tl[i][j]).reshape(1,2), norm=divnorm, cmap=current_cmap, square=True,linewidths=0.18,linecolor = \"gray\",clip_on=False,cbar=False,fmt=\"s\",annot_kws={'color':'k'})\n",
    "        \n",
    "        ax[i][j][1].set_ylabel(\"\")\n",
    "        ax[i][j][1].set_xlabel(\"\")\n",
    "        ax[i][j][2].set_xticks(np.arange(0.5,2.5))\n",
    "        ax[i][j][2].set_xticklabels(np.arange(1,3))\n",
    "        if j==0:\n",
    "            ax[i][j][0].set_yticks(np.arange(0.5,2.5))\n",
    "            ax[i][j][0].set_yticklabels(np.arange(1,3)[::-1])\n",
    "        if i==0:\n",
    "            ax[i][j][1].set_title(gene,pad=10)\n",
    "        ax[i][j][0].set_xticks([])\n",
    "        ax[i][j][2].set_yticks([])\n",
    "fig.text(0.55, 0.04, 'Measured TL quantiles', ha='center',fontsize=15)\n",
    "fig.text(0.08, 0.5, 'PRS quantiles', va='center', rotation='vertical',fontsize=15)\n",
    "cbar_ax.tick_params(axis='y', which='both',direction='in',colors=\"k\")\n",
    "cbar_ax.set_yticklabels(['0.25', '0.5','0.75','1','2','3', '4'],fontsize=8,color=\"k\")\n",
    "cbar_ax.set_title('OR')\n",
    "#plt.savefig(\"or_prs_tl_with_counts_2q_vaf_lt_0.1.pdf\",bbox_inches=\"tight\")\n",
    "#plt.savefig(\"or_prs_tl_with_counts_2q_vaf_lt_0.1.svg\",bbox_inches=\"tight\")\n",
    "\n",
    "\n",
    "fig.canvas.draw()\n",
    "transFigure = fig.transFigure.inverted()\n",
    "\n",
    "for i in [0,1]:\n",
    "    coord1 = transFigure.transform(ax[0][i][1].transData.transform([2,0.5]))\n",
    "    coord2 = transFigure.transform(ax[0][i][1].transData.transform([2.15,0.5]))\n",
    "    coord3 = transFigure.transform(ax[1][i][1].transData.transform([2.15,0.5]))\n",
    "    coord4 = transFigure.transform(ax[1][i][1].transData.transform([2,0.5]))\n",
    "\n",
    "    line1 = matplotlib.lines.Line2D((coord1[0],coord2[0]),(coord1[1],coord2[1]), transform=fig.transFigure,color=\"k\",linestyle=\"--\",linewidth=0.7)\n",
    "    line2 = matplotlib.lines.Line2D((coord2[0],coord3[0]),(coord2[1],coord3[1]), transform=fig.transFigure,color=\"k\",linestyle=\"--\",linewidth=0.7)\n",
    "    line3 = matplotlib.lines.Line2D((coord3[0],coord4[0]),(coord3[1],coord4[1]), transform=fig.transFigure,color=\"k\",linestyle=\"--\",linewidth=0.5)\n",
    "\n",
    "    fig.lines=fig.lines+[line1,line2,line3]\n",
    "    \n",
    "for i in [2,3]:\n",
    "    coord1 = transFigure.transform(ax[0][i][1].transData.transform([2,1.5]))\n",
    "    coord2 = transFigure.transform(ax[0][i][1].transData.transform([2.15,1.5]))\n",
    "    coord3 = transFigure.transform(ax[1][i][1].transData.transform([2.15,1.5]))\n",
    "    coord4 = transFigure.transform(ax[1][i][1].transData.transform([2,1.5]))\n",
    "\n",
    "    line1 = matplotlib.lines.Line2D((coord1[0],coord2[0]),(coord1[1],coord2[1]), transform=fig.transFigure,color=\"k\",linestyle=\"--\",linewidth=0.7)\n",
    "    line2 = matplotlib.lines.Line2D((coord2[0],coord3[0]),(coord2[1],coord3[1]), transform=fig.transFigure,color=\"k\",linestyle=\"--\",linewidth=0.7)\n",
    "    line3 = matplotlib.lines.Line2D((coord3[0],coord4[0]),(coord3[1],coord4[1]), transform=fig.transFigure,color=\"k\",linestyle=\"--\",linewidth=0.5)\n",
    "\n",
    "    fig.lines=fig.lines+[line1,line2,line3]\n",
    "#plt.savefig(\"../figures_revision1/tl_prs_qq_2_singlegene_noHM.svg\")\n",
    "#plt.savefig(\"../figures_revision1/tl_prs_qq_2_singlegene_noHM.pdf\")   \n",
    "#plt.savefig(\"../figures_revision1/tl_prs_qq_2_singlegene_noHM.png\")   \n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f47cbe2c",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "########## Plot MR results############\n",
    "\n",
    "\n",
    "mr_res=pd.read_csv(\"../polygenic_risk_score/farm/regression_coef/revision1_singlegene/tl_as_exp_ch_nowbc_flip_final_r1_singlegene.txt\",usecols=[\"CH_type\",\"method\",\"or\",\"or_lci95\",\"or_uci95\",\"pval\"],header=\"infer\",sep=\"\\t\")\n",
    "\n",
    "display(mr_res)\n",
    "mr_res=mr_res[(mr_res[\"method\"]==\"Inverse variance weighted\") & (mr_res[\"CH_type\"].isin([\"dnmt3a\",\"tet2\",\"asxl1\",\"ppm1d\",\"splicing\",\"mpn\",\"long_telomere\",\"short_telomere\",\"all\",\"Y_loss\",\"X_loss\",\"autosomal\"]))]\n",
    "display(mr_res)\n",
    "\n",
    "mr_res=mr_res.sort_values('CH_type', key=lambda s: s.apply([\"dnmt3a\",\"tet2\",\"asxl1\",\"ppm1d\",\"splicing\",\"mpn\",\"long_telomere\",\"short_telomere\",\"all\",\"Y_loss\",\"X_loss\",\"autosomal\"].index), ignore_index=True)\n",
    "print(mr_res)\n",
    "fig, axes = plt.subplots(figsize=(8,6.5))\n",
    "axes.axvline(x=1,linestyle='--',linewidth=0.5)\n",
    "data=mr_res.copy(deep=True).reset_index()\n",
    "\n",
    "axes.errorbar(data.loc[data[\"pval\"]<0.05,\"or\"],data[data[\"pval\"]<0.05].index, xerr=[(data.loc[data[\"pval\"]<0.05,\"or\"]-data.loc[data[\"pval\"]<0.05,\"or_lci95\"]),(data.loc[data[\"pval\"]<0.05,\"or_uci95\"]-data.loc[data[\"pval\"]<0.05,\"or\"])], fmt='o',label=\"P value < 0.05\",markersize=9,elinewidth=1,c=\"#483737\")\n",
    "display(data)\n",
    "axes.errorbar(data.loc[(data[\"pval\"]>=0.05),\"or\"],data[(data[\"pval\"]>=0.05)].index, xerr=[(data.loc[(data[\"pval\"]>=0.05),\"or\"]-data.loc[(data[\"pval\"]>=0.05),\"or_lci95\"]),(data.loc[(data[\"pval\"]>=0.05),\"or_uci95\"]-data.loc[(data[\"pval\"]>=0.05),\"or\"])], fmt='o',label=\"P value \"+r'$\\mathrm{\\geq} $'+\" 0.05\",markersize=9,elinewidth=1,c=\"#ac9393\")\n",
    "#axes.set_xlim([0,5])\n",
    "axes.set_xlabel(\"OR\",fontsize=15)\n",
    "axes.set_ylim([-0.8,11.8])\n",
    "axes.set_yticks(np.arange(12))\n",
    "axes.set_yticklabels([\"DNMT3A\",\"TET2\",\"ASXL1\",\"PPM1D\",\"Splicing\",\"JAK2,CALR\",\"Long telomere\",\"Short telomere\",\"All CH\",\"LOY\",\"LOX\",\"autosomal\"])\n",
    "#axes.ticklabelparams(fontsize=16)\n",
    "#axes.set_xticks([0.2,0.3,0.6,1,2,3,4,6])\n",
    "#axes.set_xticklabels([0.2,0.3,0.6,1,2,3,4,6])\n",
    "axes.legend(fontsize=12)\n",
    "\n",
    "\n",
    "axes.set_xscale('log')\n",
    "axes.tick_params(axis=\"both\",which='major',labelsize=12)\n",
    "axes.tick_params(axis=\"both\",which='minor',labelsize=12)\n",
    "\n",
    "#axes.get_xaxis().get_major_formatter().labelOnlyBase = False\n",
    "axes.get_xaxis().set_major_formatter(matplotlib.ticker.ScalarFormatter())\n",
    "axes.get_xaxis().set_minor_formatter(matplotlib.ticker.ScalarFormatter())\n",
    "axes.spines[['right', 'top']].set_visible(False)\n",
    "\n",
    "#        k+=1\n",
    "#plt.savefig(\"../figures_revision1/MR_flip_final_singlegene.png\",bbox_inches=\"tight\",dpi=600)\n",
    "#plt.savefig(\"../figures_revision1/MR_flip_final_singlegene.pdf\",bbox_inches=\"tight\")\n",
    "#plt.savefig(\"../figures_revision1/MR_flip_final_singlegene.svg\",bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ff789fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "data=pd.concat([data,pd.DataFrame(multipletests(data[\"pval\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1])],axis=1)\n",
    "data=data.rename(columns={0:\"FDR_all\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6a319a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "data[\"CH_type\"]=[\"DNMT3A\",\"TET2\",\"ASXL1\",\"PPM1D\",\"Splicing\",\"JAK2,CALR\",\"Long telomere\",\"Short telomere\",\"All CH\",\"LOY\",\"LOX\",\"autosomal\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a55dfe55",
   "metadata": {},
   "outputs": [],
   "source": [
    "#data[[x for x in data.columns if (x!=\"index\" and x!=\"method\") ]].to_csv(\"mr_res_r1_singlegene.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df2c2ba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################Supplementary figures##################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02399f6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "######## Association with paternal age  ##################################\n",
    "\n",
    "######### Load data ###################\n",
    "parents_age=pd.read_csv(\"father_mother_age.tsv\",sep=\"\\t\",dtype={'f.eid':\"str\"})\n",
    "parents_age.columns=[\"eid\",\"maternal_age\",\"paternal_age\"]\n",
    "parents_age[\"maternal_age\"]=parents_age[\"maternal_age\"].map(lambda row:np.nan if row<0 else row)\n",
    "parents_age[\"paternal_age\"]=parents_age[\"paternal_age\"].map(lambda row:np.nan if row<0 else row)\n",
    "ukb500k_ts_som=ukb500k_ts_som.merge(parents_age,how=\"left\",on=\"eid\")\n",
    "ukb500k_ts_som[\"paternal_age_at_birth\"]=ukb500k_ts_som[\"paternal_age\"]-ukb500k_ts_som[\"age\"]\n",
    "ukb500k_ts_som[\"maternal_age_at_birth\"]=ukb500k_ts_som[\"maternal_age\"]-ukb500k_ts_som[\"age\"]\n",
    "ukb500k_ts_som[\"paternal_age_at_birth\"]=ukb500k_ts_som[\"paternal_age_at_birth\"].map(lambda row:np.nan if row<5 else row)\n",
    "ukb500k_ts_som[\"maternal_age_at_birth\"]=ukb500k_ts_som[\"maternal_age_at_birth\"].map(lambda row:np.nan if row<5 else row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0777dde0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#################### Define Long/short groups #################\n",
    "ukb500k_ts_som[\"telomere_group\"]=\"Control\"\n",
    "ukb500k_ts_som.loc[(ukb500k_ts_som[\"single_gene\"].isin([\"DNMT3A_other\",\"DNMT3A_R882\",\"JAK2\",\"GNAS\",\"GNB1\",\"TET2\",\"CBL\"])),\"telomere_group\"]=\"Long\"\n",
    "ukb500k_ts_som.loc[(ukb500k_ts_som[\"single_gene\"].isin([\"PPM1D\",\"SF3B1\",\"SRSF2\"])),\"telomere_group\"]=\"Short\"\n",
    "ukb500k_ts_som.loc[((ukb500k_ts_som[\"single_gene\"]!=\"Control\")  & (ukb500k_ts_som[\"telomere_group\"]==\"Control\")),\"telomere_group\"]=\"Others\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb41e539",
   "metadata": {},
   "outputs": [],
   "source": [
    "############### Perform association analysis ##############\n",
    "temp=ukb500k_ts_som.copy(deep=True)\n",
    "\n",
    "formula = 'telomere_group ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10','paternal_age_at_birth'])\n",
    "gene_list=list(temp[\"telomere_group\"].unique())\n",
    "gene_list.remove(\"Control\")\n",
    "\n",
    "or_table_paternal_age=[]\n",
    "\n",
    "for gene in gene_list:\n",
    "    data=temp[[\"paternal_age_at_birth\", \"sex\", \"age\", \"smoking_status\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\",\"telomere_group\"]].copy(deep=True)\n",
    "    data=data.dropna()\n",
    "    data=data[(data[\"telomere_group\"].isin([gene,\"Control\"]))]\n",
    "    data[\"telomere_group\"]=data[\"telomere_group\"].map({gene:1,\"Control\":0})\n",
    "    print(data[\"telomere_group\"].value_counts())\n",
    "    lin_model=smf.logit(formula=formula,data=data)\n",
    "    result=lin_model.fit(max_iterations= 100000)\n",
    "    display(result.summary2())\n",
    "    params = np.exp(result.params)\n",
    "    conf = np.exp(result.conf_int())\n",
    "    print(result.pvalues)\n",
    "    print([gene,params[\"paternal_age_at_birth\"],conf.loc[\"paternal_age_at_birth\",0],conf.loc[\"paternal_age_at_birth\",1],result.pvalues[\"paternal_age_at_birth\"]])\n",
    "    or_table_paternal_age.append([gene,params[\"paternal_age_at_birth\"],conf.loc[\"paternal_age_at_birth\",0],conf.loc[\"paternal_age_at_birth\",1],result.pvalues[\"paternal_age_at_birth\"]])\n",
    "or_table_paternal_age=pd.DataFrame(or_table_paternal_age)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dffc81e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "or_table_paternal_age"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccdc6bd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "########### Generate figure ######################3\n",
    "plt.figure(figsize=(8,3))\n",
    "or_table_paternal_age=or_table_paternal_age[or_table_paternal_age[0]!=\"Others\"]\n",
    "or_table_paternal_age[0]=[\"Long\\n(2204)\", \"Short\\n(98)\"]\n",
    "\n",
    "eb=plt.errorbar(or_table_paternal_age.loc[0,1],[0],xerr=[[or_table_paternal_age.loc[0,1]-or_table_paternal_age.loc[0,2]],[or_table_paternal_age.loc[0,3]-or_table_paternal_age.loc[0,1]]],fmt=\"o\",label=\"P\"+r'$\\mathrm{<}$0.05', markersize=10,elinewidth=1,c=\"#73264d\")\n",
    "eb=plt.errorbar(or_table_paternal_age.loc[2,1],[1],xerr=[[or_table_paternal_age.loc[2,1]-or_table_paternal_age.loc[2,2]],[or_table_paternal_age.loc[2,3]-or_table_paternal_age.loc[2,1]]],fmt=\"o\",label=\"P\"+r'$\\mathrm{\\geq}0.05$', markersize=10,elinewidth=1,c=\"#df9fbf\")\n",
    "\n",
    "\n",
    "plt.yticks(np.arange(2),or_table_paternal_age[0])\n",
    "plt.axvline(x=1,c=\"k\",linestyle=\"--\",linewidth=0.8)\n",
    "plt.xlabel(\"OR\",fontsize=16)\n",
    "plt.xticks(fontsize=13)\n",
    "plt.yticks(fontsize=13)\n",
    "plt.ylim([-0.5,1.5])\n",
    "plt.legend()\n",
    "\n",
    "#plt.savefig(\"../figures_new/fathers_age_ass_ch_or.png\",bbox_inches=\"tight\",dpi=600)\n",
    "#plt.savefig(\"../figures_revision1/fathers_age_assn_telomeregroup_or_singlegene.png\",bbox_inches=\"tight\",dpi=600)\n",
    "#plt.savefig(\"../figures_revision1/fathers_age_assn_telomeregroup_or_singlegene.svg\",bbox_inches=\"tight\")\n",
    "#plt.savefig(\"../figures_revision1/fathers_age_assn_telomeregroup_or_singlegene.pdf\",bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47afd6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "or_table_paternal_age\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b58e77ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "or_table_paternal_age.columns=[\"CH type\",\"OR\",\"OR lower CI (2.5%)\",\"OR lower CI (97.5%)\",\"P-value\"]\n",
    "or_table_paternal_age[\"CH type\"]=[\"Long CH\",\"Short CH\"]\n",
    "#or_table_paternal_age[\"FDR\"]=multipletests(or_table_paternal_age[\"P-value\"],alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n",
    "or_table_paternal_age.to_csv(\"SupplementaryTable_extfig2.ts\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa014d6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "or_table_paternal_age"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36d79b29",
   "metadata": {},
   "outputs": [],
   "source": [
    "######## Add time to diagnosis data for haematological malignancies ###################\n",
    "tt_diagnosis=pd.read_csv(\"/home/sruthi/Downloads/tt_diagnosis.tsv\",delimiter=\"\\t\",dtype={\"eid\":\"str\"})\n",
    "tt_diagnosis.columns=[x+\"_tt\" for x in tt_diagnosis.columns]\n",
    "tt_diagnosis=tt_diagnosis.rename(columns={\"eid_tt\":\"eid\"})\n",
    "tt_diagnosis.replace(0,np.nan,inplace=True)\n",
    "tt_diagnosis.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "087d1bc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "############# Combine with main data set #####################\n",
    "ukb500k_ts_som=ukb500k_ts_som.merge(tt_diagnosis,on=\"eid\",how=\"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32238157",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ukb500k_ts_som.drop(columns=['MPN', 'MDS', 'CMML', 'AML'],inplace=True)\n",
    "ukb500k_ts_som[\"myeloid_malignancy\"]=ukb500k_ts_som[[\"AML_tt\",\"MDS_tt\",\"MPN_tt\",\"CMML_tt\"]].apply(lambda row: \"False\" if row.isna().all() else row.idxmin(),axis=1)\n",
    "ukb500k_ts_som[\"Time_to_malignancy\"]=ukb500k_ts_som[['AML_tt','MDS_tt',\"MPN_tt\",\"CMML_tt\"]].min(axis=1)\n",
    "display(ukb500k_ts_som[\"myeloid_malignancy\"].value_counts())\n",
    "ukb500k_ts_som[\"myeloid_malignancy\"]=ukb500k_ts_som[\"myeloid_malignancy\"].map({\"AML_tt\":\"AML\",\"MDS_tt\":\"MDS\",\"MPN_tt\":\"MPN\",\"CMML_tt\":\"CMML\",\"False\":np.nan},na_action='ignore')\n",
    "ukb500k_ts_som[\"Time_to_chemo\"]=(ukb500k_ts_som[\"chemo_date\"]-ukb500k_ts_som[\"daac\"]).dt.days\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05d92b63",
   "metadata": {},
   "outputs": [],
   "source": [
    "######## perform association analysis between myeloid malignancies and prs/ltl.\n",
    "def run_reg(temp_data,MM_list_i,tl,formula):\n",
    "    d=temp_data.copy(deep=True)\n",
    "    d=d[(d[\"myeloid_malignancy\"].isin(MM_list_i+[\"Control\"]))]\n",
    "    print(d[\"myeloid_malignancy\"].value_counts(),MM_list_i)\n",
    "    d[\"mm\"]=d[\"myeloid_malignancy\"].map(lambda row:1 if row in MM_list_i else 0)\n",
    "    d=d.astype({\"mm\": int})\n",
    "    print(formula)\n",
    "    log_model=smf.glm(formula=formula,data=d,family=sm.families.Binomial())\n",
    "    result=log_model.fit(max_iterations= 100000)\n",
    "    display(result.summary2())\n",
    "    params = result.params\n",
    "    conf = result.conf_int()\n",
    "    return([np.exp(params[tl]),result.pvalues[tl],np.exp(conf.loc[tl,0]),np.exp(conf.loc[tl,1])])   \n",
    "    \n",
    "\n",
    "MM_list=[[\"MDS\"],[\"AML\"], [\"MPN\"],[\"MDS\",\"MPN\",\"CMML\",\"AML\"]]\n",
    "names_MM=[\"MDS\",\"AML\",\"MPN\",\"myeloid_malignancy\"]\n",
    "\n",
    "data_prs=ukb500k_ts_som.loc[(ukb500k_ts_som[\"Time_to_chemo\"]>ukb500k_ts_som[\"Time_to_malignancy\"]) |  (ukb500k_ts_som[\"Time_to_chemo\"].isna()) |  (ukb500k_ts_som[\"Time_to_malignancy\"].isna()),[\"total_PRS\",\"sex\", \"age\", \"smoking_status\",\"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\",\"myeloid_malignancy\"]]\n",
    "\n",
    "data_prs.loc[(data_prs[\"myeloid_malignancy\"].isna()),\"myeloid_malignancy\"]=\"Control\"\n",
    "data_prs=data_prs.dropna()\n",
    "print(data_prs[\"myeloid_malignancy\"].value_counts())\n",
    "\n",
    "data_tl=ukb500k_ts_som.loc[((ukb500k_ts_som[\"Time_to_chemo\"]>ukb500k_ts_som[\"Time_to_malignancy\"]) |  (ukb500k_ts_som[\"Time_to_chemo\"].isna())  |  (ukb500k_ts_som[\"Time_to_malignancy\"].isna())) & ((ukb500k_ts_som[\"HM_tt\"]>0) |(ukb500k_ts_som[\"HM_tt\"].isna()) ) & ((ukb500k_ts_som[\"Time_to_malignancy\"].isna())|(ukb500k_ts_som[\"Time_to_malignancy\"]>0)),[\"sex\", \"age\", \"smoking_status\",\"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\",\"myeloid_malignancy\",\"Z-adjusted T/S log\"]]\n",
    "data_tl=data_tl.rename(columns={\"Z-adjusted T/S log\":\"tl\"})\n",
    "\n",
    "\n",
    "data_tl.loc[(data_tl[\"myeloid_malignancy\"].isna()),\"myeloid_malignancy\"]=\"Control\"\n",
    "data_tl=data_tl.dropna()\n",
    "print(data_tl[\"myeloid_malignancy\"].value_counts())\n",
    "\n",
    "#data=data.rename(columns={\"Z-adjusted T/S log\":\"tl\"})\n",
    "formula_base = 'mm ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10'])\n",
    "\n",
    "reg_coef_mm_all=[]\n",
    "for i in range(4):\n",
    "    \n",
    "        reg_coef_mm_all.append(run_reg(data_prs,MM_list[i],\"total_PRS\",formula_base+\"+total_PRS\")+run_reg(data_tl,MM_list[i],\"tl\",formula_base+\"+tl\"))\n",
    "\n",
    "\n",
    "reg_coef_mm_all=pd.DataFrame(reg_coef_mm_all)\n",
    "reg_coef_mm_all.columns=[\"prs_or\",\"prs_pvalue\",\"prs_or_low_ci\",\"prs_or_high_ci\",\"tl_or\",\"tl_pvalue\",\"tl_or_low_ci\",\"tl_or_high_ci\"]\n",
    "reg_coef_mm_all[\"MM\"]=names_MM\n",
    "\n",
    "display(reg_coef_mm_all)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "\n",
    "reg_coef_mm_all_sel=reg_coef_mm_all.loc[(reg_coef_mm_all[\"prs_pvalue\"]<0.05),:]\n",
    "plt.errorbar(reg_coef_mm_all_sel.loc[:,\"prs_or\"],reg_coef_mm_all_sel.index,xerr=[reg_coef_mm_all_sel.loc[:,\"prs_or\"]-reg_coef_mm_all_sel.loc[:,\"prs_or_low_ci\"],reg_coef_mm_all_sel.loc[:,\"prs_or_high_ci\"]-reg_coef_mm_all_sel.loc[:,\"prs_or\"]],fmt=\"o\",label=\"LTL-PRS (P\"+r'$\\mathrm{<}0.05$'+\")\",c=\"#7a7a52\", markersize=9)\n",
    "\n",
    "\n",
    "reg_coef_mm_all_sel=reg_coef_mm_all.loc[(reg_coef_mm_all[\"prs_pvalue\"]>=0.05),:]\n",
    "eb=plt.errorbar(reg_coef_mm_all_sel.loc[:,\"prs_or\"],reg_coef_mm_all_sel.index,xerr=[reg_coef_mm_all_sel.loc[:,\"prs_or\"]-reg_coef_mm_all_sel.loc[:,\"prs_or_low_ci\"],reg_coef_mm_all_sel.loc[:,\"prs_or_high_ci\"]-reg_coef_mm_all_sel.loc[:,\"prs_or\"]],fmt=\"o\",label=\"LTL-PRS (P\"+r'$\\mathrm{\\geq}0.05$'+\")\",c=\"#b8b894\",markersize=9)\n",
    "#eb[-1][0].set_linestyle('--')\n",
    "\n",
    "\n",
    "reg_coef_mm_all_sel=reg_coef_mm_all.loc[(reg_coef_mm_all[\"tl_pvalue\"]<0.05),:]\n",
    "plt.errorbar(reg_coef_mm_all_sel.loc[:,\"tl_or\"],reg_coef_mm_all_sel.index+0.2,xerr=[reg_coef_mm_all_sel.loc[:,\"tl_or\"]-reg_coef_mm_all_sel.loc[:,\"tl_or_low_ci\"],reg_coef_mm_all_sel.loc[:,\"tl_or_high_ci\"]-reg_coef_mm_all_sel.loc[:,\"tl_or\"]],fmt=\"o\",label=\"LTL (P\"+r'$\\mathrm{<}0.05$'+\")\", c=\"#cc6699\",markersize=9)\n",
    "\n",
    "#reg_coef_mm_all_sel=reg_coef_mm_all.loc[(reg_coef_mm_all[\"tl_pvalue\"]>=0.05),:]\n",
    "#eb=plt.errorbar(reg_coef_mm_all_sel.loc[:,\"tl_or\"],reg_coef_mm_all_sel.index+0.2,xerr=[reg_coef_mm_all_sel.loc[:,\"tl_or\"]-reg_coef_mm_all_sel.loc[:,\"tl_or_low_ci\"],reg_coef_mm_all_sel.loc[:,\"tl_or_high_ci\"]-reg_coef_mm_all_sel.loc[:,\"tl_or\"]],fmt=\"o\",label=\"LTL-PRS (P\"+r'$\\mathrm{\\geq}0.05$'+\")\",c=\"#0099cc\",markersize=9)\n",
    "#eb[-1][0].set_linestyle('--')\n",
    "\n",
    "plt.yticks(np.arange(4)+0.1,names_MM,fontsize=13)\n",
    "plt.xticks(fontsize=13)\n",
    "plt.axvline(x=1,linestyle=\"--\",linewidth=1)\n",
    "plt.legend()\n",
    "\n",
    "plt.xlabel(\"OR\",fontsize=18)\n",
    "#plt.savefig(\"../figures_final/MM_ch_all_tl_prs_OR_chemo.svg\",dpi=600)\n",
    "#plt.savefig(\"../figures_revision1/MM_ch_all_tl_prs_OR_chemo_noHM.svg\",dpi=600)\n",
    "#plt.savefig(\"../figures_revision1/MM_ch_all_tl_prs_OR_chemo_noHM.pdf\",dpi=600)\n",
    "#plt.savefig(\"../figures_revision1/MM_ch_all_tl_prs_OR_chemo_noHM.svg\",dpi=600)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50b12f89",
   "metadata": {},
   "outputs": [],
   "source": [
    "prs_table=reg_coef_mm_all[[\"MM\",\"prs_or\",\"prs_pvalue\",\"prs_or_low_ci\",\"prs_or_high_ci\"]].copy(deep=True)\n",
    "prs_table[\"PRS/LTL\"]=\"PRS\"\n",
    "tl_table=reg_coef_mm_all[[\"MM\",\"tl_or\",\"tl_pvalue\",\"tl_or_low_ci\",\"tl_or_high_ci\"]].copy(deep=True)\n",
    "tl_table.columns=[\"MM\",\"prs_or\",\"prs_pvalue\",\"prs_or_low_ci\",\"prs_or_high_ci\"]\n",
    "tl_table[\"PRS/LTL\"]=\"LTL\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9329a367",
   "metadata": {},
   "outputs": [],
   "source": [
    "supptable_ext_fig1=pd.concat([prs_table,tl_table])\n",
    "supptable_ext_fig1.columns=[\"MM\",\"OR\",\"P-value\",\"OR lower CI (2.5%)\",\"OR upper CI (97.5%)\",\"PRS/LTL\"]\n",
    "supptable_ext_fig1[\"FDR\"]=multipletests(supptable_ext_fig1[\"P-value\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n",
    "supptable_ext_fig1[[\"PRS/LTL\",\"MM\",\"OR\",\"OR lower CI (2.5%)\",\"OR upper CI (97.5%)\",\"P-value\",\"FDR\"]].to_csv(\"SupplementaryTable_extfig1.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0838bda2",
   "metadata": {},
   "outputs": [],
   "source": [
    "supptable_ext_fig1[[\"PRS/LTL\",\"MM\",\"OR\",\"OR lower CI (2.5%)\",\"OR upper CI (97.5%)\",\"P-value\",\"FDR\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0f772cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "############################ Validation ##########################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47894136",
   "metadata": {},
   "outputs": [],
   "source": [
    "################# Topmed-derived PRS #################\n",
    "temp=ukb500k_ts_som_mca[[\"topmed_all_trans_total_PRS\",\"sex\", \"age\", \"smoking_status\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\",\"CHR_COPY_CHANGE\"]].copy(deep=True)\n",
    "print(temp.shape)\n",
    "temp=temp[~(temp[\"CHR_COPY_CHANGE\"].isin([\"Y:neutral\",\"X:neutral\",\"Y:gain\",\"X:gain\"]))]\n",
    "temp=temp.dropna()\n",
    "temp[\"MCA_type\"]=temp[\"CHR_COPY_CHANGE\"].map(lambda row:\"autosomal MCA\" if row not in [\"Y:loss\",\"X:loss\",\"Control\"] else row)\n",
    "y=\"topmed_all_trans_total_PRS\"\n",
    "\n",
    "counts=temp[\"MCA_type\"].value_counts()\n",
    "print(counts.sum())\n",
    "gene_order=[\"autosomal MCA\",\"Y:loss\",\"X:loss\"]\n",
    "plt.figure(figsize=(6,4))\n",
    "b=sns.violinplot(data=temp,x=\"MCA_type\",y=y,order=[\"Control\"]+gene_order,linewidth=2,palette={\"Control\":gene_color[\"Control\"],\"autosomal MCA\":\"#999999\",\"Y:loss\":\"#E69F00\",\"X:loss\":\"#56B4E9\"})\n",
    "ax=plt.gca()\n",
    "xticks=ax.get_xticklabels()\n",
    "xticks_new=[]\n",
    "for i in xticks:\n",
    "    print(i.get_text(),counts[i.get_text()])\n",
    "    xticks_new.append(i.get_text()+\"\\n(\"+str(counts[i.get_text()])+\")\")\n",
    "#plt.ylim([-5,5])\n",
    "plt.xticks(np.arange(0,len(xticks),1),xticks_new)\n",
    "plt.axhline(y=temp.loc[temp[\"MCA_type\"]==\"Control\",y].median(), color='k', linestyle='--')\n",
    "plt.grid(axis='x')\n",
    "plt.xticks(fontsize=12)\n",
    "plt.yticks(fontsize=12)\n",
    "ax.xaxis.label.set_size(18)\n",
    "ax.yaxis.label.set_size(16)\n",
    "plt.ylabel(\"PRS\")\n",
    "plt.xlabel(\"\")\n",
    "plt.show()\n",
    "\n",
    "plot_values=[]\n",
    "for mca_type in [\"autosomal MCA\",\"Y:loss\",\"X:loss\"]:\n",
    "    if mca_type==\"autosomal MCA\":\n",
    "        data=temp[temp[\"MCA_type\"].isin([\"Control\",mca_type])].copy(deep=True)\n",
    "        print(data.shape)\n",
    "        formula = 'MCA_type ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"topmed_all_trans_total_PRS\"])\n",
    "    elif mca_type==\"Y:loss\":\n",
    "        data=temp[(temp[\"MCA_type\"].isin([\"Control\",mca_type])) & (temp[\"sex\"]==\"Male\")].copy(deep=True)\n",
    "        formula = 'MCA_type ~' + ' + '.join(['age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"topmed_all_trans_total_PRS\"])\n",
    "    else:\n",
    "        data=temp[(temp[\"MCA_type\"].isin([\"Control\",mca_type])) & (temp[\"sex\"]==\"Female\")].copy(deep=True)\n",
    "        formula = 'MCA_type ~' + ' + '.join(['age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"topmed_all_trans_total_PRS\"])\n",
    "    data[\"MCA_type\"]=data[\"MCA_type\"].map({mca_type:1,\"Control\":0})\n",
    "    data[\"MCA_type\"]=data[\"MCA_type\"].astype(int)\n",
    "    \n",
    "    lin_model=smf.logit(formula=formula,data=data)\n",
    "    result=lin_model.fit(max_iterations= 100000)\n",
    "    display(result.summary2())\n",
    "    params = np.exp(result.params)\n",
    "    conf = np.exp(result.conf_int())\n",
    "    print(result.pvalues)\n",
    "    plot_values.append([params[\"topmed_all_trans_total_PRS\"],conf.loc[\"topmed_all_trans_total_PRS\",0],conf.loc[\"topmed_all_trans_total_PRS\",1],result.pvalues[\"topmed_all_trans_total_PRS\"]])\n",
    "\n",
    "plot_values=np.array(plot_values)\n",
    "data_mca=temp.copy(deep=True)\n",
    "data_mca=data_mca.rename(columns={\"Z-adjusted T/S log\":\"tl\"})\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "212062ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "supptable_topmed_mca=pd.DataFrame(plot_values)\n",
    "supptable_topmed_mca[\"Gene/mCA\"]=[\"autosomal mCA\",\"LOY\",\"LOX\"]\n",
    "supptable_topmed_mca.columns=[\"OR\",\"OR lower CI (2.5%)\",\"OR upper CI (97.5%)\",\"P-value\",\"Gene/mCA\"]\n",
    "supptable_topmed_mca=supptable_topmed_mca[[\"Gene/mCA\",\"OR\",\"OR lower CI (2.5%)\",\"OR upper CI (97.5%)\",\"P-value\"]]\n",
    "supptable_topmed_mca[\"FDR\"]=multipletests(supptable_topmed_mca[\"P-value\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc75ab8a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "############# Association with all CH and PRS ###############\n",
    "y=\"topmed_all_trans_total_PRS\"\n",
    "temp=ukb500k_ts_som.copy(deep=True)\n",
    "temp=temp.loc[(temp[y].notna())]\n",
    "\n",
    "formula = 'single_gene ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10','topmed_all_trans_total_PRS'])\n",
    "\n",
    "or_table_pr=[]\n",
    "\n",
    "data=temp[[\"topmed_all_trans_total_PRS\", \"sex\", \"age\", \"smoking_status\", \"PC1\", \"PC2\", \"PC3\", \"PC4\", \"PC5\", \"PC6\", \"PC7\", \"PC8\", \"PC9\", \"PC10\", \"single_gene\"]].copy(deep=True)\n",
    "data=data.dropna()\n",
    "print(data.shape)\n",
    "data[\"single_gene\"]=data[\"single_gene\"].map(lambda row:0 if row==\"Control\" else 1)\n",
    "data[\"single_gene\"]=data[\"single_gene\"].astype(int)\n",
    "print(data[\"single_gene\"].value_counts())\n",
    "log_model_allch=smf.logit(formula=formula,data=data)\n",
    "result_allch=log_model_allch.fit(max_iterations= 100000)\n",
    "display(result_allch.summary2())\n",
    "params_allch = np.exp(result_allch.params)\n",
    "conf_allch = np.exp(result_allch.conf_int())\n",
    "print(result_allch.pvalues)\n",
    "print([\"all CH\", params_allch[\"topmed_all_trans_total_PRS\"], conf_allch.loc[\"topmed_all_trans_total_PRS\",0], conf_allch.loc[\"topmed_all_trans_total_PRS\",1], result_allch.pvalues[\"topmed_all_trans_total_PRS\"]])\n",
    "or_table_pr.append([\"all CH\", params_allch[\"topmed_all_trans_total_PRS\"], conf_allch.loc[\"topmed_all_trans_total_PRS\",0], conf_allch.loc[\"topmed_all_trans_total_PRS\",1], result_allch.pvalues[\"topmed_all_trans_total_PRS\"]])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e3ad007",
   "metadata": {},
   "outputs": [],
   "source": [
    "y=\"topmed_all_trans_total_PRS\"\n",
    "temp=ukb500k_ts_som.copy(deep=True)\n",
    "temp=temp.loc[(temp[y].notna())]\n",
    "\n",
    "formula = 'single_gene ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10','topmed_all_trans_total_PRS'])\n",
    "gene_list=list(temp[\"single_gene\"].unique())\n",
    "gene_list.remove(\"Control\")\n",
    "\n",
    "for gene in gene_list:\n",
    "    data=temp[[\"topmed_all_trans_total_PRS\", \"sex\", \"age\", \"smoking_status\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\",\"single_gene\"]].copy(deep=True)\n",
    "    data=data.dropna()\n",
    "    print(data.shape)\n",
    "    data=data[(data[\"single_gene\"].isin([gene,\"Control\"]))]\n",
    "    data[\"single_gene\"]=data[\"single_gene\"].map({gene:1,\"Control\":0})\n",
    "    print(data[\"single_gene\"].value_counts())\n",
    "    log_model=smf.logit(formula=formula,data=data)\n",
    "    result=log_model.fit(max_iterations= 100000)\n",
    "    display(result.summary2())\n",
    "    params = np.exp(result.params)\n",
    "    conf = np.exp(result.conf_int())\n",
    "    print(result.pvalues)\n",
    "    print([gene,params[\"topmed_all_trans_total_PRS\"],conf.loc[\"topmed_all_trans_total_PRS\",0],conf.loc[\"topmed_all_trans_total_PRS\",1],result.pvalues[\"topmed_all_trans_total_PRS\"]])\n",
    "    or_table_pr.append([gene,params[\"topmed_all_trans_total_PRS\"],conf.loc[\"topmed_all_trans_total_PRS\",0],conf.loc[\"topmed_all_trans_total_PRS\",1],result.pvalues[\"topmed_all_trans_total_PRS\"]])\n",
    "or_table_pr=pd.DataFrame(or_table_pr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b4302b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "or_table_pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21fafbf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.errorbar(np.arange(15),or_table_pr.loc[:,1],yerr=[or_table_pr.loc[:,1]-or_table_pr.loc[:,2],or_table_pr.loc[:,3]-or_table_pr.loc[:,1]],fmt=\"o\",markersize=7)\n",
    "ticklabel=or_table_pr[0]\n",
    "ticklabel=[ticklabel[i]+\"\\n*\" if or_table_pr.loc[i,4]<0.05 else ticklabel[i] for i in range(15)]\n",
    "plt.xticks(np.arange(15),ticklabel,fontsize=13,rotation=90)\n",
    "plt.ylabel(\"OR\",fontsize=15)\n",
    "plt.axhline(y=1,linestyle=\"--\",lw=1)\n",
    "#plt.savefig(\"topmed_all_trans_total_prs_or.png\",dpi=600,bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b119922",
   "metadata": {},
   "outputs": [],
   "source": [
    "y=\"topmed_all_trans_total_PRS\"\n",
    "temp=ukb500k_ts_som[[\"topmed_all_trans_total_PRS\", \"sex\", \"age\", \"smoking_status\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\",\"single_gene\"]].copy(deep=True)\n",
    "temp=temp.dropna()\n",
    "gene_count=Counter(temp[\"single_gene\"])\n",
    "\n",
    "left, width = 0.01, 15*0.05\n",
    "bottom_h, height_h=0.1,0.23\n",
    "bottom, height = 0.35, .60\n",
    "left_h = left+width+0.03\n",
    "rect_0 = [left, bottom, width, height]\n",
    "rect_1 = [left_h, bottom, 4*0.05, height]\n",
    "rect_2 = [left, bottom_h, width, height_h]\n",
    "rect_3 = [left_h, bottom_h, 4*0.05, height_h]\n",
    "\n",
    "plt.figure(figsize=(22,12))\n",
    "ax1=plt.axes(rect_0)\n",
    "ax2=plt.axes(rect_1)\n",
    "ax3=plt.axes(rect_2,sharex=ax1)\n",
    "ax = [ax1, ax2, ax3, plt.axes(rect_3,sharex=ax2)]\n",
    "\n",
    "gene_color_new={}\n",
    "for key in gene_color:\n",
    "    gene_color_new[key]=sns.set_hls_values(gene_color[key],s=0.5)\n",
    "black=sns.set_hls_values('#000000',s=0.5)\n",
    "\n",
    "gene_list=[]\n",
    "gene_groups=[]\n",
    "quartiles=[]\n",
    "\n",
    "gene_list.append(\"Control\")\n",
    "gene_groups.append(np.array(temp.loc[((temp[\"single_gene\"]==\"Control\") & (temp[y].notna())),y]))\n",
    "quartiles.append(np.percentile(gene_groups[-1],[25, 50, 75]))\n",
    "\n",
    "gene_list.append(\"all CH\")\n",
    "gene_groups.append(np.array(temp.loc[((temp[\"single_gene\"].isin([\"ASXL1\",\"CALR\",\"CBL\",\"DNMT3A_R882\",\"DNMT3A_other\",\"GNAS\", \"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"U2AF1\"])) & (temp[y].notna())),y]))\n",
    "quartiles.append(np.percentile(gene_groups[-1],[25, 50, 75]))\n",
    "\n",
    "for key in [\"ASXL1\",\"CALR\",\"CBL\",\"DNMT3A_R882\",\"DNMT3A_other\",\"GNAS\", \"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"U2AF1\"]: \n",
    "        if gene_count[key]>=20:\n",
    "            gene_list.append(key)\n",
    "            gene_groups.append(np.array(temp.loc[((temp[\"single_gene\"]==key) & (temp[y].notna())),y]))\n",
    "            quartiles.append(np.percentile(gene_groups[-1],[25, 50, 75]))\n",
    "quartiles=np.array(quartiles)           \n",
    "            \n",
    "if gene_list != []:\n",
    "        temp=temp.loc[((temp[\"single_gene\"].isin(gene_list)) & (temp[y].notna())),:]\n",
    "        counts=temp[\"single_gene\"].value_counts()\n",
    "        counts[\"all CH\"]=gene_groups[1].shape[0]\n",
    "        \n",
    "        parts = ax[0].violinplot(gene_groups, showmeans=False, showmedians=False, showextrema=False,widths=0.65,positions=np.arange(0,len(gene_list),1))\n",
    "\n",
    "        for i,pc in enumerate(parts['bodies']):\n",
    "            if gene_list[i]==\"Control\":\n",
    "                pc.set_facecolor(gene_color_new[\"Control\"])\n",
    "            else:\n",
    "                pc.set_facecolor(\"#999999\")\n",
    "            pc.set_edgecolor(black)\n",
    "            pc.set_alpha(1)\n",
    "            pc.set_linewidth(1.5)\n",
    "\n",
    "        ax[0].boxplot(gene_groups,positions=np.arange(0,len(gene_list),1),showcaps=False,widths=0.06, patch_artist=True, boxprops=dict(color=black, facecolor=black),whiskerprops=dict(color=black, linewidth=2), medianprops=dict(color=\"w\", linewidth=0 ),showfliers=False)\n",
    "        inds = np.arange(0, len(quartiles))\n",
    "        ax[0].scatter(inds, quartiles[:,1], marker='o', color='white', s=30, zorder=3)\n",
    "        \n",
    "        xticks_new=[]\n",
    "        for i in gene_list:\n",
    "            print(i,counts[i])\n",
    "            if i != \"Control\":\n",
    "                pval_prs=or_table_pr.loc[(or_table_pr[0]==i),4].reset_index()\n",
    "                pval_prs=pval_prs.iloc[0,1]\n",
    "            else:\n",
    "                pval_prs=1\n",
    "            if pval_prs <0.05:\n",
    "                xticks_new.append(i[0].upper()+i[1:]+\"\\n(\"+str(counts[i])+\")\\n*\")\n",
    "            else:\n",
    "                xticks_new.append(i[0].upper()+i[1:]+\"\\n(\"+str(counts[i])+\")\")\n",
    "        print(xticks_new)\n",
    "        ax[0].set_xticks(np.arange(0,len(gene_list),1))\n",
    "        #ax[0].set_xticklabels(xticks_new,rotation=45)\n",
    "        ax[0].set_xticklabels([])\n",
    "        ax[0].tick_params(axis='y',labelsize=15,size=7)\n",
    "        ax[0].tick_params(axis='x',labelsize=12,labelbottom=False)\n",
    "        ax[0].grid(axis='x')\n",
    "        ax[0].xaxis.label.set_size(20)\n",
    "        ax[0].yaxis.label.set_size(22)\n",
    "        ax[0].set_ylabel(\"PRS\")\n",
    "        ax[0].set_ylim([-4,4])\n",
    "        #ax[0].set_xlabel(\"Gene\")\n",
    "        ax[0].axhline(y=temp.loc[temp[\"single_gene\"]==\"Control\",y].median(), color='k', linestyle='--')\n",
    "        print(temp.loc[temp[\"single_gene\"]==\"Control\",y].median())\n",
    "                        \n",
    "temp=data_mca[[y,\"MCA_type\"]].copy(deep=True)\n",
    "temp=temp.dropna()\n",
    "temp[\"MCA_type\"]=temp[\"MCA_type\"].map(lambda row:\"autosomal\" if row not in [\"Y:loss\",\"X:loss\",\"Control\"] else row)\n",
    "y=\"topmed_all_trans_total_PRS\"\n",
    "\n",
    "temp=temp.loc[(temp[y].notna()),:]\n",
    "counts=temp[\"MCA_type\"].value_counts()\n",
    "print(counts.sum())\n",
    "gene_groups=[]\n",
    "gene_order=[\"Control\",\"autosomal\",\"Y:loss\",\"X:loss\"]\n",
    "quartiles=[]\n",
    "for key in gene_order:\n",
    "    gene_groups.append(temp.loc[(temp[\"MCA_type\"]==key)& (temp[y].notna()),y])\n",
    "    quartiles.append(np.percentile(gene_groups[-1],[25, 50, 75]))\n",
    "quartiles=np.array(quartiles)           \n",
    "                       \n",
    "parts = ax[1].violinplot(gene_groups, showmeans=False, showmedians=False, showextrema=False,widths=0.65,positions=np.arange(0,len(gene_order),1))\n",
    "mca_color={\"Control\":gene_color[\"Control\"],\"autosomal\":\"#999999\",\"Y:loss\":\"#E69F00\",\"X:loss\":\"#56B4E9\"}           \n",
    "for key in mca_color:\n",
    "    mca_color[key]=sns.set_hls_values(mca_color[key],s=0.5)\n",
    "    \n",
    "for i,pc in enumerate(parts['bodies']):\n",
    "            if gene_order[i]==\"Control\":\n",
    "                pc.set_facecolor(mca_color[gene_order[i]])\n",
    "            else:\n",
    "                pc.set_facecolor(\"#8585ad\")\n",
    "            pc.set_edgecolor(black)\n",
    "            pc.set_alpha(1)\n",
    "            pc.set_linewidth(1.5)\n",
    "\n",
    "ax[1].boxplot(gene_groups,positions=np.arange(0,len(gene_order),1),showcaps=False,widths=0.06, patch_artist=True,boxprops=dict(color=black, facecolor=black),\n",
    "            whiskerprops=dict(color=black, linewidth=2),\n",
    "            medianprops=dict(color=\"w\", linewidth=0 ),showfliers=False)                       \n",
    "inds = np.arange(0, len(quartiles))\n",
    "ax[1].scatter(inds, quartiles[:,1], marker='o', color='white', s=30, zorder=3)                       \n",
    "xticks_new_mca=[]\n",
    "gene_rename={\"autosomal\":\"autosomal\",\"Y:loss\":\"LOY\",\"X:loss\":\"LOX\"}\n",
    "for j,i in enumerate(gene_order):\n",
    "    print(i,counts[i])\n",
    "    if i!=\"Control\":\n",
    "        if plot_values[j-1,3]<0.05:\n",
    "            xticks_new_mca.append(gene_rename[i]+\"\\n(\"+str(counts[i])+\")\"+\"\\n*\")\n",
    "        else:\n",
    "            xticks_new_mca.append(gene_rename[i]+\"\\n(\"+str(counts[i])+\")\")\n",
    "    else:\n",
    "        xticks_new_mca.append(i+\"\\n(\"+str(counts[i])+\")\")\n",
    "\n",
    "ax[1].set_xticks(np.arange(0,len(xticks),1))\n",
    "ax[1].set_xticklabels([])\n",
    "ax[1].grid(axis='x')\n",
    "ax[1].tick_params(axis='x',labelsize=12,labelbottom=False)\n",
    "ax[1].tick_params(axis='y',labelsize=15)\n",
    "ax[1].xaxis.label.set_size(25)\n",
    "ax[1].yaxis.label.set_size(25)\n",
    "ax[1].set_ylim([-4,4])\n",
    "\n",
    "ax[1].axhline(y=temp.loc[temp[\"MCA_type\"]==\"Control\",y].median(), color='k', linestyle='--')\n",
    "plt.subplots_adjust(wspace=0.05)\n",
    "\n",
    "#or_table_pr=or_table_pr.set_index(0).loc[[x for x in gene_list if x!=\"Control\"],:]\n",
    "#or_table_pr=or_table_pr.reset_index()\n",
    "or_table_pr=or_table_pr.set_index(0).loc[[x for x in gene_list if x!=\"Control\"],:]\n",
    "or_table_pr=or_table_pr.reset_index()\n",
    "or_significant=or_table_pr.loc[(or_table_pr[4]>=0.05),:]\n",
    "ax[2].errorbar(or_significant.index+1,or_significant.loc[:,1],yerr=[or_significant.loc[:,1]-or_significant.loc[:,2],or_significant.loc[:,3]-or_significant.loc[:,1]],fmt=\"o\",markersize=20,elinewidth=2,c=\"#b3b3b3\",label=\"P\"+r'$\\mathrm{\\geq}0.05$')\n",
    "or_significant=or_table_pr.loc[(or_table_pr[4]<0.05),:]\n",
    "ax[2].errorbar(or_significant.index+1,or_significant.loc[:,1],yerr=[or_significant.loc[:,1]-or_significant.loc[:,2],or_significant.loc[:,3]-or_significant.loc[:,1]],fmt=\"o\",markersize=20,elinewidth=2,c=\"#4d4d4d\",label=\"P\"+r'$\\mathrm{<}0.05$')\n",
    "\n",
    "ax[2].set_xticks(np.arange(0,16))\n",
    "ax[2].set_xticklabels(xticks_new,fontsize=15,rotation=35)\n",
    "ax[2].set_ylabel(\"OR\",fontsize=22)\n",
    "ax[2].set_xlabel(\"CH driver gene\",fontsize=22)\n",
    "ax[2].tick_params(axis='y',labelsize=15,size=7)\n",
    "ax[2].axhline(y=1,linestyle=\"--\",lw=1)\n",
    "ax[2].legend()\n",
    "\n",
    "\n",
    "plot_values_s=plot_values[(plot_values[:,3]<0.05),:]\n",
    "ax[3].errorbar(np.where(plot_values[:,3]<0.05)[0]+1,plot_values_s[:,0],yerr=[plot_values_s[:,0]-plot_values_s[:,1],plot_values_s[:,2]-plot_values_s[:,0]],fmt=\"o\",markersize=20,elinewidth=2,c=\"#575782\",label=\"P\"+r'$\\mathrm{<}0.05$')\n",
    "\n",
    "plot_values_n=plot_values[plot_values[:,3]>=0.05,:]\n",
    "ax[3].errorbar(np.where(plot_values[:,3]>=0.05)[0]+1,plot_values_n[:,0],yerr=[plot_values_n[:,0]-plot_values_n[:,1],plot_values_n[:,2]-plot_values_n[:,0]],fmt=\"o\",markersize=20,elinewidth=2,c=\"#b1b1cb\",label=\"P\"+r'$\\mathrm{\\geq}0.05$')\n",
    "ax[3].set_xticks(np.arange(0,4))\n",
    "ax[3].set_xticklabels(xticks_new_mca,fontsize=15,rotation=45)\n",
    "ax[3].tick_params(axis='y',labelsize=15)\n",
    "ax[3].axhline(y=1,linestyle=\"--\",lw=1)\n",
    "ax[3].set_xlabel(\"mCA type\",fontsize=22)\n",
    "ax[3].legend()\n",
    "\n",
    "for a in ax:\n",
    "    a.spines[['right', 'top']].set_visible(False)\n",
    "\n",
    "#plt.savefig(\"../figures_revision1/topmedprs_ch_or_mca_singlegene.png\",dpi=600,bbox_inches=\"tight\")\n",
    "#plt.savefig(\"../figures_revision1/topmedprs_ch_or_mca_singlegene.pdf\",bbox_inches=\"tight\")\n",
    "#plt.savefig(\"../figures_revision1/topmedprs_ch_or_mca_singlegene.svg\",bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "307480b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "supptable_topmed_gene=or_table_pr.copy(deep=True)\n",
    "supptable_topmed_gene.columns=[\"Gene/mCA\",\"OR\",\"OR lower CI (2.5%)\",\"OR upper CI (97.5%)\",\"P-value\"]\n",
    "supptable_topmed_gene=supptable_topmed_gene.sort_values(by=\"Gene/mCA\",key=lambda col: col.str.lower())\n",
    "supptable_topmed_gene[\"FDR\"]=multipletests(supptable_topmed_gene[\"P-value\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n",
    "supptable_topmed=pd.concat([supptable_topmed_gene[[\"Gene/mCA\",\"OR\",\"OR lower CI (2.5%)\",\"OR upper CI (97.5%)\",\"P-value\",\"FDR\"]],supptable_topmed_mca[[\"Gene/mCA\",\"OR\",\"OR lower CI (2.5%)\",\"OR upper CI (97.5%)\",\"P-value\",\"FDR\"]]])\n",
    "supptable_topmed[\"FDR_all\"]=multipletests(supptable_topmed[\"P-value\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c140ebd",
   "metadata": {},
   "outputs": [],
   "source": [
    "supptable_topmed.to_csv(\"prs_coef_topmed_supptable_r1_singlegene_check.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8a646e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "supptable_topmed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01dba836",
   "metadata": {},
   "outputs": [],
   "source": [
    "######### Load AllofUs data ############\n",
    "aou_highvaf_prs=pd.read_csv(\"../allofus/ltl_prs_zscores_EUR_aou_09162024.tsv\",delimiter=\"\\t\")\n",
    "aou_highvaf_or=pd.read_csv(\"../allofus/ltl_prs_OR_by_gene_EUR_aou_09162024.tsv\",delimiter=\"\\t\")\n",
    "aou_singlemut_prs=pd.read_csv(\"../allofus/ltl_prs_zscores_EUR_aou_singlemutant_09202024.tsv\",delimiter=\"\\t\")\n",
    "aou_singlemut_or=pd.read_csv(\"../allofus/ltl_prs_OR_by_gene_EUR_aou_singlemutant_09202024.tsv\",delimiter=\",\")\n",
    "aou_singlemut_or=aou_singlemut_or[[\"Gene\",\"OR\",\"CI_lower\",\"CI_upper\",\"P\",\"N\"]]\n",
    "\n",
    "aou_highvaf_or.loc[len(aou_highvaf_or.index)] = ['Splicing', 0.91665, 0.832261,1.009595,0.077368,391]\n",
    "aou_highvaf_or.loc[len(aou_highvaf_or.index)] = ['Splicing+PPM1D', 0.916923, 0.832506,1.009901,0.078403,754]\n",
    "aou_singlemut_or.loc[len(aou_singlemut_or.index)] = ['Splicing',0.919832,0.823256,1.027737,0.139799,296]\n",
    "aou_singlemut_or.loc[len(aou_singlemut_or.index)] = ['Splicing+PPM1D',0.92009,0.82349,1.028021,0.141118,603]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "051ec596",
   "metadata": {},
   "outputs": [],
   "source": [
    "aou_highvaf_or[aou_highvaf_or[\"Gene\"].isin([\"Control\",\"ASXL1\",\"CBL\",\"DNMT3A\",\"GNAS\", \"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"U2AF1\",\"Splicing\",\"Splicing+PPM1D\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4be2445b",
   "metadata": {},
   "outputs": [],
   "source": [
    "supptable_aou_gene=aou_highvaf_or[aou_highvaf_or[\"Gene\"].isin([\"Control\",\"ASXL1\",\"CBL\",\"DNMT3A\",\"GNAS\", \"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"U2AF1\",\"Splicing\",\"Splicing+PPM1D\"])].copy(deep=True)\n",
    "supptable_aou_gene[\"FDR\"]=multipletests(supptable_aou_gene[\"P\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n",
    "\n",
    "#supptable_aou_gene.to_csv(\"supptable_aou_highvaf_r1.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "964e2a0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "supptable_aou_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb3d271e",
   "metadata": {},
   "outputs": [],
   "source": [
    "supptable_aou_gene=aou_singlemut_or[aou_singlemut_or[\"Gene\"].isin([\"Control\",\"ASXL1\",\"CBL\",\"DNMT3A\",\"GNAS\", \"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"U2AF1\",\"Splicing\",\"Splicing+PPM1D\"])].copy(deep=True)\n",
    "supptable_aou_gene[\"FDR\"]=multipletests(supptable_aou_gene[\"P\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n",
    "supptable_aou_gene.loc[supptable_aou_gene[\"Gene\"].isin([\"ASXL1\",\"CBL\",\"GNAS\", \"GNB1\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TP53\",\"U2AF1\"]),\"FDR_selected\"]=multipletests(supptable_aou_gene.loc[supptable_aou_gene[\"Gene\"].isin([\"ASXL1\",\"CBL\",\"GNAS\", \"GNB1\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TP53\",\"U2AF1\"]),\"P\"], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]\n",
    "#supptable_aou_gene.to_csv(\"supptable_aou_singlemut_r1.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d484ca70",
   "metadata": {},
   "outputs": [],
   "source": [
    "y=\"ltl_prs_zscore\"\n",
    "temp=aou_singlemut_prs.copy(deep=True)\n",
    "temp=temp.rename(columns={\"Gene.refGene\":\"single_gene\"})\n",
    "temp=temp.dropna()\n",
    "gene_count=Counter(temp[\"single_gene\"])\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(2, 1,sharex=False,sharey=False,figsize=(23,12),gridspec_kw={'height_ratios':[5,1.5]})\n",
    "\n",
    "\n",
    "gene_color_new={}\n",
    "for key in gene_color:\n",
    "    gene_color_new[key]=sns.set_hls_values(gene_color[key],s=0.5)\n",
    "black=sns.set_hls_values('#000000',s=0.5)\n",
    "\n",
    "gene_list=[]\n",
    "gene_groups=[]\n",
    "quartiles=[]\n",
    "for key in [\"Control\",\"ASXL1\",\"CBL\",\"DNMT3A\",\"GNAS\", \"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"U2AF1\",\"Splicing\",\"Splicing+PPM1D\"]: \n",
    "            gene_list.append(key)\n",
    "            if key == \"Splicing\":\n",
    "                gene_groups.append(np.array(temp.loc[((temp[\"single_gene\"].isin([\"SF3B1\",\"SRSF2\",\"U2AF1\"])) & (temp[y].notna())),y]))\n",
    "            elif key==\"Splicing+PPM1D\":\n",
    "                gene_groups.append(np.array(temp.loc[((temp[\"single_gene\"].isin([\"SF3B1\",\"SRSF2\",\"U2AF1\",\"PPM1D\"])) & (temp[y].notna())),y]))\n",
    "            else:\n",
    "                gene_groups.append(np.array(temp.loc[((temp[\"single_gene\"]==key) & (temp[y].notna())),y]))\n",
    "            quartiles.append(np.percentile(gene_groups[-1],[25, 50, 75]))\n",
    "quartiles=np.array(quartiles)\n",
    "print(gene_list)\n",
    "\n",
    "or_table_pr=aou_singlemut_or.set_index(\"Gene\").loc[[x for x in gene_list if x!=\"Control\"],:]\n",
    "or_table_pr=or_table_pr.reset_index()\n",
    "\n",
    "if gene_list != []:\n",
    "        temp=temp.loc[((temp[\"single_gene\"].isin(gene_list)) & (temp[y].notna())),:]\n",
    "        counts=temp[\"single_gene\"].value_counts()\n",
    "        counts[\"Splicing\"]=gene_groups[-2].shape[0]\n",
    "        counts[\"Splicing+PPM1D\"]=gene_groups[-1].shape[0]\n",
    "        print(counts)\n",
    "        \n",
    "        parts = ax[0].violinplot(gene_groups, showmeans=False, showmedians=False, showextrema=False,widths=0.65,positions=np.arange(0,len(gene_list),1))\n",
    "\n",
    "        for i,pc in enumerate(parts['bodies']):\n",
    "            if gene_list[i]==\"Control\":\n",
    "                pc.set_facecolor(gene_color_new[\"Control\"])\n",
    "            else:\n",
    "                pc.set_facecolor(\"#999999\")\n",
    "            pc.set_edgecolor(black)\n",
    "            pc.set_alpha(1)\n",
    "            pc.set_linewidth(1.5)\n",
    "\n",
    "        ax[0].boxplot(gene_groups,positions=np.arange(0,len(gene_list),1),showcaps=False,widths=0.06, patch_artist=True, boxprops=dict(color=black, facecolor=black),whiskerprops=dict(color=black, linewidth=2), medianprops=dict(color=\"w\", linewidth=0 ),showfliers=False)\n",
    "        inds = np.arange(0, len(quartiles))\n",
    "        ax[0].scatter(inds, quartiles[:,1], marker='o', color='white', s=30, zorder=3)\n",
    "        \n",
    "        xticks_new=[]\n",
    "        for i in gene_list:\n",
    "            print(i,counts[i])\n",
    "            if i != \"Control\":\n",
    "                pval_prs=or_table_pr.loc[(or_table_pr[\"Gene\"]==i),\"P\"].reset_index()\n",
    "                pval_prs=pval_prs.iloc[0,1]\n",
    "            else:\n",
    "                pval_prs=1\n",
    "            if pval_prs <0.05:\n",
    "                xticks_new.append(i+\"\\n(\"+str(counts[i])+\")\")\n",
    "            else:\n",
    "                xticks_new.append(i+\"\\n(\"+str(counts[i])+\")\")\n",
    "        print(xticks_new)\n",
    "        ax[0].set_xticks(np.arange(0,len(gene_list),1))\n",
    "        #ax[0].set_xticklabels(xticks_new,rotation=45)\n",
    "        ax[0].set_xticklabels([])\n",
    "        ax[0].tick_params(axis='y',labelsize=15,size=7)\n",
    "        ax[0].tick_params(axis='x',labelsize=12,labelbottom=False)\n",
    "        ax[0].grid(axis='x')\n",
    "        ax[0].xaxis.label.set_size(20)\n",
    "        ax[0].yaxis.label.set_size(22)\n",
    "        ax[0].set_ylabel(\"PRS\")\n",
    "        ax[0].set_ylim([-4,4])\n",
    "        #ax[0].set_xlabel(\"Gene\")\n",
    "        ax[0].axhline(y=temp.loc[temp[\"single_gene\"]==\"Control\",y].median(), color='k', linestyle='--')\n",
    "        print(temp.loc[temp[\"single_gene\"]==\"Control\",y].median())\n",
    "                        \n",
    "\n",
    "\n",
    "or_significant=or_table_pr.loc[(or_table_pr[\"P\"]>=0.05),:]\n",
    "print(or_significant)\n",
    "ax[1].errorbar(or_significant.index+1,or_significant.loc[:,\"OR\"],yerr=[or_significant.loc[:,\"OR\"]-or_significant.loc[:,\"CI_lower\"],or_significant.loc[:,\"CI_upper\"]-or_significant.loc[:,\"OR\"]],fmt=\"o\",markersize=15,elinewidth=4,c=\"#b3b3b3\",label=\"P\"+r'$\\mathrm{\\geq}0.05$')\n",
    "or_significant=or_table_pr.loc[(or_table_pr[\"P\"]<0.05),:]\n",
    "ax[1].errorbar(or_significant.index+1,or_significant.loc[:,\"OR\"],yerr=[or_significant.loc[:,\"OR\"]-or_significant.loc[:,\"CI_lower\"],or_significant.loc[:,\"CI_upper\"]-or_significant.loc[:,\"OR\"]],fmt=\"o\",markersize=15,elinewidth=4,c=\"#4d4d4d\",label=\"P\"+r'$\\mathrm{<}0.05$')\n",
    "\n",
    "\n",
    "ax[1].set_xticks(np.arange(0,15))\n",
    "ax[1].set_xticklabels(xticks_new,fontsize=15,rotation=35)\n",
    "ax[1].set_ylabel(\"OR\",fontsize=22)\n",
    "ax[1].set_xlabel(\"CH driver gene\",fontsize=22)\n",
    "ax[1].tick_params(axis='y',labelsize=15,size=7)\n",
    "ax[1].axhline(y=1,linestyle=\"--\",lw=1)\n",
    "ax[1].legend()\n",
    "\n",
    "\n",
    "for a in ax:\n",
    "    a.spines[['right', 'top']].set_visible(False)\n",
    "\n",
    "#plt.savefig(\"../figures_revision1/prs_ch_singlemut_or.png\",dpi=600,bbox_inches=\"tight\")\n",
    "#plt.savefig(\"../figures_revision1/prs_ch_singlemut_or.pdf\",bbox_inches=\"tight\")\n",
    "#plt.savefig(\"../figures_revision1/prs_ch_singlemut_or.svg\",bbox_inches=\"tight\")\n",
    "\n",
    "\n",
    "plt.show()\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe992d1f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "470d5dae",
   "metadata": {},
   "outputs": [],
   "source": [
    "############################ Response to reviewer ###########################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b2e0225",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "############# Association with all CH and PRS for all genes with n>50 ###############\n",
    "y=\"total_PRS\"\n",
    "temp=ukb500k_ts_som_allgenes.copy(deep=True)\n",
    "temp=temp.loc[(temp[y].notna())]\n",
    "\n",
    "formula = 'single_gene ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10','total_PRS'])\n",
    "\n",
    "or_table_pr=[]\n",
    "\n",
    "data=temp[[\"total_PRS\", \"sex\", \"age\", \"smoking_status\", \"PC1\", \"PC2\", \"PC3\", \"PC4\", \"PC5\", \"PC6\", \"PC7\", \"PC8\", \"PC9\", \"PC10\", \"single_gene\"]].copy(deep=True)\n",
    "data=data.dropna()\n",
    "print(data.shape)\n",
    "data[\"single_gene\"]=data[\"single_gene\"].map(lambda row:0 if row==\"Control\" else 1)\n",
    "data[\"single_gene\"]=data[\"single_gene\"].astype(int)\n",
    "print(data[\"single_gene\"].value_counts())\n",
    "log_model_allch=smf.logit(formula=formula,data=data)\n",
    "result_allch=log_model_allch.fit(max_iterations= 100000)\n",
    "display(result_allch.summary2())\n",
    "params_allch = np.exp(result_allch.params)\n",
    "conf_allch = np.exp(result_allch.conf_int())\n",
    "print(result_allch.pvalues)\n",
    "print([\"all CH\", params_allch[\"total_PRS\"], conf_allch.loc[\"total_PRS\",0], conf_allch.loc[\"total_PRS\",1], result_allch.pvalues[\"total_PRS\"]])\n",
    "or_table_pr.append([\"all CH\", params_allch[\"total_PRS\"], conf_allch.loc[\"total_PRS\",0], conf_allch.loc[\"total_PRS\",1], result_allch.pvalues[\"total_PRS\"]])\n",
    "\n",
    "############# Association with CH and PRS for all genes with n>50 ###############\n",
    "\n",
    "\n",
    "y=\"total_PRS\"\n",
    "temp=ukb500k_ts_som_allgenes.copy(deep=True)\n",
    "temp=temp.loc[(temp[y].notna())]\n",
    "\n",
    "formula = 'single_gene ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10','total_PRS'])\n",
    "gene_list=list(temp[\"single_gene\"].unique())\n",
    "gene_list.remove(\"Control\")\n",
    "\n",
    "for gene in gene_list:\n",
    "    data=temp[[\"total_PRS\", \"sex\", \"age\", \"smoking_status\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\",\"single_gene\"]].copy(deep=True)\n",
    "    data=data.dropna()\n",
    "    print(data.shape)\n",
    "    data=data[(data[\"single_gene\"].isin([gene,\"Control\"]))]\n",
    "    data[\"single_gene\"]=data[\"single_gene\"].map({gene:1,\"Control\":0})\n",
    "    data[\"single_gene\"]=data[\"single_gene\"].astype(int)\n",
    "    print(data[\"single_gene\"].value_counts())\n",
    "    log_model=smf.logit(formula=formula,data=data)\n",
    "    result=log_model.fit(max_iterations= 100000)\n",
    "    display(result.summary2())\n",
    "    params = np.exp(result.params)\n",
    "    conf = np.exp(result.conf_int())\n",
    "    print(result.pvalues)\n",
    "    print([gene,params[\"total_PRS\"],conf.loc[\"total_PRS\",0],conf.loc[\"total_PRS\",1],result.pvalues[\"total_PRS\"]])\n",
    "    or_table_pr.append([gene,params[\"total_PRS\"],conf.loc[\"total_PRS\",0],conf.loc[\"total_PRS\",1],result.pvalues[\"total_PRS\"]])\n",
    "or_table_pr=pd.DataFrame(or_table_pr)\n",
    "\n",
    "or_table_pr\n",
    "\n",
    "or_table_pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83bdeedc",
   "metadata": {},
   "outputs": [],
   "source": [
    "y=\"total_PRS\"\n",
    "temp=ukb500k_ts_som_allgenes[[\"total_PRS\", \"sex\", \"age\", \"smoking_status\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\",\"single_gene\"]].copy(deep=True)\n",
    "temp=temp.dropna()\n",
    "gene_count=Counter(temp[\"single_gene\"])\n",
    "\n",
    "\"\"\"\n",
    "left, width = 0.02, 15*0.05\n",
    "bottom_h, height_h=0.1,0.23\n",
    "bottom, height = 0.35, .60\n",
    "left_h = left+width+0.02\n",
    "rect_0 = [left, bottom, width, height]\n",
    "rect_1 = [left_h, bottom, 4*0.05, height]\n",
    "rect_2 = [left, bottom_h, width, height_h]\n",
    "rect_3 = [left_h, bottom_h, 4*0.05, height_h]\n",
    "\n",
    "plt.figure(figsize=(22,12))\n",
    "ax1=plt.axes(rect_0)\n",
    "ax2=plt.axes(rect_1)\n",
    "ax3=plt.axes(rect_2,sharex=ax1)\n",
    "ax = [ax1, ax2, ax3, plt.axes(rect_3,sharex=ax2)]\n",
    "\"\"\"\n",
    "\n",
    "fig, ax = plt.subplots(2, 1,sharex=False,sharey=False,figsize=(26,12),gridspec_kw={'height_ratios':[5,1.5]})\n",
    "\n",
    "gene_color_new={}\n",
    "for key in gene_color_allgenes:\n",
    "    gene_color_new[key]=sns.set_hls_values(gene_color_allgenes[key],s=0.5)\n",
    "black=sns.set_hls_values('#000000',s=0.5)\n",
    "\n",
    "gene_list=[]\n",
    "gene_groups=[]\n",
    "quartiles=[]\n",
    "\n",
    "gene_list.append(\"Control\")\n",
    "gene_groups.append(np.array(temp.loc[((temp[\"single_gene\"]==\"Control\") & (temp[y].notna())),y]))\n",
    "quartiles.append(np.percentile(gene_groups[-1],[25, 50, 75]))\n",
    "\n",
    "gene_list.append(\"all CH\")\n",
    "gene_groups.append(np.array(temp.loc[((temp[\"single_gene\"].isin([\"ASXL1\",\"CALR\",\"CBL\",\"DNMT3A_other\",\"DNMT3A_R882\",\"GNAS\",\"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"U2AF1\",\"ATM\",\"STAT3\",\"BRCC3\",\"CHEK2\",\"KDM6A\",\"KRAS\",\"BCOR\", \"STAG2\",\"BCORL1\",\"IDH2\"])) & (temp[y].notna())),y]))\n",
    "quartiles.append(np.percentile(gene_groups[-1],[25, 50, 75]))\n",
    "\n",
    "for key in [\"ASXL1\",\"CALR\",\"CBL\",\"DNMT3A_other\",\"DNMT3A_R882\",\"GNAS\",\"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"U2AF1\",\"ATM\",\"STAT3\",\"BRCC3\",\"CHEK2\",\"KDM6A\",\"KRAS\",\"BCOR\", \"STAG2\",\"BCORL1\",\"IDH2\"]: \n",
    "        if gene_count[key]>=20:\n",
    "            gene_list.append(key)\n",
    "            gene_groups.append(np.array(temp.loc[((temp[\"single_gene\"]==key) & (temp[y].notna())),y]))\n",
    "            quartiles.append(np.percentile(gene_groups[-1],[25, 50, 75]))\n",
    "quartiles=np.array(quartiles)           \n",
    "            \n",
    "if gene_list != []:\n",
    "        temp=temp.loc[((temp[\"single_gene\"].isin(gene_list)) & (temp[y].notna())),:]\n",
    "        counts=temp[\"single_gene\"].value_counts()\n",
    "        counts[\"all CH\"]=gene_groups[1].shape[0]\n",
    "        \n",
    "        parts = ax[0].violinplot(gene_groups, showmeans=False, showmedians=False, showextrema=False,widths=0.65,positions=np.arange(0,len(gene_list),1))\n",
    "\n",
    "        for i,pc in enumerate(parts['bodies']):\n",
    "            if gene_list[i]==\"Control\":\n",
    "                pc.set_facecolor(gene_color_new[\"Control\"])\n",
    "            else:\n",
    "                pc.set_facecolor(\"#999999\")\n",
    "            pc.set_edgecolor(black)\n",
    "            pc.set_alpha(1)\n",
    "            pc.set_linewidth(1.5)\n",
    "\n",
    "        ax[0].boxplot(gene_groups,positions=np.arange(0,len(gene_list),1),showcaps=False,widths=0.06, patch_artist=True, boxprops=dict(color=black, facecolor=black),whiskerprops=dict(color=black, linewidth=2), medianprops=dict(color=\"w\", linewidth=0 ),showfliers=False)\n",
    "        inds = np.arange(0, len(quartiles))\n",
    "        ax[0].scatter(inds, quartiles[:,1], marker='o', color='white', s=30, zorder=3)\n",
    "        \n",
    "        xticks_new=[]\n",
    "        for i in gene_list:\n",
    "            print(i,counts[i])\n",
    "            if i != \"Control\":\n",
    "                pval_prs=or_table_pr.loc[(or_table_pr[0]==i),4].reset_index()\n",
    "                pval_prs=pval_prs.iloc[0,1]\n",
    "            else:\n",
    "                pval_prs=1\n",
    "            if pval_prs <0.05:\n",
    "                xticks_new.append(i[0].upper()+i[1:]+\"\\n(\"+str(counts[i])+\")\\n*\")\n",
    "            else:\n",
    "                xticks_new.append(i[0].upper()+i[1:]+\"\\n(\"+str(counts[i])+\")\")\n",
    "        print(xticks_new)\n",
    "        ax[0].set_xticks(np.arange(0,len(gene_list),1))\n",
    "        #ax[0].set_xticklabels(xticks_new,rotation=45)\n",
    "        ax[0].set_xticklabels([])\n",
    "        ax[0].tick_params(axis='y',labelsize=15,size=7)\n",
    "        ax[0].tick_params(axis='x',labelsize=12,labelbottom=False)\n",
    "        ax[0].grid(axis='x')\n",
    "        ax[0].xaxis.label.set_size(20)\n",
    "        ax[0].yaxis.label.set_size(22)\n",
    "        ax[0].set_ylabel(\"PRS\")\n",
    "        ax[0].set_ylim([-4,4])\n",
    "        #ax[0].set_xlabel(\"Gene\")\n",
    "        ax[0].axhline(y=temp.loc[temp[\"single_gene\"]==\"Control\",y].median(), color='k', linestyle='--')\n",
    "        print(temp.loc[temp[\"single_gene\"]==\"Control\",y].median())\n",
    "                        \n",
    "\n",
    "#plt.savefig(\"../figures_new/prs_ch_mca_new.png\",dpi=600,bbox_inches=\"tight\")\n",
    "#plt.savefig(\"../figures_new/prs_ch_mca_new.pdf\",bbox_inches=\"tight\")\n",
    "\n",
    "\n",
    "or_table_pr=or_table_pr.set_index(0).loc[[x for x in gene_list if x!=\"Control\"],:]\n",
    "or_table_pr=or_table_pr.reset_index()\n",
    "or_significant=or_table_pr.loc[(or_table_pr[4]>=0.05),:]\n",
    "ax[1].errorbar(or_significant.index+1,or_significant.loc[:,1],yerr=[or_significant.loc[:,1]-or_significant.loc[:,2],or_significant.loc[:,3]-or_significant.loc[:,1]],fmt=\"o\",markersize=15,elinewidth=1,c=\"#b3b3b3\",label=\"P\"+r'$\\mathrm{\\geq}0.05$')\n",
    "or_significant=or_table_pr.loc[(or_table_pr[4]<0.05),:]\n",
    "ax[1].errorbar(or_significant.index+1,or_significant.loc[:,1],yerr=[or_significant.loc[:,1]-or_significant.loc[:,2],or_significant.loc[:,3]-or_significant.loc[:,1]],fmt=\"o\",markersize=15,elinewidth=1,c=\"#4d4d4d\",label=\"P\"+r'$\\mathrm{<}0.05$')\n",
    "\n",
    "ax[1].set_xticks(np.arange(0,len(gene_list)))\n",
    "ax[1].set_xticklabels(xticks_new,fontsize=15,rotation=35)\n",
    "ax[1].set_ylabel(\"OR\",fontsize=22)\n",
    "ax[1].set_xlabel(\"CH driver gene\",fontsize=22)\n",
    "ax[1].tick_params(axis='y',labelsize=15,size=7)\n",
    "ax[1].axhline(y=1,linestyle=\"--\",lw=1)\n",
    "ax[1].legend()\n",
    "\n",
    "\n",
    "\n",
    "for a in ax:\n",
    "    a.spines[['right', 'top']].set_visible(False)\n",
    "\n",
    "#plt.savefig(\"../figures_revision1/prs_ch_or_allgenes_singlemut.png\",dpi=600,bbox_inches=\"tight\")\n",
    "#plt.savefig(\"../figures_revision1/prs_ch_or_allgenes_singlemut.pdf\",bbox_inches=\"tight\")\n",
    "#plt.savefig(\"../figures_revision1/prs_ch_or_allgenes_singlemut.svg\",bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "899505f3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "data=ukb500k_ts_som[[\"Z-adjusted T/S log\", \"sex\", \"age\", \"smoking_status\", \"single_gene\", \"single_gene_vaf\",\"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\", \"WBC_wins\", \"MO_perc_wins\", \"LY_perc_wins\", \"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",\"eid\",\"HM\"]].copy(deep=True)\n",
    "print(data.shape,data[data[\"HM\"].isna()].shape)\n",
    "\n",
    "\n",
    "data=data.dropna()\n",
    "data=data[(data[\"HM\"]==0)]\n",
    "print(data.shape,data[data[\"HM\"].isna()].shape)\n",
    "print(data.shape)\n",
    "\n",
    "\n",
    "data=data.rename(columns={\"Z-adjusted T/S log\":\"tl\"})\n",
    "formula = 'tl ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\"])\n",
    "\n",
    "lin_model=smf.ols(formula=formula,data=data)\n",
    "result_fullreg=lin_model.fit(max_iterations= 100000)\n",
    "display(result_fullreg.summary2())\n",
    "params_fullreg = result_fullreg.params\n",
    "conf_fullreg = result_fullreg.conf_int()\n",
    "\n",
    "print(result_fullreg.pvalues)\n",
    "data=data.merge(result_fullreg.resid.rename(\"residuals\"),left_index=True,right_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7218ff1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pvalue_table=pd.read_csv(\"/home/sruthi/Downloads/2024-12-26_three-bin-comparison_pairwise-wilcox_bonferroni3.csv\",usecols=[\"single_gene\",\"vaf_group_comparison\",\"p.value\"],dtype=\"str\")\n",
    "#pvalue_table=pd.read_csv(\"~/Downloads/2025-01-24_single-gene_noHM_three-bin-comparison_pairwise-wilcox_bonferroni3.csv\",usecols=[\"single_gene\",\"vaf_group_comparison\",\"adj.p.value\"],dtype=\"str\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de07ffa4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "y=\"residuals\"\n",
    "#pvalue_table=pd.read_csv(\"/home/sruthi/Downloads/2024-12-26_three-bin-comparison_pairwise-wilcox_bonferroni3.csv\",usecols=[\"single_gene\",\"vaf_group_comparison\",\"p.value\"])\n",
    "pvalue_table=pd.read_csv(\"~/Downloads/2025-01-24_single-gene_noHM_three-bin-comparison_pairwise-wilcox_bonferroni3.csv\",usecols=[\"single_gene\",\"vaf_group_comparison\",\"adj.p.value\"],dtype=\"str\")\n",
    "pvalue_table=pvalue_table.rename(columns={\"adj.p.value\":\"p.value\"})\n",
    "print(data.shape)\n",
    "overlapping_bins=[0,0.1,0.2]\n",
    "box_groups=[data.loc[(data[\"single_gene\"]==\"Control\") & (data[y].notna()),y]]\n",
    "tick_labels=[\"Control\"]\n",
    "box_colors=[gene_color[\"Control\"]]\n",
    "ticks=[-1+0.5]\n",
    "j=0\n",
    "plt.figure(figsize=(16,7))\n",
    "ax=plt.gca()\n",
    "box_color_pal=sns.color_palette(\"viridis\", 3)\n",
    "vaf_bin_counts=[\"Control\\n(\"+str(box_groups[0].shape[0])+\")\"]\n",
    "vaf_mean_all=[\"\"]\n",
    "for gene in [\"ASXL1\",\"CALR\",\"CBL\",\"DNMT3A_R882\",\"DNMT3A_other\",\"GNAS\",\"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"U2AF1\"]:\n",
    "    temp=data[data[\"single_gene\"]==gene]\n",
    "    temp=temp[temp[y].notna()]\n",
    "    print(temp[\"single_gene_vaf\"])\n",
    " #   box_groups=[]\n",
    "  #  tick_labels=[]\n",
    "  #  ticks=[]\n",
    "    gene_vaf_bin_counts=[]\n",
    "    vaf_mean_all=[]\n",
    "    \n",
    "    df1=temp.loc[(temp[\"single_gene_vaf\"]<=0.1),:]\n",
    "    df2=temp.loc[((temp[\"single_gene_vaf\"]>0.1)&(temp[\"single_gene_vaf\"]<=0.2)),:]\n",
    "    df3=temp.loc[(temp[\"single_gene_vaf\"]>0.2),:]\n",
    "    dfs=[df1,df2,df3]\n",
    "    #vaf_mean_all=[df1[\"single_gene_vaf\"].mean(),df2[\"single_gene_vaf\"].mean(),df3[\"single_gene_vaf\"].mean()]\n",
    "    \n",
    "    gene_pos=[]\n",
    "    for n,df in enumerate(dfs):\n",
    "        if df.shape[0]>10:\n",
    "            box_groups.append(df[y])\n",
    "            gene_vaf_bin_counts.append(str(df.shape[0]))\n",
    "            vaf_mean_all.append(\"{:.2f}\".format(df[\"single_gene_vaf\"].mean()))\n",
    "            box_colors.append(box_color_pal[n])    \n",
    "            ticks.append(j+0.2+0.2*n)\n",
    "            gene_pos.append(j+0.2+0.2*n)\n",
    "    \n",
    "    ypos={'1v2':3,'1v3':3.3,'2v3':3.6}\n",
    "    for index, row in pvalue_table.loc[pvalue_table[\"single_gene\"]==gene].iterrows():\n",
    "        if float(row[\"p.value\"])<0.05:        \n",
    "            x=str(row[\"vaf_group_comparison\"]).split(\"v\")\n",
    "            print (x)\n",
    "            #plt.plot([gene_pos[int(x[0])-1],gene_pos[int(x[1])-1],[1,2]\n",
    "            plt.plot([gene_pos[int(x[0])-1],gene_pos[int(x[1])-1]],[ypos[row[\"vaf_group_comparison\"]],ypos[row[\"vaf_group_comparison\"]]],c=\"k\",linewidth=0.5)\n",
    "            print (gene_pos[int(x[1])-1]-gene_pos[int(x[0])-1])\n",
    "            if float(row[\"p.value\"])<0.001:\n",
    "                ax.text((gene_pos[int(x[1])-1]+gene_pos[int(x[0])-1])/2,ypos[row[\"vaf_group_comparison\"]]+0.05,\"***\",horizontalalignment='center',verticalalignment='center')\n",
    "            #ax.text((gene_pos[int(x[1])-1]+gene_pos[int(x[0])-1])/2,3.1,\"*\")\n",
    "            elif float(row[\"p.value\"])<0.01:\n",
    "                ax.text((gene_pos[int(x[1])-1]+gene_pos[int(x[0])-1])/2,ypos[row[\"vaf_group_comparison\"]]+0.05,\"**\",horizontalalignment='center',verticalalignment='center')\n",
    "            #ax.text((gene_pos[int(x[1])-1]+gene_pos[int(x[0])-1])/2,3.1,\"*\")\n",
    "            else:\n",
    "                ax.text((gene_pos[int(x[1])-1]+gene_pos[int(x[0])-1])/2,ypos[row[\"vaf_group_comparison\"]]+0.05,\"*\",horizontalalignment='center',verticalalignment='center')\n",
    "            #ax.text((gene_pos[int(x[1])-1]+gene_pos[int(x[0])-1])/2,3.1,\"*\")\n",
    "            \n",
    "\n",
    "            \n",
    "            \n",
    "        \"\"\"\n",
    "        if df.shape[0]>10:\n",
    "            box_groups.append(df)\n",
    "            if i != 0.3:\n",
    "                tick_labels.append(\"[\"+ \"{:1.1f}\".format(i)+ \",\"+\"{:1.1f}\".format(i+0.2)+ \")\\n(\"+ str(box_groups[-1].shape[0])+\")\")\n",
    "            else:\n",
    "                tick_labels.append(\"[0.3-1]\"+\"\\n(\"+ str(box_groups[-1].shape[0])+\")\")\n",
    "            box_colors.append(box_color_pal[int(i*10)])    \n",
    "            ticks.append(j+i*10*0.15)\n",
    "        gene_vaf_bin_counts.append(str(df.shape[0]))\n",
    "        vaf_mean_all.append(\"{:.2f}\".format(vaf_mean))\n",
    "        \"\"\"\n",
    "    vaf_bin_counts.append(gene+\"\\n(\"+\",\".join(gene_vaf_bin_counts)+\")\")\n",
    "    #b=plt.boxplot(box_groups,positions=np.arange(0,len(box_groups),1))\n",
    "    j+=1\n",
    "bp=plt.boxplot(box_groups,positions=ticks,widths = 0.08,flierprops = {'marker':'o','markersize':1,'markerfacecolor':'none','markeredgewidth':0.1}, boxprops = dict( linestyle='none'),whiskerprops={'lw':0.3}, medianprops={'color':'r'},capprops={'linewidth':0.5},patch_artist=True)\n",
    "\n",
    "        \n",
    "\n",
    "ax=plt.gca()\n",
    "xticks=ax.get_xticklabels()\n",
    "plt.xticks(np.arange(-1,14,1)+0.5,[\"Control\",\"ASXL1\",\"CALR\",\"CBL\",\"DNMT3A_R882\",\"DNMT3A_other\",\"GNAS\",\"GNB1\",\"JAK2\",\"PPM1D\",\"SF3B1\",\"SRSF2\",\"TET2\",\"TP53\",\"U2AF1\"],fontsize=6)\n",
    "plt.xticks(np.arange(-1,14,1)+0.5,vaf_bin_counts,fontsize=6)\n",
    "i=0\n",
    "for patch in bp['boxes']:\n",
    "        patch.set(facecolor=box_colors[i]) \n",
    "        patch.set(edgecolor=box_colors[i])\n",
    "        i+=1\n",
    "\n",
    "plt.ylim([-5,5])\n",
    "#    plt.title(gene+\"(\"+str(temp.shape[0])+\")\")\n",
    "#    plt.show()\n",
    "\n",
    "legend_patch = [mpatches.Patch(color=box_color_pal[0], label='<=0.1'),mpatches.Patch(color=box_color_pal[1], label='0.1-0.2'),mpatches.Patch(color=box_color_pal[2], label='>0.2')]\n",
    "plt.legend(handles=legend_patch,title=\"VAF bins\",fontsize=6,title_fontsize=6,ncol=2,loc='upper left',bbox_to_anchor=(0.04, 0.995))\n",
    "plt.yticks(fontsize=8)\n",
    "plt.axhline(y=data.loc[data[\"single_gene\"]==\"Control\",y].median(), color='k', linestyle='--',linewidth=0.5)\n",
    "plt.ylabel(\"Telomere length\",fontsize=10)\n",
    "ax.text(0.2,0.95,\"*   P<0.05\\n**   P<0.01\\n***  P<0.001\",fontsize=6,transform=ax.transAxes)\n",
    "\n",
    "#plt.savefig(\"../figures_revision1/vaf_vs_tl_multibin_reviewer1_noHM.png\",bbox_inches=\"tight\",dpi=600)\n",
    "#plt.savefig(\"../figures_revision1/vaf_vs_tl_multibin_reviewer1_noHM.pdf\",bbox_inches=\"tight\")\n",
    "#plt.savefig(\"../figures_revision1/vaf_vs_tl_multibin_reviewer1_noHM.svg\",bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f9d4ba9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bin_vaf(v):\n",
    "    if v != \"NA\":\n",
    "        if v<=0.1:\n",
    "            return 1\n",
    "        elif v>0.1 and v<=0.2:\n",
    "            return 2\n",
    "        else:\n",
    "            return 3\n",
    "        \n",
    "data[\"vaf_group\"]=data[\"single_gene_vaf\"].map(lambda row:bin_vaf(row))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b81d7ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "#data[[\"eid\",\"residuals\",\"single_gene\",\"single_gene_vaf\",\"vaf_group\"]].to_csv(\"ltl_residual_multibin_vaf_singlegene.tsv\",sep=\"\\t\",index=False,na_rep=\"NA\")\n",
    "\n",
    "\n",
    "#data[[\"eid\",\"residuals\",\"single_gene\",\"single_gene_vaf\",\"vaf_group\"]].to_csv(\"ltl_residual_multibin_vaf_singlegene_noHM.tsv\",sep=\"\\t\",index=False,na_rep=\"NA\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cb1c330",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "######## LTL~PRS+VAF+PRS*VAF ########## \n",
    "####### gene specific models for LTL ##############\n",
    "\n",
    "temp=ukb500k_ts_som.copy(deep=True)\n",
    "y=\"tl\"\n",
    "temp= temp.rename(columns={'Z-adjusted T/S log': 'tl'})\n",
    "temp=temp.loc[(temp[y].notna())]\n",
    "\n",
    "data=temp[[\"total_PRS\", \"sex\", \"age\", \"smoking_status\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\",\"single_gene\",\"single_gene_vaf\",\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",\"tl\",\"HM\"]].copy(deep=True)\n",
    "data=data.dropna()\n",
    "data=data[(data[\"HM\"]==0)]\n",
    "total_prs=np.arange(-10,10.25,0.25)\n",
    "vaf_values=np.arange(0.05,1.05,0.05)\n",
    "inputs_base=pd.DataFrame(list(product(total_prs,vaf_values)))\n",
    "inputs_base.columns=[\"total_PRS\",\"single_gene_vaf\"]\n",
    "\n",
    "for col in data.columns:\n",
    "    if col not in ([\"total_PRS\",\"single_gene_vaf\",\"tl\",\"sex\",\"smoking_status\",\"single_gene\",\"HM\"]):\n",
    "        inputs_base[col]=data[col].mean()\n",
    "print(inputs_base.columns)\n",
    "all_inputs=[]\n",
    "\n",
    "for values in data[\"sex\"].unique():\n",
    "    inputs_base[\"sex\"]=values\n",
    "    for ss in data[\"smoking_status\"].unique():\n",
    "        inputs_base[\"smoking_status\"]=ss\n",
    "        all_inputs.append(inputs_base.copy(deep=True))\n",
    "all_inputs=pd.concat(all_inputs).reset_index(drop=True)\n",
    "\n",
    "\n",
    "formula_0 = 'tl ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\"])\n",
    "\n",
    "formula_1 = 'tl ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10','total_PRS',\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\"])\n",
    "\n",
    "formula_2 = 'tl ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"single_gene_vaf\",\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\"])\n",
    "\n",
    "formula_3 = 'tl ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10','total_PRS',\"single_gene_vaf\",\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\"])\n",
    "\n",
    "formula_4 = 'tl ~' + ' + '.join(['C(sex,Treatment(reference=\"Female\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10','total_PRS',\"single_gene_vaf\",\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\"]) + \"+ total_PRS * single_gene_vaf\"\n",
    "\n",
    "gene_list=list(temp[\"single_gene\"].unique())\n",
    "gene_list.remove(\"Control\")\n",
    "gene_order={\"DNMT3A_other\":0,\"TET2\":1,\"SRSF2\":2,\"PPM1D\":3}\n",
    "\n",
    "coef_list_all=[]\n",
    "adj_r_all=[]\n",
    "g=0\n",
    "max_tl,min_tl=[],[]\n",
    "tl_sens_all={}\n",
    "for gene in gene_list:\n",
    "    data=temp[[\"total_PRS\", \"sex\", \"age\", \"smoking_status\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\",\"single_gene\",\"single_gene_vaf\",\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",\"tl\",\"HM\"]].copy(deep=True)\n",
    "    data=data.dropna()\n",
    "    data=data[(data[\"HM\"]==0)]\n",
    "    \n",
    "    print(data.shape)\n",
    "    data=data[(data[\"single_gene\"]==gene)]\n",
    "    data[\"single_gene_vaf\"]=data[\"single_gene_vaf\"].astype(float)\n",
    "    data[\"total_PRS\"]=data[\"total_PRS\"].astype(float)\n",
    "    print(data[\"single_gene\"].value_counts())\n",
    "    \n",
    "    adj_r=[gene]\n",
    "    coef_list=[gene]\n",
    "    \n",
    "    lin_model=smf.ols(formula=formula_0,data=data)\n",
    "    result=lin_model.fit(max_iterations= 100000)\n",
    "    display(result.summary2())\n",
    "    params = result.params\n",
    "    conf = result.conf_int()\n",
    "    print(result.pvalues)\n",
    "    adj_r+=[result.rsquared_adj]\n",
    "    \n",
    "    \n",
    "    lin_model=smf.ols(formula=formula_1,data=data)\n",
    "    result=lin_model.fit(max_iterations= 100000)\n",
    "    display(result.summary2())\n",
    "    params = result.params\n",
    "    conf = result.conf_int()\n",
    "    adj_r+=[result.rsquared_adj]\n",
    "    coef_list+=[params[\"total_PRS\"],conf.loc[\"total_PRS\",0],conf.loc[\"total_PRS\",1],result.pvalues[\"total_PRS\"]]\n",
    "    \n",
    "    lin_model=smf.ols(formula=formula_2,data=data)\n",
    "    result=lin_model.fit(max_iterations= 100000)\n",
    "    display(result.summary2())\n",
    "    params = result.params\n",
    "    conf = result.conf_int()\n",
    "    adj_r+=[result.rsquared_adj]\n",
    "    coef_list+=[params[\"single_gene_vaf\"],conf.loc[\"single_gene_vaf\",0],conf.loc[\"single_gene_vaf\",1],result.pvalues[\"single_gene_vaf\"]]\n",
    "    \n",
    "    lin_model_3=smf.ols(formula=formula_3,data=data)\n",
    "    result_3=lin_model_3.fit(max_iterations= 100000)\n",
    "    display(result_3.summary2())\n",
    "    params = result_3.params\n",
    "    print(params)\n",
    "    conf = result_3.conf_int()\n",
    "    adj_r+=[result_3.rsquared_adj]\n",
    "    coef_list+=[params[\"total_PRS\"],conf.loc[\"total_PRS\",0],conf.loc[\"total_PRS\",1],result_3.pvalues[\"total_PRS\"],params[\"single_gene_vaf\"],conf.loc[\"single_gene_vaf\",0],conf.loc[\"single_gene_vaf\",1],result_3.pvalues[\"single_gene_vaf\"]]\n",
    "    \n",
    "    lin_model_4=smf.ols(formula=formula_4,data=data)\n",
    "    result_4=lin_model_4.fit(max_iterations= 100000)\n",
    "    display(result_4.summary2())\n",
    "    params = result_4.params\n",
    "    conf = result_4.conf_int()\n",
    "    coef_list+=[params[\"total_PRS\"],conf.loc[\"total_PRS\",0],conf.loc[\"total_PRS\",1],result_4.pvalues[\"total_PRS\"],params[\"single_gene_vaf\"],conf.loc[\"single_gene_vaf\",0],conf.loc[\"single_gene_vaf\",1],result_4.pvalues[\"single_gene_vaf\"],params[\"total_PRS:single_gene_vaf\"],conf.loc[\"total_PRS:single_gene_vaf\",0],conf.loc[\"total_PRS:single_gene_vaf\",1],result_4.pvalues[\"total_PRS:single_gene_vaf\"]]\n",
    "    anovaResults = anova_lm(result_3, result_4)\n",
    "    adj_r+=[result_4.rsquared_adj,anovaResults.loc[1,\"Pr(>F)\"]]\n",
    "    print(anovaResults)\n",
    "    adj_r_all.append(adj_r)\n",
    "    coef_list_all.append(coef_list)\n",
    "    \n",
    "    if gene in [\"DNMT3A_other\",\"TET2\",\"SRSF2\",\"PPM1D\"]:\n",
    "        tl_pred=(result_3.predict(all_inputs)).reset_index(drop=True)\n",
    "        max_tl.append(tl_pred.max())\n",
    "        min_tl.append(tl_pred.min())\n",
    "        tl_sens=pd.concat([all_inputs,tl_pred],axis=1).rename(columns={0:\"pred\"})\n",
    "        tl_sens=tl_sens.groupby([\"total_PRS\",\"single_gene_vaf\"]).agg({\"pred\":'mean'}).reset_index()\n",
    "        tl_sens=tl_sens.pivot_table(columns=[\"total_PRS\"],index=[\"single_gene_vaf\"])\n",
    "        tl_sens_all[gene]=tl_sens\n",
    "\n",
    "adj_r_all=pd.DataFrame(adj_r_all)\n",
    "adj_r_all.columns=[\"Gene\",\"Basic\",\"PRS\",\"VAF\",\"PRS+VAF\",\"PRS_VAF_interaction\",\"ANOVA\"]\n",
    "\n",
    "coef_list_all=pd.DataFrame(coef_list_all)\n",
    "coef_list_all.columns=[\"Gene\",\"PRS_only_coef\",\"PRS_only_coef_lowerci\",\"PRS_only_coef_higherci\",\"PRS_only_pvalue\",\"VAF_only_coef\",\"VAF_only_coef_lowerci\",\"VAF_only_coef_higherci\",\"VAF_only_pvalue\",\"PRS_plus_VAF_PRS_coef\",\"PRS_plus_VAF_PRS_coef_lowerci\",\"PRS_plus_VAF_PRS_coef_higherci\",\"PRS_plus_VAF_PRS_pvalue\",\"PRS_plus_VAF_VAF_coef\",\"PRS_plus_VAF_VAF_coef_lowerci\",\"PRS_plus_VAF_VAF_coef_higherci\",\"PRS_plus_VAF_VAF_pvalue\",\"PRS_VAF_int_PRS_coef\",\"PRS_VAF_int_PRS_coef_lowerci\",\"PRS_VAF_int_PRS_coef_higherci\",\"PRS_VAF_int_PRS_pvalue\",\"PRS_VAF_int_VAF_coef\",\"PRS_VAF_int_VAF_coef_lowerci\",\"PRS_VAF_int_VAF_coef_higherci\",\"PRS_VAF_int_VAF_pvalue\",\"PRS_VAF_int_PRS:VAF_coef\",\"PRS_VAF_int_PRS:VAF_coef_lowerci\",\"PRS_VAF_int_PRS:VAF_coef_higherci\",\"PRS_VAF_int_PRS:VAF_pvalue\"]\n",
    "\n",
    "display(coef_list_all)\n",
    "\n",
    "display(adj_r_all)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c84aac2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Sensitivity analysis figure ##############\n",
    "\n",
    "fig, axes = plt.subplots(2, 2,sharey=True,sharex=True,figsize=(12,12))#,gridspec_kw={'height_ratios':[5,5,1.5],'width_ratios':[15,4]})0\n",
    "axes=[axes[0,0],axes[0,1],axes[1,0],axes[1,1]]\n",
    "cbar_ax = fig.add_axes([0.93, 0.15, 0.02, 0.7])        \n",
    "\n",
    "cmap = matplotlib.cm.YlGnBu_r\n",
    "norm = matplotlib.colors.Normalize(vmin=min(min_tl), vmax=max(max_tl))\n",
    "\n",
    "fig.colorbar(matplotlib.cm.ScalarMappable(norm=norm, cmap=cmap),\n",
    "             cax=cbar_ax,label='LTL')\n",
    "\n",
    "\n",
    "for g,gene in enumerate([\"DNMT3A_other\",\"TET2\",\"SRSF2\",\"PPM1D\"]):\n",
    "    #if gene != \"PPM1D\":\n",
    "    cbar=sns.heatmap(np.asarray(tl_sens_all[gene][::-1]),ax=axes[g], cbar=False,cmap=cmap,norm=norm, clip_on=False,linecolor=None,linewidths=0.0, rasterized=True)\n",
    "    #else:\n",
    "        #cbar=sns.heatmap(np.asarray(tl_sens_all[gene][::-1]),ax=axes[g],cmap=cmap,norm=norm,clip_on=False,cbar=False)\n",
    "    axes[g].set_xticks(np.arange(0,84,8)+0.5,np.arange(-10,11,2))\n",
    "    axes[g].set_yticks(np.arange(0,20,4)+0.5,[1,0.8,0.6,0.4,0.2])\n",
    "    #fig.colorbar(cbar,ax=cbar_ax)\n",
    "    axes[g].set_title(gene)\n",
    "\n",
    "fig.add_subplot(111, frameon=False)\n",
    "# hide tick and tick label of the big axes\n",
    "plt.tick_params(labelcolor='none', top=False, bottom=False, left=False, right=False)\n",
    "plt.grid(False)\n",
    "plt.xlabel(\"PRS\",labelpad=20,fontsize=20)\n",
    "plt.ylabel(\"VAF\",labelpad=20,fontsize=20)\n",
    "\n",
    "plt.savefig(\"../figures_revision1/sensitivity_analysis.pdf\",bbox_inches=\"tight\",dpi=600)\n",
    "plt.savefig(\"../figures_revision1/sensitivity_analysis.svg\",bbox_inches=\"tight\")\n",
    "plt.savefig(\"../figures_revision1/sensitivity_analysis.png\",bbox_inches=\"tight\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "818a1232",
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_r_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b55271de",
   "metadata": {},
   "outputs": [],
   "source": [
    "for pval in [\"PRS_only_pvalue\",\"PRS_plus_VAF_PRS_pvalue\",\"PRS_plus_VAF_VAF_pvalue\",\"PRS_VAF_int_PRS_pvalue\",\"PRS_VAF_int_VAF_pvalue\",\"PRS_VAF_int_PRS:VAF_pvalue\"]:\n",
    "    coef_list_all[pval+\"_FDR\"]=multipletests(coef_list_all[pval], alpha=0.05, method='fdr_bh',returnsorted=False,is_sorted=False)[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f790e4c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#adj_r_all.to_csv(\"adj_r_singlegene_noHM.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d432460a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#coef_list_all[[\"Gene\",\"PRS_only_coef\",\"PRS_only_pvalue\",\"VAF_only_coef\",\"VAF_only_pvalue\",\"PRS_plus_VAF_PRS_coef\",\"PRS_plus_VAF_PRS_pvalue\",\"PRS_plus_VAF_VAF_coef\",\"PRS_plus_VAF_VAF_pvalue\",\"PRS_VAF_int_PRS_coef\",\"PRS_VAF_int_PRS_pvalue\",\"PRS_VAF_int_VAF_coef\",\"PRS_VAF_int_VAF_pvalue\",\"PRS_VAF_int_PRS:VAF_coef\",\"PRS_VAF_int_PRS:VAF_pvalue\"]].to_csv(\"prs_vaf_interaction_singlegene_noHM.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6b27bd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "coef_list_all[[\"Gene\",\"PRS_only_coef\",\"PRS_only_pvalue\",\"PRS_only_pvalue_FDR\",\"PRS_plus_VAF_PRS_coef\",\"PRS_plus_VAF_PRS_pvalue\",\"PRS_plus_VAF_PRS_pvalue_FDR\",\"PRS_plus_VAF_VAF_coef\",\"PRS_plus_VAF_VAF_pvalue\",\"PRS_plus_VAF_VAF_pvalue_FDR\",\"PRS_VAF_int_PRS_coef\",\"PRS_VAF_int_PRS_pvalue\",\"PRS_VAF_int_PRS_pvalue_FDR\",\"PRS_VAF_int_VAF_coef\",\"PRS_VAF_int_VAF_pvalue\",\"PRS_VAF_int_VAF_pvalue_FDR\",\"PRS_VAF_int_PRS:VAF_coef\",\"PRS_VAF_int_PRS:VAF_pvalue\", \"PRS_VAF_int_PRS:VAF_pvalue_FDR\"]].to_csv(\"prs_vaf_interaction_singlegene_noHM_FDR.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff8b1087",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "########### Double mutant analysis ###########\n",
    "########### Regression for LTL~covariates so as to get the adjusted LTL ##########\n",
    "y='Z-adjusted T/S log'\n",
    "temp=ukb500k_ts_som_multigene[[\"Z-adjusted T/S log\", \"sex\", \"age\", \"smoking_status\", \"PC1\", \"PC2\", \"PC3\", \"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\",\"multi_gene_TET2\",\"total_PRS\",\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",\"eid\",\"HM\"]].copy(deep=True)\n",
    "print(temp.shape)\n",
    "\n",
    "formula = 'tl~' + ' + '.join(['C(sex,Treatment(reference=\"Male\"))','age','C(smoking_status,Treatment(reference=\"Never\"))','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10',\"WBC_wins\",\"MO_perc_wins\",\"LY_perc_wins\",\"BA_perc_wins\",\"NE_perc_wins\",\"EO_perc_wins\",\"total_PRS\"])\n",
    "\n",
    "\n",
    "temp=temp.rename(columns={\"Z-adjusted T/S log\":\"tl\"})\n",
    "temp=temp.dropna()\n",
    "temp=temp[(temp[\"HM\"]==0)]\n",
    "lin_model=smf.ols(formula=formula,data=temp)\n",
    "result=lin_model.fit(max_iterations= 100000)\n",
    "display(result.summary2())\n",
    "params = result.params\n",
    "conf = result.conf_int()\n",
    "print(result.pvalues)\n",
    "temp=temp.merge(result.resid.rename(\"tl_residuals\"),left_index=True,right_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6eddcdd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "####### Get the differences in the VAF  between DNMT3A and TET2 clones ########\n",
    "\n",
    "def get_vaf_diff(data_ukb):\n",
    "    d=data_ukb.copy(deep=True)\n",
    "    d=d[(d[\"multi_gene_TET2\"]==\"DNMT3A,TET2-TET2\") | (d[\"multi_gene_TET2\"]==\"DNMT3A,TET2-DNMT3A\")]\n",
    "    d[\"tet2_dnmt3a_vaf\"]=0\n",
    "    d[\"dnmt3a_tet2_vaf\"]=0\n",
    "    d[\"dnmt3a_tet2_diff\"]=0\n",
    "    for index, row in d.iterrows():\n",
    "        v=np.array(row[\"vaf\"].split(\",\")).astype(float)\n",
    "        all_genes_present=np.array(row[\"symbol\"].split(\",\"))\n",
    "        d.loc[index,\"tet2_dnmt3a_vaf\"]=np.max(v[all_genes_present==\"DNMT3A\"])\n",
    "        d.loc[index,\"dnmt3a_tet2_vaf\"]=np.max(v[all_genes_present==\"TET2\"])\n",
    "        d.loc[index,\"dnmt3a_tet2_diff\"]=np.max(v[all_genes_present==\"DNMT3A\"])-np.max(v[all_genes_present==\"TET2\"])\n",
    "    data_ukb=data_ukb.merge(d[[\"eid\",\"dnmt3a_tet2_vaf\",\"tet2_dnmt3a_vaf\",\"dnmt3a_tet2_diff\"]],on=\"eid\",how='left')\n",
    "    \n",
    "    d=data_ukb.copy(deep=True)\n",
    "    d=d[(d[\"multi_gene_TET2\"]==\"SRSF2,TET2-TET2\") | (d[\"multi_gene_TET2\"]==\"SRSF2,TET2-SRSF2\")]\n",
    "    d[\"tet2_srsf2_vaf\"]=0\n",
    "    d[\"srsf2_tet2_vaf\"]=0\n",
    "    d[\"srsf2_tet2_diff\"]=0\n",
    "    for index, row in d.iterrows():\n",
    "        v=np.array(row[\"vaf\"].split(\",\")).astype(float)\n",
    "        all_genes_present=np.array(row[\"symbol\"].split(\",\"))\n",
    "        d.loc[index,\"tet2_srsf2_vaf\"]=np.max(v[all_genes_present==\"SRSF2\"])\n",
    "        d.loc[index,\"srsf2_tet2_vaf\"]=np.max(v[all_genes_present==\"TET2\"])\n",
    "        d.loc[index,\"srsf2_tet2_diff\"]=np.max(v[all_genes_present==\"SRSF2\"])-np.max(v[all_genes_present==\"TET2\"])\n",
    "    data_ukb=data_ukb.merge(d[[\"eid\",\"srsf2_tet2_vaf\",\"tet2_srsf2_vaf\",\"srsf2_tet2_diff\"]],on=\"eid\",how='left')\n",
    "    return data_ukb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0944590a",
   "metadata": {},
   "outputs": [],
   "source": [
    "vaf_diff=get_vaf_diff(ukb500k_ts_som_multigene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14732365",
   "metadata": {},
   "outputs": [],
   "source": [
    "########### Plot the adjusted residuals ############\n",
    "fig, axes = plt.subplots(1, 2,sharex=False,sharey=True,figsize=(10,4))\n",
    "eid_sel=vaf_diff.loc[((vaf_diff[\"srsf2_tet2_diff\"]<=-0.035) | (vaf_diff[\"srsf2_tet2_diff\"]>=0.035)) & (vaf_diff[\"srsf2_tet2_vaf\"]<0.6) & (vaf_diff[\"tet2_srsf2_vaf\"]<0.6),\"eid\"]\n",
    "\n",
    "bp = temp[((temp[\"multi_gene_TET2\"]==\"SRSF2,TET2-SRSF2\") | (temp[\"multi_gene_TET2\"]==\"SRSF2,TET2-TET2\")) & (temp[\"eid\"].isin(eid_sel))].boxplot(column='tl_residuals', by='multi_gene_TET2', grid=False,ax=axes[0],widths=0.5,boxprops={'color':'#b3cce6','facecolor':'#b3cce6'}, patch_artist=True,zorder=1,medianprops={'color':'r'},whiskerprops={'color':'k'})\n",
    "g=[\"SRSF2,TET2-SRSF2\",\"SRSF2,TET2-TET2\"]\n",
    "\n",
    "#vaf_diff.loc[((vaf_diff[\"srsf2_tet2_diff\"]<=-0.035) | (vaf_diff[\"srsf2_tet2_diff\"]>=0.035)) & (vaf_diff[\"srsf2_tet2_vaf\"]<0.6) & (vaf_diff[\"tet2_srsf2_vaf\"]<0.6) & (vaf_diff[\"eid\"].isin(temp[\"eid\"])),[\"eid\",\"symbol\",\"vaf\",\"srsf2_tet2_vaf\",\"tet2_srsf2_vaf\",\"srsf2_tet2_diff\"]].to_csv(\"srsf2_tet2_data.tsv\",sep=\"\\t\",index=False,header=[\"eid\",\"symbol\",\"vaf\",\"tet2_vaf\",\"srsf2_vaf\",\"srsf2_tet2_diff\"])\n",
    "\n",
    "for i in range(1,3):\n",
    "    y = temp.loc[(temp[\"multi_gene_TET2\"]==g[i-1]) & (temp[\"eid\"].isin(eid_sel)),\"tl_residuals\"].dropna()\n",
    "    print(y.shape)\n",
    "    np.random.seed(0)\n",
    "    # Add some random \"jitter\" to the x-axis\n",
    "    z = np.random.normal(i, 0.06, size=len(y))\n",
    "    #axes[0].scatter(z, y, alpha=0.5,s=20,edgecolor='#800000',facecolors='None',linewidths=1.5)\n",
    "axes[0].set_xticks([1,2])\n",
    "axes[0].set_xticklabels([\"SRSF2\",\"TET2\"],fontsize=11)\n",
    "axes[0].set_ylim([-4,4])\n",
    "axes[0].set_ylabel(\"LTL\",fontsize=13)\n",
    "axes[0].set_xlabel(\"Gene with larger clone\",fontsize=13)\n",
    "axes[0].set_title(\"\")\n",
    "axes[0].text(2,3,\"P=0.01\",fontsize=12)\n",
    "axes[0].text(0.3,4.5,\"a\",fontsize=15,fontweight=\"bold\")\n",
    "\n",
    "\n",
    "\n",
    "eid_sel=vaf_diff.loc[((vaf_diff[\"dnmt3a_tet2_diff\"]<=-0.035) | (vaf_diff[\"dnmt3a_tet2_diff\"]>=0.035)) & (vaf_diff[\"dnmt3a_tet2_vaf\"]<0.6) & (vaf_diff[\"tet2_dnmt3a_vaf\"]<0.6),\"eid\"]\n",
    "bp = temp[((temp[\"multi_gene_TET2\"]==\"DNMT3A,TET2-DNMT3A\") | (temp[\"multi_gene_TET2\"]==\"DNMT3A,TET2-TET2\")) & (temp[\"eid\"].isin(eid_sel))].boxplot(column='tl_residuals', by='multi_gene_TET2', grid=False,ax=axes[1],widths=0.5,boxprops={'color':'#b3cce6','facecolor':'#b3cce6'}, patch_artist=True,zorder=1,medianprops={'color':'r'},whiskerprops={'color':'k'},)\n",
    "\n",
    "print(temp.loc[((temp[\"multi_gene_TET2\"]==\"DNMT3A,TET2-DNMT3A\") | (temp[\"multi_gene_TET2\"]==\"DNMT3A,TET2-TET2\")) & (temp[\"eid\"].isin(eid_sel)),\"multi_gene_TET2\"].value_counts())\n",
    "\n",
    "g=[\"DNMT3A,TET2-DNMT3A\",\"DNMT3A,TET2-TET2\"]\n",
    "\n",
    "#vaf_diff.loc[((vaf_diff[\"dnmt3a_tet2_diff\"]<=-0.035) | (vaf_diff[\"dnmt3a_tet2_diff\"]>=0.035)) & (vaf_diff[\"dnmt3a_tet2_vaf\"]<0.6) & (vaf_diff[\"tet2_dnmt3a_vaf\"]<0.6) & (vaf_diff[\"eid\"].isin(temp[\"eid\"])),[\"eid\",\"symbol\",\"vaf\",\"dnmt3a_tet2_vaf\",\"tet2_dnmt3a_vaf\",\"dnmt3a_tet2_diff\"]].to_csv(\"dnmt3a_tet2_data.tsv\",sep=\"\\t\",index=False,header=[\"eid\",\"symbol\",\"vaf\",\"tet2_vaf\",\"dnmt3a_vaf\",\"dnmt3a_tet2_diff\"])\n",
    "             \n",
    "for i in range(1,3):\n",
    "    y = temp.loc[(temp[\"multi_gene_TET2\"]==g[i-1]) & (temp[\"eid\"].isin(eid_sel)),\"tl_residuals\"].dropna()\n",
    "    np.random.seed(0)\n",
    "    # Add some random \"jitter\" to the x-axis\n",
    "    z = np.random.normal(i, 0.06, size=len(y))\n",
    "    #axes[1].scatter(z, y, alpha=0.5,s=20,edgecolor='#800000',facecolors='None',linewidths=1.5)\n",
    "axes[1].set_xticks([1,2])\n",
    "axes[1].set_xticklabels([\"DNMT3A\",\"TET2\"],fontsize=11)\n",
    "axes[1].set_xlabel(\"Gene with larger clone\",fontsize=13)\n",
    "axes[1].set_title(\"\")\n",
    "axes[1].text(2,3,\"P=0.90\",fontsize=12)\n",
    "axes[1].text(0.4,4.5,\"b\",fontsize=15,fontweight=\"bold\")\n",
    "#axes[1].set_yticks(np.arange(-4,5,1))\n",
    "axes[1].set_facecolor('white')\n",
    "fig.suptitle('')\n",
    "plt.subplots_adjust(wspace=0.1)\n",
    "\n",
    "#plt.savefig(\"../figures_revision1/double_mut_comparison_noHM.pdf\",bbox_inches=\"tight\")\n",
    "#plt.savefig(\"../figures_revision1/double_mut_comparison.png\",bbox_inches=\"tight\",dpi=600)\n",
    "#plt.savefig(\"../figures_revision1/double_mut_comparison.svg\",bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0796533f",
   "metadata": {},
   "outputs": [],
   "source": [
    "################ Compare the adjusted LTLs - srsf2, tet2 ##############\n",
    "eid_sel=vaf_diff.loc[((vaf_diff[\"srsf2_tet2_diff\"]<=-0.035) | (vaf_diff[\"srsf2_tet2_diff\"]>=0.035)) & (vaf_diff[\"srsf2_tet2_vaf\"]<0.6) & (vaf_diff[\"tet2_srsf2_vaf\"]<0.6),\"eid\"]\n",
    "print(eid_sel.shape)\n",
    "print(temp.loc[((temp[\"multi_gene_TET2\"]==\"SRSF2,TET2-SRSF2\") & (temp[\"eid\"].isin(eid_sel))),\"tl_residuals\"].shape,temp.loc[((temp[\"multi_gene_TET2\"]==\"SRSF2,TET2-TET2\") & (temp[\"eid\"].isin(eid_sel))),\"tl_residuals\"].shape)\n",
    "\n",
    "mannwhitneyu(temp.loc[((temp[\"multi_gene_TET2\"]==\"SRSF2,TET2-SRSF2\") & (temp[\"eid\"].isin(eid_sel))),\"tl_residuals\"],temp.loc[((temp[\"multi_gene_TET2\"]==\"SRSF2,TET2-TET2\") & (temp[\"eid\"].isin(eid_sel))),\"tl_residuals\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90b85fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "################ Compare the adjusted LTLs - dnmt3a, tet2 ##############\n",
    "\n",
    "eid_sel=vaf_diff.loc[((vaf_diff[\"dnmt3a_tet2_diff\"]<=-0.035) | (vaf_diff[\"dnmt3a_tet2_diff\"]>=0.035)) & (vaf_diff[\"dnmt3a_tet2_vaf\"]<0.6) & (vaf_diff[\"tet2_dnmt3a_vaf\"]<0.6),\"eid\"]\n",
    "\n",
    "print(temp.loc[((temp[\"multi_gene_TET2\"]==\"DNMT3A,TET2-DNMT3A\") & (temp[\"eid\"].isin(eid_sel))),\"tl_residuals\"].shape, temp.loc[((temp[\"multi_gene_TET2\"]==\"DNMT3A,TET2-TET2\") & (temp[\"eid\"].isin(eid_sel))),\"tl_residuals\"].shape)\n",
    "\n",
    "mannwhitneyu(temp.loc[((temp[\"multi_gene_TET2\"]==\"DNMT3A,TET2-DNMT3A\") & (temp[\"eid\"].isin(eid_sel))),\"tl_residuals\"], temp.loc[((temp[\"multi_gene_TET2\"]==\"DNMT3A,TET2-TET2\") & (temp[\"eid\"].isin(eid_sel))),\"tl_residuals\"])\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
